# Feature selection

## Background

Here we apply feature selection methods to identify highly variable genes (HVGs) for further downstream analyses.


## Previous steps

*Code to run steps from the previous chapters to generate the `SpatialExperiment` object required for this chapter.*

```{r previous_steps, message=FALSE, results='hide'}
# LOAD DATA

library(SpatialExperiment)
library(STexampleData)
spe <- Visium_humanDLPFC()

# QUALITY CONTROL (QC)

library(scater)
# subset to keep only spots over tissue
spe <- spe[, spatialData(spe)$in_tissue == 1]
# identify mitochondrial genes
is_mito <- grepl("(^MT-)|(^mt-)", rowData(spe)$gene_name)
# calculate per-spot QC metrics
spe <- addPerCellQC(spe, subsets = list(mito = is_mito))
# select QC thresholds
qc_lib_size <- colData(spe)$sum < 500
qc_detected <- colData(spe)$detected < 250
qc_mito <- colData(spe)$subsets_mito_percent > 30
qc_cell_count <- colData(spe)$cell_count > 12
# combined set of discarded spots
discard <- qc_lib_size | qc_detected | qc_mito | qc_cell_count
colData(spe)$discard <- discard
# filter low-quality spots
spe <- spe[, !colData(spe)$discard]

# NORMALIZATION

library(scran)
# quick clustering for pool-based size factors
set.seed(123)
qclus <- quickCluster(spe)
# calculate size factors
spe <- computeSumFactors(spe, cluster = qclus)
# calculate logcounts (log-transformed normalized counts)
spe <- logNormCounts(spe)
```


## Highly variable genes (HVGs)

We use methods from `scran` [@Lun2016] to identify a set of top highly variable genes (HVGs), which will be used to define cell types. These methods were originally developed for single-cell RNA sequencing, so here we assume that spots can be treated as equivalent to cells, and we use only molecular features (gene expression).

First, we remove mitochondrial genes, since these are very highly expressed in this dataset, and are not of biological interest.

```{r remove_mito, message=FALSE}
# remove mitochondrial genes
spe <- spe[!is_mito, ]
dim(spe)
```


Then, apply methods from `scran`. This gives us a list of HVGs, which can be used for further downstream analyses.

```{r select_HVGs, message=FALSE, fig.height=5}
library(scran)

# fit mean-variance relationship
dec <- modelGeneVar(spe)

# visualize mean-variance relationship
fit <- metadata(dec)
plot(fit$mean, fit$var, 
     xlab = "mean of log-expression", ylab = "variance of log-expression")
curve(fit$trend(x), col = "dodgerblue", add = TRUE, lwd = 2)

# select top HVGs
top_hvgs <- getTopHVGs(dec, prop = 0.1)
length(top_hvgs)
```


## Spatially variable genes (SVGs)

Alternatively, in ST data, we can apply methods to identify spatially variable genes (SVGs) instead of HVGs.

Standard statistical measures such as [Moran's I](https://en.wikipedia.org/wiki/Moran%27s_I) can be used to rank genes by the observed spatial autocorrelation, after making some assumptions such as the bandwidth. Here, we will demonstrate some examples.

Several new statistical methods adapted to identify SVGs in ST data have also recently been developed. These include [SpatialDE](https://github.com/Teichlab/SpatialDE), [SPARK](https://xzhoulab.github.io/SPARK/), and [SPARK-X](https://xzhoulab.github.io/SPARK/). SpatialDE and SPARK do not scale to large datasets containing thousands of spatial coordinates, while SPARK-X is available from GitHub and therefore (due to Bioconductor guidelines) cannot be included in the interactive examples here. We encourage developers to contribute their packages to managed repositories (i.e. CRAN, Bioconductor, and pip), and will expand this section in the future to include additional examples as these methods become available.

A recent benchmark paper ([Li et al. 2021](https://www.biorxiv.org/content/10.1101/2021.08.27.457741v1)) has shown that integrating HVGs and SVGs can improve downstream clustering performance in ST data, confirming that SVGs contain additional biologically relevant information that is not captured by HVGs in ST data.

