<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 13 Human DLPFC workflow | Best Practices for Spatial Transcriptomics Analysis with Bioconductor</title>
  <meta name="description" content="Online book ‘Best Practices for Spatial Transcriptomics Analysis with Bioconductor’" />
  <meta name="generator" content="bookdown 0.35 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 13 Human DLPFC workflow | Best Practices for Spatial Transcriptomics Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="Online book ‘Best Practices for Spatial Transcriptomics Analysis with Bioconductor’" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 13 Human DLPFC workflow | Best Practices for Spatial Transcriptomics Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Online book ‘Best Practices for Spatial Transcriptomics Analysis with Bioconductor’" />
  



<meta name="date" content="2023-09-12" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="workflows.html"/>
<link rel="next" href="mouse-coronal-workflow.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 1em;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Best Practices ST</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#contents"><i class="fa fa-check"></i><b>1.1</b> Contents</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#who-this-book-is-for"><i class="fa fa-check"></i><b>1.2</b> Who this book is for</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#bioconductor"><i class="fa fa-check"></i><b>1.3</b> Bioconductor</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#contributions"><i class="fa fa-check"></i><b>1.4</b> Contributions</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html"><i class="fa fa-check"></i><b>2</b> Spatially-resolved transcriptomics</a>
<ul>
<li class="chapter" data-level="2.1" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#overview"><i class="fa fa-check"></i><b>2.1</b> Overview</a></li>
<li class="chapter" data-level="2.2" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#sequencing-based-platforms"><i class="fa fa-check"></i><b>2.2</b> Sequencing-based platforms</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#x-genomics-visium"><i class="fa fa-check"></i><b>2.2.1</b> 10x Genomics Visium</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#molecule-based-platforms"><i class="fa fa-check"></i><b>2.3</b> Molecule-based platforms</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#x-genomics-xenium"><i class="fa fa-check"></i><b>2.3.1</b> 10x Genomics Xenium</a></li>
<li class="chapter" data-level="2.3.2" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#vizgen-merscope"><i class="fa fa-check"></i><b>2.3.2</b> Vizgen MERSCOPE</a></li>
<li class="chapter" data-level="2.3.3" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#nanostring-cosmx"><i class="fa fa-check"></i><b>2.3.3</b> NanoString CosMx</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="bioconductor-data-classes.html"><a href="bioconductor-data-classes.html"><i class="fa fa-check"></i><b>3</b> Bioconductor data classes</a>
<ul>
<li class="chapter" data-level="3.1" data-path="bioconductor-data-classes.html"><a href="bioconductor-data-classes.html#spatialexperiment-class"><i class="fa fa-check"></i><b>3.1</b> SpatialExperiment class</a></li>
<li class="chapter" data-level="3.2" data-path="bioconductor-data-classes.html"><a href="bioconductor-data-classes.html#molecule-based-data"><i class="fa fa-check"></i><b>3.2</b> Molecule-based data</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="bioconductor-data-classes.html"><a href="bioconductor-data-classes.html#moleculeexperiment"><i class="fa fa-check"></i><b>3.2.1</b> MoleculeExperiment</a></li>
<li class="chapter" data-level="3.2.2" data-path="bioconductor-data-classes.html"><a href="bioconductor-data-classes.html#spatialfeatureexperiment"><i class="fa fa-check"></i><b>3.2.2</b> SpatialFeatureExperiment</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>II Analysis steps</b></span></li>
<li class="chapter" data-level="4" data-path="analysis-steps.html"><a href="analysis-steps.html"><i class="fa fa-check"></i><b>4</b> Analysis steps</a>
<ul>
<li class="chapter" data-level="4.1" data-path="analysis-steps.html"><a href="analysis-steps.html#load-data"><i class="fa fa-check"></i><b>4.1</b> Load data</a></li>
<li class="chapter" data-level="4.2" data-path="analysis-steps.html"><a href="analysis-steps.html#spatialexperiment-object"><i class="fa fa-check"></i><b>4.2</b> SpatialExperiment object</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>5</b> Quality control</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quality-control.html"><a href="quality-control.html#background"><i class="fa fa-check"></i><b>5.1</b> Background</a></li>
<li class="chapter" data-level="5.2" data-path="quality-control.html"><a href="quality-control.html#load-data-1"><i class="fa fa-check"></i><b>5.2</b> Load data</a></li>
<li class="chapter" data-level="5.3" data-path="quality-control.html"><a href="quality-control.html#plot-data"><i class="fa fa-check"></i><b>5.3</b> Plot data</a></li>
<li class="chapter" data-level="5.4" data-path="quality-control.html"><a href="quality-control.html#calculate-qc-metrics"><i class="fa fa-check"></i><b>5.4</b> Calculate QC metrics</a></li>
<li class="chapter" data-level="5.5" data-path="quality-control.html"><a href="quality-control.html#selecting-thresholds"><i class="fa fa-check"></i><b>5.5</b> Selecting thresholds</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="quality-control.html"><a href="quality-control.html#library-size"><i class="fa fa-check"></i><b>5.5.1</b> Library size</a></li>
<li class="chapter" data-level="5.5.2" data-path="quality-control.html"><a href="quality-control.html#number-of-expressed-features"><i class="fa fa-check"></i><b>5.5.2</b> Number of expressed features</a></li>
<li class="chapter" data-level="5.5.3" data-path="quality-control.html"><a href="quality-control.html#proportion-of-mitochondrial-reads"><i class="fa fa-check"></i><b>5.5.3</b> Proportion of mitochondrial reads</a></li>
<li class="chapter" data-level="5.5.4" data-path="quality-control.html"><a href="quality-control.html#number-of-cells-per-spot"><i class="fa fa-check"></i><b>5.5.4</b> Number of cells per spot</a></li>
<li class="chapter" data-level="5.5.5" data-path="quality-control.html"><a href="quality-control.html#remove-low-quality-spots"><i class="fa fa-check"></i><b>5.5.5</b> Remove low-quality spots</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="quality-control.html"><a href="quality-control.html#zero-cell-and-single-cell-spots"><i class="fa fa-check"></i><b>5.6</b> Zero-cell and single-cell spots</a></li>
<li class="chapter" data-level="5.7" data-path="quality-control.html"><a href="quality-control.html#quality-control-at-gene-level"><i class="fa fa-check"></i><b>5.7</b> Quality control at gene level</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="normalization.html"><a href="normalization.html"><i class="fa fa-check"></i><b>6</b> Normalization</a>
<ul>
<li class="chapter" data-level="6.1" data-path="normalization.html"><a href="normalization.html#background-1"><i class="fa fa-check"></i><b>6.1</b> Background</a></li>
<li class="chapter" data-level="6.2" data-path="normalization.html"><a href="normalization.html#previous-steps"><i class="fa fa-check"></i><b>6.2</b> Previous steps</a></li>
<li class="chapter" data-level="6.3" data-path="normalization.html"><a href="normalization.html#logcounts"><i class="fa fa-check"></i><b>6.3</b> Logcounts</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="feature-selection.html"><a href="feature-selection.html"><i class="fa fa-check"></i><b>7</b> Feature selection</a>
<ul>
<li class="chapter" data-level="7.1" data-path="feature-selection.html"><a href="feature-selection.html#background-2"><i class="fa fa-check"></i><b>7.1</b> Background</a></li>
<li class="chapter" data-level="7.2" data-path="feature-selection.html"><a href="feature-selection.html#previous-steps-1"><i class="fa fa-check"></i><b>7.2</b> Previous steps</a></li>
<li class="chapter" data-level="7.3" data-path="feature-selection.html"><a href="feature-selection.html#highly-variable-genes-hvgs"><i class="fa fa-check"></i><b>7.3</b> Highly variable genes (HVGs)</a></li>
<li class="chapter" data-level="7.4" data-path="feature-selection.html"><a href="feature-selection.html#spatially-variable-genes-svgs"><i class="fa fa-check"></i><b>7.4</b> Spatially variable genes (SVGs)</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html"><i class="fa fa-check"></i><b>8</b> Dimensionality reduction</a>
<ul>
<li class="chapter" data-level="8.1" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#background-3"><i class="fa fa-check"></i><b>8.1</b> Background</a></li>
<li class="chapter" data-level="8.2" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#previous-steps-2"><i class="fa fa-check"></i><b>8.2</b> Previous steps</a></li>
<li class="chapter" data-level="8.3" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>8.3</b> Principal component analysis (PCA)</a></li>
<li class="chapter" data-level="8.4" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#uniform-manifold-approximation-and-projection-umap"><i class="fa fa-check"></i><b>8.4</b> Uniform Manifold Approximation and Projection (UMAP)</a></li>
<li class="chapter" data-level="8.5" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#visualizations"><i class="fa fa-check"></i><b>8.5</b> Visualizations</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>9</b> Clustering</a>
<ul>
<li class="chapter" data-level="9.1" data-path="clustering.html"><a href="clustering.html#overview-1"><i class="fa fa-check"></i><b>9.1</b> Overview</a></li>
<li class="chapter" data-level="9.2" data-path="clustering.html"><a href="clustering.html#previous-steps-3"><i class="fa fa-check"></i><b>9.2</b> Previous steps</a></li>
<li class="chapter" data-level="9.3" data-path="clustering.html"><a href="clustering.html#non-spatial-clustering-on-hvgs"><i class="fa fa-check"></i><b>9.3</b> Non-spatial clustering on HVGs</a></li>
<li class="chapter" data-level="9.4" data-path="clustering.html"><a href="clustering.html#spatially-aware-clustering"><i class="fa fa-check"></i><b>9.4</b> Spatially-aware clustering</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="marker-genes.html"><a href="marker-genes.html"><i class="fa fa-check"></i><b>10</b> Marker genes</a>
<ul>
<li class="chapter" data-level="10.1" data-path="marker-genes.html"><a href="marker-genes.html#background-4"><i class="fa fa-check"></i><b>10.1</b> Background</a></li>
<li class="chapter" data-level="10.2" data-path="marker-genes.html"><a href="marker-genes.html#previous-steps-4"><i class="fa fa-check"></i><b>10.2</b> Previous steps</a></li>
<li class="chapter" data-level="10.3" data-path="marker-genes.html"><a href="marker-genes.html#marker-genes-1"><i class="fa fa-check"></i><b>10.3</b> Marker genes</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="spot-level-deconvolution.html"><a href="spot-level-deconvolution.html"><i class="fa fa-check"></i><b>11</b> Spot-level deconvolution</a>
<ul>
<li class="chapter" data-level="11.1" data-path="spot-level-deconvolution.html"><a href="spot-level-deconvolution.html#background-5"><i class="fa fa-check"></i><b>11.1</b> Background</a></li>
<li class="chapter" data-level="11.2" data-path="spot-level-deconvolution.html"><a href="spot-level-deconvolution.html#previous-steps-5"><i class="fa fa-check"></i><b>11.2</b> Previous steps</a></li>
<li class="chapter" data-level="11.3" data-path="spot-level-deconvolution.html"><a href="spot-level-deconvolution.html#number-of-cells-per-spot-1"><i class="fa fa-check"></i><b>11.3</b> Number of cells per spot</a></li>
</ul></li>
<li class="part"><span><b>III Workflows</b></span></li>
<li class="chapter" data-level="12" data-path="workflows.html"><a href="workflows.html"><i class="fa fa-check"></i><b>12</b> Workflows</a></li>
<li class="chapter" data-level="13" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html"><i class="fa fa-check"></i><b>13</b> Human DLPFC workflow</a>
<ul>
<li class="chapter" data-level="13.1" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#description-of-dataset"><i class="fa fa-check"></i><b>13.1</b> Description of dataset</a></li>
<li class="chapter" data-level="13.2" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#load-data-2"><i class="fa fa-check"></i><b>13.2</b> Load data</a></li>
<li class="chapter" data-level="13.3" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#plot-data-1"><i class="fa fa-check"></i><b>13.3</b> Plot data</a></li>
<li class="chapter" data-level="13.4" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#quality-control-qc"><i class="fa fa-check"></i><b>13.4</b> Quality control (QC)</a></li>
<li class="chapter" data-level="13.5" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#normalization-1"><i class="fa fa-check"></i><b>13.5</b> Normalization</a></li>
<li class="chapter" data-level="13.6" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#feature-selection-1"><i class="fa fa-check"></i><b>13.6</b> Feature selection</a></li>
<li class="chapter" data-level="13.7" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#spatially-aware-feature-selection"><i class="fa fa-check"></i><b>13.7</b> Spatially-aware feature selection</a></li>
<li class="chapter" data-level="13.8" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>13.8</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="13.9" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#clustering-1"><i class="fa fa-check"></i><b>13.9</b> Clustering</a></li>
<li class="chapter" data-level="13.10" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#marker-genes-2"><i class="fa fa-check"></i><b>13.10</b> Marker genes</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html"><i class="fa fa-check"></i><b>14</b> Mouse coronal workflow</a>
<ul>
<li class="chapter" data-level="14.1" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#description-of-dataset-1"><i class="fa fa-check"></i><b>14.1</b> Description of dataset</a></li>
<li class="chapter" data-level="14.2" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#load-data-3"><i class="fa fa-check"></i><b>14.2</b> Load data</a></li>
<li class="chapter" data-level="14.3" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#plot-data-2"><i class="fa fa-check"></i><b>14.3</b> Plot data</a></li>
<li class="chapter" data-level="14.4" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#quality-control-qc-1"><i class="fa fa-check"></i><b>14.4</b> Quality control (QC)</a></li>
<li class="chapter" data-level="14.5" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#normalization-2"><i class="fa fa-check"></i><b>14.5</b> Normalization</a></li>
<li class="chapter" data-level="14.6" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#feature-selection-2"><i class="fa fa-check"></i><b>14.6</b> Feature selection</a></li>
<li class="chapter" data-level="14.7" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#spatially-aware-feature-selection-1"><i class="fa fa-check"></i><b>14.7</b> Spatially-aware feature selection</a></li>
<li class="chapter" data-level="14.8" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#dimensionality-reduction-2"><i class="fa fa-check"></i><b>14.8</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="14.9" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#clustering-2"><i class="fa fa-check"></i><b>14.9</b> Clustering</a></li>
<li class="chapter" data-level="14.10" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#marker-genes-3"><i class="fa fa-check"></i><b>14.10</b> Marker genes</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html"><i class="fa fa-check"></i><b>15</b> spatialLIBD workflow</a>
<ul>
<li class="chapter" data-level="15.1" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#overview-2"><i class="fa fa-check"></i><b>15.1</b> Overview</a></li>
<li class="chapter" data-level="15.2" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#spatiallibd"><i class="fa fa-check"></i><b>15.2</b> spatialLIBD</a>
<ul>
<li class="chapter" data-level="15.2.1" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#why-use-spatiallibd"><i class="fa fa-check"></i><b>15.2.1</b> Why use spatialLIBD?</a></li>
<li class="chapter" data-level="15.2.2" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#want-to-learn-more-about-spatiallibd"><i class="fa fa-check"></i><b>15.2.2</b> Want to learn more about spatialLIBD?</a></li>
</ul></li>
<li class="chapter" data-level="15.3" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#code-prerequisites"><i class="fa fa-check"></i><b>15.3</b> Code prerequisites</a></li>
<li class="chapter" data-level="15.4" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#prepare-for-spatiallibd"><i class="fa fa-check"></i><b>15.4</b> Prepare for spatialLIBD</a>
<ul>
<li class="chapter" data-level="15.4.1" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#basic-information"><i class="fa fa-check"></i><b>15.4.1</b> Basic information</a></li>
<li class="chapter" data-level="15.4.2" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#gene-annotation"><i class="fa fa-check"></i><b>15.4.2</b> Gene annotation</a></li>
<li class="chapter" data-level="15.4.3" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#extra-information-and-filtering"><i class="fa fa-check"></i><b>15.4.3</b> Extra information and filtering</a></li>
</ul></li>
<li class="chapter" data-level="15.5" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#explore-the-data"><i class="fa fa-check"></i><b>15.5</b> Explore the data</a></li>
<li class="chapter" data-level="15.6" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#sharing-your-website"><i class="fa fa-check"></i><b>15.6</b> Sharing your website</a></li>
<li class="chapter" data-level="15.7" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#wrapping-up"><i class="fa fa-check"></i><b>15.7</b> Wrapping up</a></li>
<li class="chapter" data-level="" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#r-session-information"><i class="fa fa-check"></i>R session information</a></li>
</ul></li>
<li class="part"><span><b>IV Appendix</b></span></li>
<li class="chapter" data-level="16" data-path="related-resources.html"><a href="related-resources.html"><i class="fa fa-check"></i><b>16</b> Related resources</a>
<ul>
<li class="chapter" data-level="16.1" data-path="related-resources.html"><a href="related-resources.html#data-preprocessing-procedures"><i class="fa fa-check"></i><b>16.1</b> Data preprocessing procedures</a></li>
<li class="chapter" data-level="16.2" data-path="related-resources.html"><a href="related-resources.html#resources-for-other-spatially-resolved-platforms"><i class="fa fa-check"></i><b>16.2</b> Resources for other spatially-resolved platforms</a></li>
<li class="chapter" data-level="16.3" data-path="related-resources.html"><a href="related-resources.html#data-structures"><i class="fa fa-check"></i><b>16.3</b> Data structures</a></li>
<li class="chapter" data-level="16.4" data-path="related-resources.html"><a href="related-resources.html#statistical-concepts"><i class="fa fa-check"></i><b>16.4</b> Statistical concepts</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="contributors-and-acknowledgments.html"><a href="contributors-and-acknowledgments.html"><i class="fa fa-check"></i><b>17</b> Contributors and acknowledgments</a>
<ul>
<li class="chapter" data-level="17.1" data-path="contributors-and-acknowledgments.html"><a href="contributors-and-acknowledgments.html#best-practices-for-spatial-transcriptomics-analysis-with-bioconductor"><i class="fa fa-check"></i><b>17.1</b> Best Practices for Spatial Transcriptomics Analysis with Bioconductor</a></li>
<li class="chapter" data-level="17.2" data-path="contributors-and-acknowledgments.html"><a href="contributors-and-acknowledgments.html#spatialexperiment"><i class="fa fa-check"></i><b>17.2</b> SpatialExperiment</a></li>
<li class="chapter" data-level="17.3" data-path="contributors-and-acknowledgments.html"><a href="contributors-and-acknowledgments.html#github-actions-workflow"><i class="fa fa-check"></i><b>17.3</b> GitHub Actions workflow</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>18</b> References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Best Practices for Spatial Transcriptomics Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="human-dlpfc-workflow" class="section level1 hasAnchor" number="13">
<h1><span class="header-section-number">Chapter 13</span> Human DLPFC workflow<a href="human-dlpfc-workflow.html#human-dlpfc-workflow" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>This workflow analyzes one sample of human brain from the dorsolateral prefrontal cortex (DLPFC) region, measured using the 10x Genomics Visium platform. This is a condensed version of the analyses shown in the individual analysis chapters in the previous part. For more details on the individual steps, see the previous chapters.</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="human-dlpfc-workflow.html#cb120-1" tabindex="-1"></a><span class="co"># clear workspace from previous chapters</span></span>
<span id="cb120-2"><a href="human-dlpfc-workflow.html#cb120-2" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>(<span class="at">all =</span> <span class="cn">TRUE</span>))</span></code></pre></div>
<div id="description-of-dataset" class="section level2 hasAnchor" number="13.1">
<h2><span class="header-section-number">13.1</span> Description of dataset<a href="human-dlpfc-workflow.html#description-of-dataset" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This is a 10x Genomics Visium dataset generated from healthy human brain samples from the dorsolateral prefrontal cortex (DLPFC) region.</p>
<p>In the full dataset, there are 12 samples in total, from 3 individuals, with 2 pairs of spatially adjacent replicates (serial sections) per individual (4 samples per individual). The individuals and spatially adjacent replicates can be used as blocking factors. Each sample spans the six layers of the cortex plus white matter in a perpendicular tissue section.</p>
<p>For the examples in this workflow and the analysis chapters, we use a single sample from this dataset (sample 151673), to keep the computational requirements to compile the book manageable.</p>
<p>For more details on the dataset, see <span class="citation">Maynard et al. (<a href="#ref-Maynard2021">2021</a>)</span>. The full dataset is publicly available through the <a href="http://bioconductor.org/packages/spatialLIBD">spatialLIBD</a> Bioconductor package. The dataset can also be explored interactively through the <a href="http://spatial.libd.org/spatialLIBD/">spatialLIBD Shiny web app</a>.</p>
</div>
<div id="load-data-2" class="section level2 hasAnchor" number="13.2">
<h2><span class="header-section-number">13.2</span> Load data<a href="human-dlpfc-workflow.html#load-data-2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Here, we load a single sample from this dataset (sample 151673), which is available as a <code>SpatialExperiment</code> object from the <a href="https://bioconductor.org/packages/STexampleData">STexampleData</a> package.</p>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb121-1"><a href="human-dlpfc-workflow.html#cb121-1" tabindex="-1"></a><span class="fu">library</span>(SpatialExperiment)</span>
<span id="cb121-2"><a href="human-dlpfc-workflow.html#cb121-2" tabindex="-1"></a><span class="fu">library</span>(STexampleData)</span>
<span id="cb121-3"><a href="human-dlpfc-workflow.html#cb121-3" tabindex="-1"></a></span>
<span id="cb121-4"><a href="human-dlpfc-workflow.html#cb121-4" tabindex="-1"></a><span class="co"># load object</span></span>
<span id="cb121-5"><a href="human-dlpfc-workflow.html#cb121-5" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">Visium_humanDLPFC</span>()</span>
<span id="cb121-6"><a href="human-dlpfc-workflow.html#cb121-6" tabindex="-1"></a>spe</span></code></pre></div>
<pre><code>## class: SpatialExperiment 
## dim: 33538 4992 
## metadata(0):
## assays(1): counts
## rownames(33538): ENSG00000243485 ENSG00000237613 ... ENSG00000277475
##   ENSG00000268674
## rowData names(3): gene_id gene_name feature_type
## colnames(4992): AAACAACGAATAGTTC-1 AAACAAGTATCTCCCA-1 ...
##   TTGTTTGTATTACACG-1 TTGTTTGTGTAAATTC-1
## colData names(7): barcode_id sample_id ... ground_truth cell_count
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## spatialCoords names(2) : pxl_col_in_fullres pxl_row_in_fullres
## imgData names(4): sample_id image_id data scaleFactor</code></pre>
</div>
<div id="plot-data-1" class="section level2 hasAnchor" number="13.3">
<h2><span class="header-section-number">13.3</span> Plot data<a href="human-dlpfc-workflow.html#plot-data-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>As an initial check, plot the spatial coordinates (spots) in x-y dimensions on the tissue slide, to check that the object has loaded correctly and that the orientation is as expected.</p>
<p>We use visualization functions from the <a href="https://bioconductor.org/packages/ggspavis">ggspavis</a> package to generate plots.</p>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb123-1"><a href="human-dlpfc-workflow.html#cb123-1" tabindex="-1"></a><span class="fu">library</span>(ggspavis)</span></code></pre></div>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="human-dlpfc-workflow.html#cb124-1" tabindex="-1"></a><span class="co"># plot spatial coordinates (spots)</span></span>
<span id="cb124-2"><a href="human-dlpfc-workflow.html#cb124-2" tabindex="-1"></a><span class="fu">plotSpots</span>(spe)</span></code></pre></div>
<p><img src="BestPracticesST_files/figure-html/unnamed-chunk-60-1.png" width="672" /></p>
</div>
<div id="quality-control-qc" class="section level2 hasAnchor" number="13.4">
<h2><span class="header-section-number">13.4</span> Quality control (QC)<a href="human-dlpfc-workflow.html#quality-control-qc" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>First, we subset the object to keep only spots over tissue. The remaining spots are background spots, which we exclude.</p>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb125-1"><a href="human-dlpfc-workflow.html#cb125-1" tabindex="-1"></a><span class="co"># subset to keep only spots over tissue</span></span>
<span id="cb125-2"><a href="human-dlpfc-workflow.html#cb125-2" tabindex="-1"></a>spe <span class="ot">&lt;-</span> spe[, <span class="fu">colData</span>(spe)<span class="sc">$</span>in_tissue <span class="sc">==</span> <span class="dv">1</span>]</span>
<span id="cb125-3"><a href="human-dlpfc-workflow.html#cb125-3" tabindex="-1"></a><span class="fu">dim</span>(spe)</span></code></pre></div>
<pre><code>## [1] 33538  3639</code></pre>
<p>Next, calculate spot-level QC metrics using the <code>scater</code> package <span class="citation">(<a href="#ref-McCarthy2017">McCarthy et al. 2017</a>)</span>, and store the QC metrics in <code>colData</code>. See <a href="quality-control.html#quality-control">Quality control</a> for more details, including explanations of the QC metrics.</p>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="human-dlpfc-workflow.html#cb127-1" tabindex="-1"></a><span class="fu">library</span>(scater)</span></code></pre></div>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="human-dlpfc-workflow.html#cb128-1" tabindex="-1"></a><span class="co"># identify mitochondrial genes</span></span>
<span id="cb128-2"><a href="human-dlpfc-workflow.html#cb128-2" tabindex="-1"></a>is_mito <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">&quot;(^MT-)|(^mt-)&quot;</span>, <span class="fu">rowData</span>(spe)<span class="sc">$</span>gene_name)</span>
<span id="cb128-3"><a href="human-dlpfc-workflow.html#cb128-3" tabindex="-1"></a><span class="fu">table</span>(is_mito)</span></code></pre></div>
<pre><code>## is_mito
## FALSE  TRUE 
## 33525    13</code></pre>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb130-1"><a href="human-dlpfc-workflow.html#cb130-1" tabindex="-1"></a><span class="fu">rowData</span>(spe)<span class="sc">$</span>gene_name[is_mito]</span></code></pre></div>
<pre><code>##  [1] &quot;MT-ND1&quot;  &quot;MT-ND2&quot;  &quot;MT-CO1&quot;  &quot;MT-CO2&quot;  &quot;MT-ATP8&quot; &quot;MT-ATP6&quot; &quot;MT-CO3&quot; 
##  [8] &quot;MT-ND3&quot;  &quot;MT-ND4L&quot; &quot;MT-ND4&quot;  &quot;MT-ND5&quot;  &quot;MT-ND6&quot;  &quot;MT-CYB&quot;</code></pre>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="human-dlpfc-workflow.html#cb132-1" tabindex="-1"></a><span class="co"># calculate per-spot QC metrics and store in colData</span></span>
<span id="cb132-2"><a href="human-dlpfc-workflow.html#cb132-2" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">addPerCellQC</span>(spe, <span class="at">subsets =</span> <span class="fu">list</span>(<span class="at">mito =</span> is_mito))</span>
<span id="cb132-3"><a href="human-dlpfc-workflow.html#cb132-3" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colData</span>(spe), <span class="dv">3</span>)</span></code></pre></div>
<pre><code>## DataFrame with 3 rows and 13 columns
##                            barcode_id     sample_id in_tissue array_row
##                           &lt;character&gt;   &lt;character&gt; &lt;integer&gt; &lt;integer&gt;
## AAACAAGTATCTCCCA-1 AAACAAGTATCTCCCA-1 sample_151673         1        50
## AAACAATCTACTAGCA-1 AAACAATCTACTAGCA-1 sample_151673         1         3
## AAACACCAATAACTGC-1 AAACACCAATAACTGC-1 sample_151673         1        59
##                    array_col ground_truth cell_count       sum  detected
##                    &lt;integer&gt;  &lt;character&gt;  &lt;integer&gt; &lt;numeric&gt; &lt;numeric&gt;
## AAACAAGTATCTCCCA-1       102       Layer3          6      8458      3586
## AAACAATCTACTAGCA-1        43       Layer1         16      1667      1150
## AAACACCAATAACTGC-1        19           WM          5      3769      1960
##                    subsets_mito_sum subsets_mito_detected subsets_mito_percent
##                           &lt;numeric&gt;             &lt;numeric&gt;            &lt;numeric&gt;
## AAACAAGTATCTCCCA-1             1407                    13              16.6351
## AAACAATCTACTAGCA-1              204                    11              12.2376
## AAACACCAATAACTGC-1              430                    13              11.4089
##                        total
##                    &lt;numeric&gt;
## AAACAAGTATCTCCCA-1      8458
## AAACAATCTACTAGCA-1      1667
## AAACACCAATAACTGC-1      3769</code></pre>
<p>Select filtering thresholds for the QC metrics by examining distributions using histograms. For additional details, including further exploratory visualizations to select the thresholds, see <a href="quality-control.html#quality-control">Quality control</a>. Here, we use relatively relaxed thresholds, since the additional exploratory visualizations showed that more stringent thresholds tended to remove groups of spots corresponding to biologically meaningful regions.</p>
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb134-1"><a href="human-dlpfc-workflow.html#cb134-1" tabindex="-1"></a><span class="co"># histograms of QC metrics</span></span>
<span id="cb134-2"><a href="human-dlpfc-workflow.html#cb134-2" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">4</span>))</span>
<span id="cb134-3"><a href="human-dlpfc-workflow.html#cb134-3" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe)<span class="sc">$</span>sum, <span class="at">xlab =</span> <span class="st">&quot;sum&quot;</span>, <span class="at">main =</span> <span class="st">&quot;UMIs per spot&quot;</span>)</span>
<span id="cb134-4"><a href="human-dlpfc-workflow.html#cb134-4" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe)<span class="sc">$</span>detected, <span class="at">xlab =</span> <span class="st">&quot;detected&quot;</span>, <span class="at">main =</span> <span class="st">&quot;Genes per spot&quot;</span>)</span>
<span id="cb134-5"><a href="human-dlpfc-workflow.html#cb134-5" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe)<span class="sc">$</span>subsets_mito_percent, <span class="at">xlab =</span> <span class="st">&quot;percent mitochondrial&quot;</span>, <span class="at">main =</span> <span class="st">&quot;Percent mito UMIs&quot;</span>)</span>
<span id="cb134-6"><a href="human-dlpfc-workflow.html#cb134-6" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">colData</span>(spe)<span class="sc">$</span>cell_count, <span class="at">xlab =</span> <span class="st">&quot;number of cells&quot;</span>, <span class="at">main =</span> <span class="st">&quot;No. cells per spot&quot;</span>)</span></code></pre></div>
<p><img src="BestPracticesST_files/figure-html/unnamed-chunk-64-1.png" width="672" /></p>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb135-1"><a href="human-dlpfc-workflow.html#cb135-1" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>))</span>
<span id="cb135-2"><a href="human-dlpfc-workflow.html#cb135-2" tabindex="-1"></a></span>
<span id="cb135-3"><a href="human-dlpfc-workflow.html#cb135-3" tabindex="-1"></a><span class="co"># select QC thresholds</span></span>
<span id="cb135-4"><a href="human-dlpfc-workflow.html#cb135-4" tabindex="-1"></a>qc_lib_size <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe)<span class="sc">$</span>sum <span class="sc">&lt;</span> <span class="dv">600</span></span>
<span id="cb135-5"><a href="human-dlpfc-workflow.html#cb135-5" tabindex="-1"></a>qc_detected <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe)<span class="sc">$</span>detected <span class="sc">&lt;</span> <span class="dv">400</span></span>
<span id="cb135-6"><a href="human-dlpfc-workflow.html#cb135-6" tabindex="-1"></a>qc_mito <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe)<span class="sc">$</span>subsets_mito_percent <span class="sc">&gt;</span> <span class="dv">28</span></span>
<span id="cb135-7"><a href="human-dlpfc-workflow.html#cb135-7" tabindex="-1"></a>qc_cell_count <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe)<span class="sc">$</span>cell_count <span class="sc">&gt;</span> <span class="dv">10</span></span>
<span id="cb135-8"><a href="human-dlpfc-workflow.html#cb135-8" tabindex="-1"></a></span>
<span id="cb135-9"><a href="human-dlpfc-workflow.html#cb135-9" tabindex="-1"></a><span class="co"># number of discarded spots for each metric</span></span>
<span id="cb135-10"><a href="human-dlpfc-workflow.html#cb135-10" tabindex="-1"></a><span class="fu">apply</span>(<span class="fu">cbind</span>(qc_lib_size, qc_detected, qc_mito, qc_cell_count), <span class="dv">2</span>, sum)</span></code></pre></div>
<pre><code>##   qc_lib_size   qc_detected       qc_mito qc_cell_count 
##             8             7            17            90</code></pre>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb137-1"><a href="human-dlpfc-workflow.html#cb137-1" tabindex="-1"></a><span class="co"># combined set of discarded spots</span></span>
<span id="cb137-2"><a href="human-dlpfc-workflow.html#cb137-2" tabindex="-1"></a>discard <span class="ot">&lt;-</span> qc_lib_size <span class="sc">|</span> qc_detected <span class="sc">|</span> qc_mito <span class="sc">|</span> qc_cell_count</span>
<span id="cb137-3"><a href="human-dlpfc-workflow.html#cb137-3" tabindex="-1"></a><span class="fu">table</span>(discard)</span></code></pre></div>
<pre><code>## discard
## FALSE  TRUE 
##  3524   115</code></pre>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb139-1"><a href="human-dlpfc-workflow.html#cb139-1" tabindex="-1"></a><span class="co"># store in object</span></span>
<span id="cb139-2"><a href="human-dlpfc-workflow.html#cb139-2" tabindex="-1"></a><span class="fu">colData</span>(spe)<span class="sc">$</span>discard <span class="ot">&lt;-</span> discard</span></code></pre></div>
<p>Plot the set of discarded spots in the spatial x-y coordinates, to confirm that the spatial distribution of the discarded spots does not correspond to any biologically meaningful regions, which would indicate that we are removing biologically informative spots.</p>
<div class="sourceCode" id="cb140"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb140-1"><a href="human-dlpfc-workflow.html#cb140-1" tabindex="-1"></a><span class="co"># check spatial pattern of discarded spots</span></span>
<span id="cb140-2"><a href="human-dlpfc-workflow.html#cb140-2" tabindex="-1"></a><span class="fu">plotQC</span>(spe, <span class="at">type =</span> <span class="st">&quot;spots&quot;</span>, <span class="at">discard =</span> <span class="st">&quot;discard&quot;</span>)</span></code></pre></div>
<p><img src="BestPracticesST_files/figure-html/unnamed-chunk-65-1.png" width="672" /></p>
<p>There is some concentration of discarded spots at the edge of the tissue region, which may be due to tissue damage. Importantly, the discarded spots do not correspond to any of the cortical layers of interest.</p>
<p>We filter out the low-quality spots from the object.</p>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="human-dlpfc-workflow.html#cb141-1" tabindex="-1"></a><span class="co"># filter low-quality spots</span></span>
<span id="cb141-2"><a href="human-dlpfc-workflow.html#cb141-2" tabindex="-1"></a>spe <span class="ot">&lt;-</span> spe[, <span class="sc">!</span><span class="fu">colData</span>(spe)<span class="sc">$</span>discard]</span>
<span id="cb141-3"><a href="human-dlpfc-workflow.html#cb141-3" tabindex="-1"></a><span class="fu">dim</span>(spe)</span></code></pre></div>
<pre><code>## [1] 33538  3524</code></pre>
</div>
<div id="normalization-1" class="section level2 hasAnchor" number="13.5">
<h2><span class="header-section-number">13.5</span> Normalization<a href="human-dlpfc-workflow.html#normalization-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Calculate log-transformed normalized counts (logcounts) with the library size factors methodology, using methods from <code>scater</code> <span class="citation">(<a href="#ref-McCarthy2017">McCarthy et al. 2017</a>)</span> and <code>scran</code> <span class="citation">(<a href="#ref-Lun2016">Lun, McCarthy, and Marioni 2016</a>)</span>, making the assumption that spots can be treated as equivalent to cells. For more details, see <a href="normalization.html#normalization">Normalization</a>.</p>
<div class="sourceCode" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb143-1"><a href="human-dlpfc-workflow.html#cb143-1" tabindex="-1"></a><span class="fu">library</span>(scran)</span></code></pre></div>
<div class="sourceCode" id="cb144"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb144-1"><a href="human-dlpfc-workflow.html#cb144-1" tabindex="-1"></a><span class="co"># calculate library size factors</span></span>
<span id="cb144-2"><a href="human-dlpfc-workflow.html#cb144-2" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">computeLibraryFactors</span>(spe)</span>
<span id="cb144-3"><a href="human-dlpfc-workflow.html#cb144-3" tabindex="-1"></a></span>
<span id="cb144-4"><a href="human-dlpfc-workflow.html#cb144-4" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">sizeFactors</span>(spe))</span></code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##  0.1321  0.6312  0.9000  1.0000  1.2849  3.7582</code></pre>
<div class="sourceCode" id="cb146"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb146-1"><a href="human-dlpfc-workflow.html#cb146-1" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">sizeFactors</span>(spe), <span class="at">breaks =</span> <span class="dv">20</span>)</span></code></pre></div>
<p><img src="BestPracticesST_files/figure-html/unnamed-chunk-68-1.png" width="672" /></p>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="human-dlpfc-workflow.html#cb147-1" tabindex="-1"></a><span class="co"># calculate logcounts and store in object</span></span>
<span id="cb147-2"><a href="human-dlpfc-workflow.html#cb147-2" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(spe)</span>
<span id="cb147-3"><a href="human-dlpfc-workflow.html#cb147-3" tabindex="-1"></a></span>
<span id="cb147-4"><a href="human-dlpfc-workflow.html#cb147-4" tabindex="-1"></a><span class="fu">assayNames</span>(spe)</span></code></pre></div>
<pre><code>## [1] &quot;counts&quot;    &quot;logcounts&quot;</code></pre>
</div>
<div id="feature-selection-1" class="section level2 hasAnchor" number="13.6">
<h2><span class="header-section-number">13.6</span> Feature selection<a href="human-dlpfc-workflow.html#feature-selection-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Identify a set of top highly variable genes (HVGs), which will be used to define cell types. We use methods from <code>scran</code> <span class="citation">(<a href="#ref-Lun2016">Lun, McCarthy, and Marioni 2016</a>)</span>, treating spots as equivalent to single cells, and considering only molecular features (gene expression) as described in <a href="feature-selection.html#feature-selection">Feature selection</a>. We also first filter out mitochondrial genes, since these are very highly expressed and not of main biological interest here.</p>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb149-1"><a href="human-dlpfc-workflow.html#cb149-1" tabindex="-1"></a><span class="co"># remove mitochondrial genes</span></span>
<span id="cb149-2"><a href="human-dlpfc-workflow.html#cb149-2" tabindex="-1"></a>spe <span class="ot">&lt;-</span> spe[<span class="sc">!</span>is_mito, ]</span>
<span id="cb149-3"><a href="human-dlpfc-workflow.html#cb149-3" tabindex="-1"></a><span class="fu">dim</span>(spe)</span></code></pre></div>
<pre><code>## [1] 33525  3524</code></pre>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="human-dlpfc-workflow.html#cb151-1" tabindex="-1"></a><span class="co"># fit mean-variance relationship</span></span>
<span id="cb151-2"><a href="human-dlpfc-workflow.html#cb151-2" tabindex="-1"></a>dec <span class="ot">&lt;-</span> <span class="fu">modelGeneVar</span>(spe)</span>
<span id="cb151-3"><a href="human-dlpfc-workflow.html#cb151-3" tabindex="-1"></a></span>
<span id="cb151-4"><a href="human-dlpfc-workflow.html#cb151-4" tabindex="-1"></a><span class="co"># visualize mean-variance relationship</span></span>
<span id="cb151-5"><a href="human-dlpfc-workflow.html#cb151-5" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">metadata</span>(dec)</span>
<span id="cb151-6"><a href="human-dlpfc-workflow.html#cb151-6" tabindex="-1"></a><span class="fu">plot</span>(fit<span class="sc">$</span>mean, fit<span class="sc">$</span>var, </span>
<span id="cb151-7"><a href="human-dlpfc-workflow.html#cb151-7" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">&quot;mean of log-expression&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;variance of log-expression&quot;</span>)</span>
<span id="cb151-8"><a href="human-dlpfc-workflow.html#cb151-8" tabindex="-1"></a><span class="fu">curve</span>(fit<span class="sc">$</span><span class="fu">trend</span>(x), <span class="at">col =</span> <span class="st">&quot;dodgerblue&quot;</span>, <span class="at">add =</span> <span class="cn">TRUE</span>, <span class="at">lwd =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="BestPracticesST_files/figure-html/unnamed-chunk-70-1.png" width="672" /></p>
<div class="sourceCode" id="cb152"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb152-1"><a href="human-dlpfc-workflow.html#cb152-1" tabindex="-1"></a><span class="co"># select top HVGs</span></span>
<span id="cb152-2"><a href="human-dlpfc-workflow.html#cb152-2" tabindex="-1"></a>top_hvgs <span class="ot">&lt;-</span> <span class="fu">getTopHVGs</span>(dec, <span class="at">prop =</span> <span class="fl">0.1</span>)</span>
<span id="cb152-3"><a href="human-dlpfc-workflow.html#cb152-3" tabindex="-1"></a><span class="fu">length</span>(top_hvgs)</span></code></pre></div>
<pre><code>## [1] 1438</code></pre>
</div>
<div id="spatially-aware-feature-selection" class="section level2 hasAnchor" number="13.7">
<h2><span class="header-section-number">13.7</span> Spatially-aware feature selection<a href="human-dlpfc-workflow.html#spatially-aware-feature-selection" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Alternatively, run <a href="https://bioconductor.org/packages/nnSVG">nnSVG</a> to identify a set of top spatially variable genes (SVGs) instead of HVGs.</p>
<p>Here, we run nnSVG using a small subset of the dataset for faster runtime. We select a subset of the data by subsampling on the set of spots and including stringent filtering for low-expressed genes. For a full analysis, we recommend running <code>nnSVG</code> on all spots and using default filtering parameters (for Visium data from human brain tissue), which takes around 45 minutes for one Visium slide on a standard laptop using multiple cores.</p>
<div class="sourceCode" id="cb154"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb154-1"><a href="human-dlpfc-workflow.html#cb154-1" tabindex="-1"></a><span class="fu">library</span>(nnSVG)</span></code></pre></div>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb155-1"><a href="human-dlpfc-workflow.html#cb155-1" tabindex="-1"></a><span class="co"># subsample spots</span></span>
<span id="cb155-2"><a href="human-dlpfc-workflow.html#cb155-2" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb155-3"><a href="human-dlpfc-workflow.html#cb155-3" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb155-4"><a href="human-dlpfc-workflow.html#cb155-4" tabindex="-1"></a>ix <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">seq_len</span>(n), n)</span>
<span id="cb155-5"><a href="human-dlpfc-workflow.html#cb155-5" tabindex="-1"></a></span>
<span id="cb155-6"><a href="human-dlpfc-workflow.html#cb155-6" tabindex="-1"></a>spe_nnSVG <span class="ot">&lt;-</span> spe[, ix]</span>
<span id="cb155-7"><a href="human-dlpfc-workflow.html#cb155-7" tabindex="-1"></a></span>
<span id="cb155-8"><a href="human-dlpfc-workflow.html#cb155-8" tabindex="-1"></a><span class="co"># filter low-expressed and mitochondrial genes</span></span>
<span id="cb155-9"><a href="human-dlpfc-workflow.html#cb155-9" tabindex="-1"></a><span class="co"># using very stringent filtering parameters for faster runtime in this example</span></span>
<span id="cb155-10"><a href="human-dlpfc-workflow.html#cb155-10" tabindex="-1"></a><span class="co"># note: for a full analysis, use alternative filtering parameters (e.g. defaults)</span></span>
<span id="cb155-11"><a href="human-dlpfc-workflow.html#cb155-11" tabindex="-1"></a>spe_nnSVG <span class="ot">&lt;-</span> <span class="fu">filter_genes</span>(</span>
<span id="cb155-12"><a href="human-dlpfc-workflow.html#cb155-12" tabindex="-1"></a>  spe_nnSVG, <span class="at">filter_genes_ncounts =</span> <span class="dv">10</span>, <span class="at">filter_genes_pcspots =</span> <span class="dv">3</span></span>
<span id="cb155-13"><a href="human-dlpfc-workflow.html#cb155-13" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## Gene filtering: removing mitochondrial genes</code></pre>
<pre><code>## removed 0 mitochondrial genes</code></pre>
<pre><code>## Gene filtering: retaining genes with at least 10 counts in at least 3% (n = 3) of spatial locations</code></pre>
<pre><code>## removed 33353 out of 33525 genes due to low expression</code></pre>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb160-1"><a href="human-dlpfc-workflow.html#cb160-1" tabindex="-1"></a><span class="co"># re-calculate logcounts after filtering</span></span>
<span id="cb160-2"><a href="human-dlpfc-workflow.html#cb160-2" tabindex="-1"></a><span class="co"># using library size factors</span></span>
<span id="cb160-3"><a href="human-dlpfc-workflow.html#cb160-3" tabindex="-1"></a>spe_nnSVG <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(spe_nnSVG)</span>
<span id="cb160-4"><a href="human-dlpfc-workflow.html#cb160-4" tabindex="-1"></a></span>
<span id="cb160-5"><a href="human-dlpfc-workflow.html#cb160-5" tabindex="-1"></a><span class="co"># run nnSVG</span></span>
<span id="cb160-6"><a href="human-dlpfc-workflow.html#cb160-6" tabindex="-1"></a><span class="co"># using a single core for compatibility on build system</span></span>
<span id="cb160-7"><a href="human-dlpfc-workflow.html#cb160-7" tabindex="-1"></a><span class="co"># note: for a full analysis, use multiple cores</span></span>
<span id="cb160-8"><a href="human-dlpfc-workflow.html#cb160-8" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb160-9"><a href="human-dlpfc-workflow.html#cb160-9" tabindex="-1"></a>spe_nnSVG <span class="ot">&lt;-</span> <span class="fu">nnSVG</span>(spe_nnSVG, <span class="at">n_threads =</span> <span class="dv">1</span>)</span>
<span id="cb160-10"><a href="human-dlpfc-workflow.html#cb160-10" tabindex="-1"></a></span>
<span id="cb160-11"><a href="human-dlpfc-workflow.html#cb160-11" tabindex="-1"></a><span class="co"># investigate results</span></span>
<span id="cb160-12"><a href="human-dlpfc-workflow.html#cb160-12" tabindex="-1"></a></span>
<span id="cb160-13"><a href="human-dlpfc-workflow.html#cb160-13" tabindex="-1"></a><span class="co"># show results</span></span>
<span id="cb160-14"><a href="human-dlpfc-workflow.html#cb160-14" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rowData</span>(spe_nnSVG), <span class="dv">3</span>)</span></code></pre></div>
<pre><code>## DataFrame with 3 rows and 17 columns
##                         gene_id   gene_name    feature_type   sigma.sq
##                     &lt;character&gt; &lt;character&gt;     &lt;character&gt;  &lt;numeric&gt;
## ENSG00000074800 ENSG00000074800        ENO1 Gene Expression 0.00840777
## ENSG00000171603 ENSG00000171603      CLSTN1 Gene Expression 0.43102027
## ENSG00000162545 ENSG00000162545     CAMK2N1 Gene Expression 0.15373440
##                    tau.sq       phi    loglik   runtime      mean       var
##                 &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## ENSG00000074800  0.518011  15.82454  -109.808     0.024   1.75820  0.530929
## ENSG00000171603  0.298698  18.12310  -123.212     0.026   2.07433  0.741077
## ENSG00000162545  0.547736   2.38046  -119.576     0.018   2.63576  0.702846
##                     spcov   prop_sv loglik_lm   LR_stat      rank      pval
##                 &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## ENSG00000074800 0.0521523 0.0159716  -109.735 -0.145454       167 1.0000000
## ENSG00000171603 0.3164987 0.5906665  -126.409  6.394009        86 0.0408845
## ENSG00000162545 0.1487576 0.2191603  -123.760  8.368951        70 0.0152302
##                      padj
##                 &lt;numeric&gt;
## ENSG00000074800 1.0000000
## ENSG00000171603 0.0817690
## ENSG00000162545 0.0374227</code></pre>
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb162-1"><a href="human-dlpfc-workflow.html#cb162-1" tabindex="-1"></a><span class="co"># number of significant SVGs</span></span>
<span id="cb162-2"><a href="human-dlpfc-workflow.html#cb162-2" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">rowData</span>(spe_nnSVG)<span class="sc">$</span>padj <span class="sc">&lt;=</span> <span class="fl">0.05</span>)</span></code></pre></div>
<pre><code>## 
## FALSE  TRUE 
##    96    76</code></pre>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="human-dlpfc-workflow.html#cb164-1" tabindex="-1"></a><span class="co"># show results for top n SVGs</span></span>
<span id="cb164-2"><a href="human-dlpfc-workflow.html#cb164-2" tabindex="-1"></a><span class="fu">rowData</span>(spe_nnSVG)[<span class="fu">order</span>(<span class="fu">rowData</span>(spe_nnSVG)<span class="sc">$</span>rank)[<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>], ]</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 17 columns
##                         gene_id   gene_name    feature_type  sigma.sq    tau.sq
##                     &lt;character&gt; &lt;character&gt;     &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt;
## ENSG00000197971 ENSG00000197971         MBP Gene Expression   3.92430  0.175336
## ENSG00000123560 ENSG00000123560        PLP1 Gene Expression   3.23544  0.461590
## ENSG00000109846 ENSG00000109846       CRYAB Gene Expression   1.89968  0.281795
## ENSG00000173786 ENSG00000173786         CNP Gene Expression   2.28793  0.402524
## ENSG00000131095 ENSG00000131095        GFAP Gene Expression   2.23709  0.461297
## ENSG00000160307 ENSG00000160307       S100B Gene Expression   1.24179  0.155737
##                       phi    loglik   runtime      mean       var     spcov
##                 &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## ENSG00000197971   0.93014  -118.372     0.025   3.78073   2.76535  0.523969
## ENSG00000123560   1.03983  -140.269     0.022   2.86143   3.02555  0.628613
## ENSG00000109846   1.88555  -126.692     0.021   1.86058   1.88123  0.740785
## ENSG00000173786   1.02675  -129.206     0.020   1.79558   1.97274  0.842396
## ENSG00000131095   2.49083  -149.004     0.022   1.94543   2.71281  0.768823
## ENSG00000160307   5.76279  -129.722     0.019   1.82695   1.46946  0.609953
##                   prop_sv loglik_lm   LR_stat      rank        pval        padj
##                 &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;
## ENSG00000197971  0.957231  -192.250  147.7556         1 0.00000e+00 0.00000e+00
## ENSG00000123560  0.875146  -196.746  112.9549         2 0.00000e+00 0.00000e+00
## ENSG00000109846  0.870824  -172.988   92.5906         3 0.00000e+00 0.00000e+00
## ENSG00000173786  0.850388  -175.363   92.3137         4 0.00000e+00 0.00000e+00
## ENSG00000131095  0.829047  -191.291   84.5739         5 0.00000e+00 0.00000e+00
## ENSG00000160307  0.888562  -160.636   61.8281         6 3.75255e-14 1.07573e-12</code></pre>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb166-1"><a href="human-dlpfc-workflow.html#cb166-1" tabindex="-1"></a><span class="co"># identify top-ranked SVG</span></span>
<span id="cb166-2"><a href="human-dlpfc-workflow.html#cb166-2" tabindex="-1"></a><span class="fu">rowData</span>(spe_nnSVG)<span class="sc">$</span>gene_name[<span class="fu">which</span>(<span class="fu">rowData</span>(spe_nnSVG)<span class="sc">$</span>rank <span class="sc">==</span> <span class="dv">1</span>)]</span></code></pre></div>
<pre><code>## [1] &quot;MBP&quot;</code></pre>
</div>
<div id="dimensionality-reduction-1" class="section level2 hasAnchor" number="13.8">
<h2><span class="header-section-number">13.8</span> Dimensionality reduction<a href="human-dlpfc-workflow.html#dimensionality-reduction-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Run principal component analysis (PCA) on the set of top HVGs, and retain the top 50 principal components (PCs) for further downstream analyses. This is done both to reduce noise and to improve computational efficiency. We also run UMAP on the set of top 50 PCs and retain the top 2 UMAP components for visualization purposes.</p>
<p>We use the computationally efficient implementation of PCA available in <code>scater</code> <span class="citation">(<a href="#ref-McCarthy2017">McCarthy et al. 2017</a>)</span>, which uses randomization, and therefore requires setting a random seed for reproducibility.</p>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb168-1"><a href="human-dlpfc-workflow.html#cb168-1" tabindex="-1"></a><span class="co"># compute PCA</span></span>
<span id="cb168-2"><a href="human-dlpfc-workflow.html#cb168-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb168-3"><a href="human-dlpfc-workflow.html#cb168-3" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(spe, <span class="at">subset_row =</span> top_hvgs)</span>
<span id="cb168-4"><a href="human-dlpfc-workflow.html#cb168-4" tabindex="-1"></a></span>
<span id="cb168-5"><a href="human-dlpfc-workflow.html#cb168-5" tabindex="-1"></a><span class="fu">reducedDimNames</span>(spe)</span></code></pre></div>
<pre><code>## [1] &quot;PCA&quot;</code></pre>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="human-dlpfc-workflow.html#cb170-1" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">reducedDim</span>(spe, <span class="st">&quot;PCA&quot;</span>))</span></code></pre></div>
<pre><code>## [1] 3524   50</code></pre>
<div class="sourceCode" id="cb172"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb172-1"><a href="human-dlpfc-workflow.html#cb172-1" tabindex="-1"></a><span class="co"># compute UMAP on top 50 PCs</span></span>
<span id="cb172-2"><a href="human-dlpfc-workflow.html#cb172-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb172-3"><a href="human-dlpfc-workflow.html#cb172-3" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">runUMAP</span>(spe, <span class="at">dimred =</span> <span class="st">&quot;PCA&quot;</span>)</span>
<span id="cb172-4"><a href="human-dlpfc-workflow.html#cb172-4" tabindex="-1"></a></span>
<span id="cb172-5"><a href="human-dlpfc-workflow.html#cb172-5" tabindex="-1"></a><span class="fu">reducedDimNames</span>(spe)</span></code></pre></div>
<pre><code>## [1] &quot;PCA&quot;  &quot;UMAP&quot;</code></pre>
<div class="sourceCode" id="cb174"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb174-1"><a href="human-dlpfc-workflow.html#cb174-1" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">reducedDim</span>(spe, <span class="st">&quot;UMAP&quot;</span>))</span></code></pre></div>
<pre><code>## [1] 3524    2</code></pre>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="human-dlpfc-workflow.html#cb176-1" tabindex="-1"></a><span class="co"># update column names for easier plotting</span></span>
<span id="cb176-2"><a href="human-dlpfc-workflow.html#cb176-2" tabindex="-1"></a><span class="fu">colnames</span>(<span class="fu">reducedDim</span>(spe, <span class="st">&quot;UMAP&quot;</span>)) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;UMAP&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)</span></code></pre></div>
</div>
<div id="clustering-1" class="section level2 hasAnchor" number="13.9">
<h2><span class="header-section-number">13.9</span> Clustering<a href="human-dlpfc-workflow.html#clustering-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Next, we perform clustering to define cell types. Here, we use molecular features (gene expression) only, as described in <a href="clustering.html#clustering">Clustering</a>. We apply graph-based clustering using the Walktrap method implemented in <code>scran</code> <span class="citation">(<a href="#ref-Lun2016">Lun, McCarthy, and Marioni 2016</a>)</span>, applied to the top 50 PCs calculated on the set of top HVGs.</p>
<div class="sourceCode" id="cb177"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb177-1"><a href="human-dlpfc-workflow.html#cb177-1" tabindex="-1"></a><span class="co"># graph-based clustering</span></span>
<span id="cb177-2"><a href="human-dlpfc-workflow.html#cb177-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb177-3"><a href="human-dlpfc-workflow.html#cb177-3" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb177-4"><a href="human-dlpfc-workflow.html#cb177-4" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">buildSNNGraph</span>(spe, <span class="at">k =</span> k, <span class="at">use.dimred =</span> <span class="st">&quot;PCA&quot;</span>)</span>
<span id="cb177-5"><a href="human-dlpfc-workflow.html#cb177-5" tabindex="-1"></a>g_walk <span class="ot">&lt;-</span> igraph<span class="sc">::</span><span class="fu">cluster_walktrap</span>(g)</span>
<span id="cb177-6"><a href="human-dlpfc-workflow.html#cb177-6" tabindex="-1"></a>clus <span class="ot">&lt;-</span> g_walk<span class="sc">$</span>membership</span>
<span id="cb177-7"><a href="human-dlpfc-workflow.html#cb177-7" tabindex="-1"></a><span class="fu">table</span>(clus)</span></code></pre></div>
<pre><code>## clus
##    1    2    3    4    5    6    7 
##  338  312 1146  978  274  116  360</code></pre>
<div class="sourceCode" id="cb179"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb179-1"><a href="human-dlpfc-workflow.html#cb179-1" tabindex="-1"></a><span class="co"># store cluster labels in column &#39;label&#39; in colData</span></span>
<span id="cb179-2"><a href="human-dlpfc-workflow.html#cb179-2" tabindex="-1"></a><span class="fu">colLabels</span>(spe) <span class="ot">&lt;-</span> <span class="fu">factor</span>(clus)</span></code></pre></div>
<p>Visualize the clusters by plotting in spatial (x-y) coordinates on the tissue slide, and in UMAP dimensions.</p>
<p>From the visualizations, we can see that the clustering reproduces the known biological structure (cortical layers), although not perfectly. The clusters are also separated in UMAP space, but again not perfectly.</p>
<div class="sourceCode" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb180-1"><a href="human-dlpfc-workflow.html#cb180-1" tabindex="-1"></a><span class="co"># plot clusters in spatial x-y coordinates</span></span>
<span id="cb180-2"><a href="human-dlpfc-workflow.html#cb180-2" tabindex="-1"></a><span class="fu">plotSpots</span>(spe, <span class="at">annotate =</span> <span class="st">&quot;label&quot;</span>, </span>
<span id="cb180-3"><a href="human-dlpfc-workflow.html#cb180-3" tabindex="-1"></a>          <span class="at">palette =</span> <span class="st">&quot;libd_layer_colors&quot;</span>)</span></code></pre></div>
<p><img src="BestPracticesST_files/figure-html/unnamed-chunk-76-1.png" width="672" /></p>
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="human-dlpfc-workflow.html#cb181-1" tabindex="-1"></a><span class="co"># plot ground truth labels in spatial coordinates</span></span>
<span id="cb181-2"><a href="human-dlpfc-workflow.html#cb181-2" tabindex="-1"></a><span class="fu">plotSpots</span>(spe, <span class="at">annotate =</span> <span class="st">&quot;ground_truth&quot;</span>, </span>
<span id="cb181-3"><a href="human-dlpfc-workflow.html#cb181-3" tabindex="-1"></a>          <span class="at">palette =</span> <span class="st">&quot;libd_layer_colors&quot;</span>)</span></code></pre></div>
<p><img src="BestPracticesST_files/figure-html/unnamed-chunk-76-2.png" width="672" /></p>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="human-dlpfc-workflow.html#cb182-1" tabindex="-1"></a><span class="co"># plot clusters in UMAP reduced dimensions</span></span>
<span id="cb182-2"><a href="human-dlpfc-workflow.html#cb182-2" tabindex="-1"></a><span class="fu">plotDimRed</span>(spe, <span class="at">type =</span> <span class="st">&quot;UMAP&quot;</span>, </span>
<span id="cb182-3"><a href="human-dlpfc-workflow.html#cb182-3" tabindex="-1"></a>           <span class="at">annotate =</span> <span class="st">&quot;label&quot;</span>, <span class="at">palette =</span> <span class="st">&quot;libd_layer_colors&quot;</span>)</span></code></pre></div>
<p><img src="BestPracticesST_files/figure-html/unnamed-chunk-77-1.png" width="480" /></p>
</div>
<div id="marker-genes-2" class="section level2 hasAnchor" number="13.10">
<h2><span class="header-section-number">13.10</span> Marker genes<a href="human-dlpfc-workflow.html#marker-genes-2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Identify marker genes by testing for differential gene expression between clusters. We use the <code>findMarkers</code> implementation in <code>scran</code> <span class="citation">(<a href="#ref-Lun2016">Lun, McCarthy, and Marioni 2016</a>)</span>, using a binomial test, which tests for genes that differ in the proportion expressed vs. not expressed between clusters. This is a more stringent test than the default t-tests, and tends to select genes that are easier to interpret and validate experimentally.</p>
<div class="sourceCode" id="cb183"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb183-1"><a href="human-dlpfc-workflow.html#cb183-1" tabindex="-1"></a><span class="co"># set gene names as row names for easier plotting</span></span>
<span id="cb183-2"><a href="human-dlpfc-workflow.html#cb183-2" tabindex="-1"></a><span class="fu">rownames</span>(spe) <span class="ot">&lt;-</span> <span class="fu">rowData</span>(spe)<span class="sc">$</span>gene_name</span>
<span id="cb183-3"><a href="human-dlpfc-workflow.html#cb183-3" tabindex="-1"></a></span>
<span id="cb183-4"><a href="human-dlpfc-workflow.html#cb183-4" tabindex="-1"></a><span class="co"># test for marker genes</span></span>
<span id="cb183-5"><a href="human-dlpfc-workflow.html#cb183-5" tabindex="-1"></a>markers <span class="ot">&lt;-</span> <span class="fu">findMarkers</span>(spe, <span class="at">test =</span> <span class="st">&quot;binom&quot;</span>, <span class="at">direction =</span> <span class="st">&quot;up&quot;</span>)</span>
<span id="cb183-6"><a href="human-dlpfc-workflow.html#cb183-6" tabindex="-1"></a></span>
<span id="cb183-7"><a href="human-dlpfc-workflow.html#cb183-7" tabindex="-1"></a><span class="co"># returns a list with one DataFrame per cluster</span></span>
<span id="cb183-8"><a href="human-dlpfc-workflow.html#cb183-8" tabindex="-1"></a>markers</span></code></pre></div>
<pre><code>## List of length 7
## names(7): 1 2 3 4 5 6 7</code></pre>
<div class="sourceCode" id="cb185"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb185-1"><a href="human-dlpfc-workflow.html#cb185-1" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span></code></pre></div>
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="human-dlpfc-workflow.html#cb186-1" tabindex="-1"></a><span class="co"># plot log-fold changes for one cluster over all other clusters</span></span>
<span id="cb186-2"><a href="human-dlpfc-workflow.html#cb186-2" tabindex="-1"></a><span class="co"># selecting cluster 1</span></span>
<span id="cb186-3"><a href="human-dlpfc-workflow.html#cb186-3" tabindex="-1"></a>interesting <span class="ot">&lt;-</span> markers[[<span class="dv">1</span>]]</span>
<span id="cb186-4"><a href="human-dlpfc-workflow.html#cb186-4" tabindex="-1"></a>best_set <span class="ot">&lt;-</span> interesting[interesting<span class="sc">$</span>Top <span class="sc">&lt;=</span> <span class="dv">5</span>, ]</span>
<span id="cb186-5"><a href="human-dlpfc-workflow.html#cb186-5" tabindex="-1"></a>logFCs <span class="ot">&lt;-</span> <span class="fu">getMarkerEffects</span>(best_set)</span>
<span id="cb186-6"><a href="human-dlpfc-workflow.html#cb186-6" tabindex="-1"></a></span>
<span id="cb186-7"><a href="human-dlpfc-workflow.html#cb186-7" tabindex="-1"></a><span class="fu">pheatmap</span>(logFCs, <span class="at">breaks =</span> <span class="fu">seq</span>(<span class="sc">-</span><span class="dv">5</span>, <span class="dv">5</span>, <span class="at">length.out =</span> <span class="dv">101</span>))</span></code></pre></div>
<p><img src="BestPracticesST_files/figure-html/unnamed-chunk-80-1.png" width="480" /></p>
<div class="sourceCode" id="cb187"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb187-1"><a href="human-dlpfc-workflow.html#cb187-1" tabindex="-1"></a><span class="co"># plot log-transformed normalized expression of top genes for one cluster</span></span>
<span id="cb187-2"><a href="human-dlpfc-workflow.html#cb187-2" tabindex="-1"></a>top_genes <span class="ot">&lt;-</span> <span class="fu">head</span>(<span class="fu">rownames</span>(interesting))</span>
<span id="cb187-3"><a href="human-dlpfc-workflow.html#cb187-3" tabindex="-1"></a></span>
<span id="cb187-4"><a href="human-dlpfc-workflow.html#cb187-4" tabindex="-1"></a><span class="fu">plotExpression</span>(spe, <span class="at">x =</span> <span class="st">&quot;label&quot;</span>, <span class="at">features =</span> top_genes)</span></code></pre></div>
<p><img src="BestPracticesST_files/figure-html/unnamed-chunk-81-1.png" width="672" /></p>

</div>
</div>
<h3> References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Lun2016" class="csl-entry">
Lun, Aaron T L, Davis J McCarthy, and John C Marioni. 2016. <span>“A Step-by-Step Workflow for Low-Level Analysis of Single-Cell <span class="nocase">RNA-seq</span> Data with <span>Bioconductor</span>.”</span> <em>F1000Research</em> 5: 2122. <a href="https://doi.org/10.12688/f1000research.9501.2">https://doi.org/10.12688/f1000research.9501.2</a>.
</div>
<div id="ref-Maynard2021" class="csl-entry">
Maynard, Kristen R., Leonardo Collado-Torres, Lukas M. Weber, Cedric Uytingco, Brianna K. Barry, Stephen R. Williams, Joseph L. Catallini II, et al. 2021. <span>“Transcriptome-Scale Spatial Gene Expression in the Human Dorsolateral Prefrontal Cortex.”</span> <em>Nature Neuroscience</em>. <a href="https://doi.org/10.1038/s41593-020-00787-0">https://doi.org/10.1038/s41593-020-00787-0</a>.
</div>
<div id="ref-McCarthy2017" class="csl-entry">
McCarthy, Davis J, Kieran R Campbell, Aaron T L Lun, and Quin F Wills. 2017. <span>“Scater: Pre-Processing, Quality Control, Normalization and Visualization of Single-Cell <span class="nocase">RNA-seq</span> Data in <span>R</span>.”</span> <em>Bioinformatics</em> 33 (8): 1179–86. <a href="https://doi.org/10.1093/bioinformatics/btw777">https://doi.org/10.1093/bioinformatics/btw777</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="workflows.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="mouse-coronal-workflow.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
},
"lib_dir": "book_assets"
});
});
</script>

</body>

</html>
