<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 10 Marker genes | Best Practices for Spatial Transcriptomics Analysis with Bioconductor</title>
  <meta name="description" content="Online book ‘Best Practices for Spatial Transcriptomics Analysis with Bioconductor’" />
  <meta name="generator" content="bookdown 0.35 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 10 Marker genes | Best Practices for Spatial Transcriptomics Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="Online book ‘Best Practices for Spatial Transcriptomics Analysis with Bioconductor’" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 10 Marker genes | Best Practices for Spatial Transcriptomics Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Online book ‘Best Practices for Spatial Transcriptomics Analysis with Bioconductor’" />
  



<meta name="date" content="2023-09-12" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="clustering.html"/>
<link rel="next" href="spot-level-deconvolution.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 1em;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Best Practices ST</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#contents"><i class="fa fa-check"></i><b>1.1</b> Contents</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#who-this-book-is-for"><i class="fa fa-check"></i><b>1.2</b> Who this book is for</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#bioconductor"><i class="fa fa-check"></i><b>1.3</b> Bioconductor</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#contributions"><i class="fa fa-check"></i><b>1.4</b> Contributions</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html"><i class="fa fa-check"></i><b>2</b> Spatially-resolved transcriptomics</a>
<ul>
<li class="chapter" data-level="2.1" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#overview"><i class="fa fa-check"></i><b>2.1</b> Overview</a></li>
<li class="chapter" data-level="2.2" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#sequencing-based-platforms"><i class="fa fa-check"></i><b>2.2</b> Sequencing-based platforms</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#x-genomics-visium"><i class="fa fa-check"></i><b>2.2.1</b> 10x Genomics Visium</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#molecule-based-platforms"><i class="fa fa-check"></i><b>2.3</b> Molecule-based platforms</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#x-genomics-xenium"><i class="fa fa-check"></i><b>2.3.1</b> 10x Genomics Xenium</a></li>
<li class="chapter" data-level="2.3.2" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#vizgen-merscope"><i class="fa fa-check"></i><b>2.3.2</b> Vizgen MERSCOPE</a></li>
<li class="chapter" data-level="2.3.3" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#nanostring-cosmx"><i class="fa fa-check"></i><b>2.3.3</b> NanoString CosMx</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="bioconductor-data-classes.html"><a href="bioconductor-data-classes.html"><i class="fa fa-check"></i><b>3</b> Bioconductor data classes</a>
<ul>
<li class="chapter" data-level="3.1" data-path="bioconductor-data-classes.html"><a href="bioconductor-data-classes.html#spatialexperiment-class"><i class="fa fa-check"></i><b>3.1</b> SpatialExperiment class</a></li>
<li class="chapter" data-level="3.2" data-path="bioconductor-data-classes.html"><a href="bioconductor-data-classes.html#molecule-based-data"><i class="fa fa-check"></i><b>3.2</b> Molecule-based data</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="bioconductor-data-classes.html"><a href="bioconductor-data-classes.html#moleculeexperiment"><i class="fa fa-check"></i><b>3.2.1</b> MoleculeExperiment</a></li>
<li class="chapter" data-level="3.2.2" data-path="bioconductor-data-classes.html"><a href="bioconductor-data-classes.html#spatialfeatureexperiment"><i class="fa fa-check"></i><b>3.2.2</b> SpatialFeatureExperiment</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>II Analysis steps</b></span></li>
<li class="chapter" data-level="4" data-path="analysis-steps.html"><a href="analysis-steps.html"><i class="fa fa-check"></i><b>4</b> Analysis steps</a>
<ul>
<li class="chapter" data-level="4.1" data-path="analysis-steps.html"><a href="analysis-steps.html#load-data"><i class="fa fa-check"></i><b>4.1</b> Load data</a></li>
<li class="chapter" data-level="4.2" data-path="analysis-steps.html"><a href="analysis-steps.html#spatialexperiment-object"><i class="fa fa-check"></i><b>4.2</b> SpatialExperiment object</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>5</b> Quality control</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quality-control.html"><a href="quality-control.html#background"><i class="fa fa-check"></i><b>5.1</b> Background</a></li>
<li class="chapter" data-level="5.2" data-path="quality-control.html"><a href="quality-control.html#load-data-1"><i class="fa fa-check"></i><b>5.2</b> Load data</a></li>
<li class="chapter" data-level="5.3" data-path="quality-control.html"><a href="quality-control.html#plot-data"><i class="fa fa-check"></i><b>5.3</b> Plot data</a></li>
<li class="chapter" data-level="5.4" data-path="quality-control.html"><a href="quality-control.html#calculate-qc-metrics"><i class="fa fa-check"></i><b>5.4</b> Calculate QC metrics</a></li>
<li class="chapter" data-level="5.5" data-path="quality-control.html"><a href="quality-control.html#selecting-thresholds"><i class="fa fa-check"></i><b>5.5</b> Selecting thresholds</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="quality-control.html"><a href="quality-control.html#library-size"><i class="fa fa-check"></i><b>5.5.1</b> Library size</a></li>
<li class="chapter" data-level="5.5.2" data-path="quality-control.html"><a href="quality-control.html#number-of-expressed-features"><i class="fa fa-check"></i><b>5.5.2</b> Number of expressed features</a></li>
<li class="chapter" data-level="5.5.3" data-path="quality-control.html"><a href="quality-control.html#proportion-of-mitochondrial-reads"><i class="fa fa-check"></i><b>5.5.3</b> Proportion of mitochondrial reads</a></li>
<li class="chapter" data-level="5.5.4" data-path="quality-control.html"><a href="quality-control.html#number-of-cells-per-spot"><i class="fa fa-check"></i><b>5.5.4</b> Number of cells per spot</a></li>
<li class="chapter" data-level="5.5.5" data-path="quality-control.html"><a href="quality-control.html#remove-low-quality-spots"><i class="fa fa-check"></i><b>5.5.5</b> Remove low-quality spots</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="quality-control.html"><a href="quality-control.html#zero-cell-and-single-cell-spots"><i class="fa fa-check"></i><b>5.6</b> Zero-cell and single-cell spots</a></li>
<li class="chapter" data-level="5.7" data-path="quality-control.html"><a href="quality-control.html#quality-control-at-gene-level"><i class="fa fa-check"></i><b>5.7</b> Quality control at gene level</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="normalization.html"><a href="normalization.html"><i class="fa fa-check"></i><b>6</b> Normalization</a>
<ul>
<li class="chapter" data-level="6.1" data-path="normalization.html"><a href="normalization.html#background-1"><i class="fa fa-check"></i><b>6.1</b> Background</a></li>
<li class="chapter" data-level="6.2" data-path="normalization.html"><a href="normalization.html#previous-steps"><i class="fa fa-check"></i><b>6.2</b> Previous steps</a></li>
<li class="chapter" data-level="6.3" data-path="normalization.html"><a href="normalization.html#logcounts"><i class="fa fa-check"></i><b>6.3</b> Logcounts</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="feature-selection.html"><a href="feature-selection.html"><i class="fa fa-check"></i><b>7</b> Feature selection</a>
<ul>
<li class="chapter" data-level="7.1" data-path="feature-selection.html"><a href="feature-selection.html#background-2"><i class="fa fa-check"></i><b>7.1</b> Background</a></li>
<li class="chapter" data-level="7.2" data-path="feature-selection.html"><a href="feature-selection.html#previous-steps-1"><i class="fa fa-check"></i><b>7.2</b> Previous steps</a></li>
<li class="chapter" data-level="7.3" data-path="feature-selection.html"><a href="feature-selection.html#highly-variable-genes-hvgs"><i class="fa fa-check"></i><b>7.3</b> Highly variable genes (HVGs)</a></li>
<li class="chapter" data-level="7.4" data-path="feature-selection.html"><a href="feature-selection.html#spatially-variable-genes-svgs"><i class="fa fa-check"></i><b>7.4</b> Spatially variable genes (SVGs)</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html"><i class="fa fa-check"></i><b>8</b> Dimensionality reduction</a>
<ul>
<li class="chapter" data-level="8.1" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#background-3"><i class="fa fa-check"></i><b>8.1</b> Background</a></li>
<li class="chapter" data-level="8.2" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#previous-steps-2"><i class="fa fa-check"></i><b>8.2</b> Previous steps</a></li>
<li class="chapter" data-level="8.3" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>8.3</b> Principal component analysis (PCA)</a></li>
<li class="chapter" data-level="8.4" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#uniform-manifold-approximation-and-projection-umap"><i class="fa fa-check"></i><b>8.4</b> Uniform Manifold Approximation and Projection (UMAP)</a></li>
<li class="chapter" data-level="8.5" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#visualizations"><i class="fa fa-check"></i><b>8.5</b> Visualizations</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>9</b> Clustering</a>
<ul>
<li class="chapter" data-level="9.1" data-path="clustering.html"><a href="clustering.html#overview-1"><i class="fa fa-check"></i><b>9.1</b> Overview</a></li>
<li class="chapter" data-level="9.2" data-path="clustering.html"><a href="clustering.html#previous-steps-3"><i class="fa fa-check"></i><b>9.2</b> Previous steps</a></li>
<li class="chapter" data-level="9.3" data-path="clustering.html"><a href="clustering.html#non-spatial-clustering-on-hvgs"><i class="fa fa-check"></i><b>9.3</b> Non-spatial clustering on HVGs</a></li>
<li class="chapter" data-level="9.4" data-path="clustering.html"><a href="clustering.html#spatially-aware-clustering"><i class="fa fa-check"></i><b>9.4</b> Spatially-aware clustering</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="marker-genes.html"><a href="marker-genes.html"><i class="fa fa-check"></i><b>10</b> Marker genes</a>
<ul>
<li class="chapter" data-level="10.1" data-path="marker-genes.html"><a href="marker-genes.html#background-4"><i class="fa fa-check"></i><b>10.1</b> Background</a></li>
<li class="chapter" data-level="10.2" data-path="marker-genes.html"><a href="marker-genes.html#previous-steps-4"><i class="fa fa-check"></i><b>10.2</b> Previous steps</a></li>
<li class="chapter" data-level="10.3" data-path="marker-genes.html"><a href="marker-genes.html#marker-genes-1"><i class="fa fa-check"></i><b>10.3</b> Marker genes</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="spot-level-deconvolution.html"><a href="spot-level-deconvolution.html"><i class="fa fa-check"></i><b>11</b> Spot-level deconvolution</a>
<ul>
<li class="chapter" data-level="11.1" data-path="spot-level-deconvolution.html"><a href="spot-level-deconvolution.html#background-5"><i class="fa fa-check"></i><b>11.1</b> Background</a></li>
<li class="chapter" data-level="11.2" data-path="spot-level-deconvolution.html"><a href="spot-level-deconvolution.html#previous-steps-5"><i class="fa fa-check"></i><b>11.2</b> Previous steps</a></li>
<li class="chapter" data-level="11.3" data-path="spot-level-deconvolution.html"><a href="spot-level-deconvolution.html#number-of-cells-per-spot-1"><i class="fa fa-check"></i><b>11.3</b> Number of cells per spot</a></li>
</ul></li>
<li class="part"><span><b>III Workflows</b></span></li>
<li class="chapter" data-level="12" data-path="workflows.html"><a href="workflows.html"><i class="fa fa-check"></i><b>12</b> Workflows</a></li>
<li class="chapter" data-level="13" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html"><i class="fa fa-check"></i><b>13</b> Human DLPFC workflow</a>
<ul>
<li class="chapter" data-level="13.1" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#description-of-dataset"><i class="fa fa-check"></i><b>13.1</b> Description of dataset</a></li>
<li class="chapter" data-level="13.2" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#load-data-2"><i class="fa fa-check"></i><b>13.2</b> Load data</a></li>
<li class="chapter" data-level="13.3" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#plot-data-1"><i class="fa fa-check"></i><b>13.3</b> Plot data</a></li>
<li class="chapter" data-level="13.4" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#quality-control-qc"><i class="fa fa-check"></i><b>13.4</b> Quality control (QC)</a></li>
<li class="chapter" data-level="13.5" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#normalization-1"><i class="fa fa-check"></i><b>13.5</b> Normalization</a></li>
<li class="chapter" data-level="13.6" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#feature-selection-1"><i class="fa fa-check"></i><b>13.6</b> Feature selection</a></li>
<li class="chapter" data-level="13.7" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#spatially-aware-feature-selection"><i class="fa fa-check"></i><b>13.7</b> Spatially-aware feature selection</a></li>
<li class="chapter" data-level="13.8" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>13.8</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="13.9" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#clustering-1"><i class="fa fa-check"></i><b>13.9</b> Clustering</a></li>
<li class="chapter" data-level="13.10" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#marker-genes-2"><i class="fa fa-check"></i><b>13.10</b> Marker genes</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html"><i class="fa fa-check"></i><b>14</b> Mouse coronal workflow</a>
<ul>
<li class="chapter" data-level="14.1" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#description-of-dataset-1"><i class="fa fa-check"></i><b>14.1</b> Description of dataset</a></li>
<li class="chapter" data-level="14.2" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#load-data-3"><i class="fa fa-check"></i><b>14.2</b> Load data</a></li>
<li class="chapter" data-level="14.3" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#plot-data-2"><i class="fa fa-check"></i><b>14.3</b> Plot data</a></li>
<li class="chapter" data-level="14.4" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#quality-control-qc-1"><i class="fa fa-check"></i><b>14.4</b> Quality control (QC)</a></li>
<li class="chapter" data-level="14.5" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#normalization-2"><i class="fa fa-check"></i><b>14.5</b> Normalization</a></li>
<li class="chapter" data-level="14.6" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#feature-selection-2"><i class="fa fa-check"></i><b>14.6</b> Feature selection</a></li>
<li class="chapter" data-level="14.7" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#spatially-aware-feature-selection-1"><i class="fa fa-check"></i><b>14.7</b> Spatially-aware feature selection</a></li>
<li class="chapter" data-level="14.8" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#dimensionality-reduction-2"><i class="fa fa-check"></i><b>14.8</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="14.9" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#clustering-2"><i class="fa fa-check"></i><b>14.9</b> Clustering</a></li>
<li class="chapter" data-level="14.10" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#marker-genes-3"><i class="fa fa-check"></i><b>14.10</b> Marker genes</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html"><i class="fa fa-check"></i><b>15</b> spatialLIBD workflow</a>
<ul>
<li class="chapter" data-level="15.1" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#overview-2"><i class="fa fa-check"></i><b>15.1</b> Overview</a></li>
<li class="chapter" data-level="15.2" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#spatiallibd"><i class="fa fa-check"></i><b>15.2</b> spatialLIBD</a>
<ul>
<li class="chapter" data-level="15.2.1" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#why-use-spatiallibd"><i class="fa fa-check"></i><b>15.2.1</b> Why use spatialLIBD?</a></li>
<li class="chapter" data-level="15.2.2" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#want-to-learn-more-about-spatiallibd"><i class="fa fa-check"></i><b>15.2.2</b> Want to learn more about spatialLIBD?</a></li>
</ul></li>
<li class="chapter" data-level="15.3" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#code-prerequisites"><i class="fa fa-check"></i><b>15.3</b> Code prerequisites</a></li>
<li class="chapter" data-level="15.4" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#prepare-for-spatiallibd"><i class="fa fa-check"></i><b>15.4</b> Prepare for spatialLIBD</a>
<ul>
<li class="chapter" data-level="15.4.1" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#basic-information"><i class="fa fa-check"></i><b>15.4.1</b> Basic information</a></li>
<li class="chapter" data-level="15.4.2" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#gene-annotation"><i class="fa fa-check"></i><b>15.4.2</b> Gene annotation</a></li>
<li class="chapter" data-level="15.4.3" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#extra-information-and-filtering"><i class="fa fa-check"></i><b>15.4.3</b> Extra information and filtering</a></li>
</ul></li>
<li class="chapter" data-level="15.5" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#explore-the-data"><i class="fa fa-check"></i><b>15.5</b> Explore the data</a></li>
<li class="chapter" data-level="15.6" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#sharing-your-website"><i class="fa fa-check"></i><b>15.6</b> Sharing your website</a></li>
<li class="chapter" data-level="15.7" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#wrapping-up"><i class="fa fa-check"></i><b>15.7</b> Wrapping up</a></li>
<li class="chapter" data-level="" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#r-session-information"><i class="fa fa-check"></i>R session information</a></li>
</ul></li>
<li class="part"><span><b>IV Appendix</b></span></li>
<li class="chapter" data-level="16" data-path="related-resources.html"><a href="related-resources.html"><i class="fa fa-check"></i><b>16</b> Related resources</a>
<ul>
<li class="chapter" data-level="16.1" data-path="related-resources.html"><a href="related-resources.html#data-preprocessing-procedures"><i class="fa fa-check"></i><b>16.1</b> Data preprocessing procedures</a></li>
<li class="chapter" data-level="16.2" data-path="related-resources.html"><a href="related-resources.html#resources-for-other-spatially-resolved-platforms"><i class="fa fa-check"></i><b>16.2</b> Resources for other spatially-resolved platforms</a></li>
<li class="chapter" data-level="16.3" data-path="related-resources.html"><a href="related-resources.html#data-structures"><i class="fa fa-check"></i><b>16.3</b> Data structures</a></li>
<li class="chapter" data-level="16.4" data-path="related-resources.html"><a href="related-resources.html#statistical-concepts"><i class="fa fa-check"></i><b>16.4</b> Statistical concepts</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="contributors-and-acknowledgments.html"><a href="contributors-and-acknowledgments.html"><i class="fa fa-check"></i><b>17</b> Contributors and acknowledgments</a>
<ul>
<li class="chapter" data-level="17.1" data-path="contributors-and-acknowledgments.html"><a href="contributors-and-acknowledgments.html#best-practices-for-spatial-transcriptomics-analysis-with-bioconductor"><i class="fa fa-check"></i><b>17.1</b> Best Practices for Spatial Transcriptomics Analysis with Bioconductor</a></li>
<li class="chapter" data-level="17.2" data-path="contributors-and-acknowledgments.html"><a href="contributors-and-acknowledgments.html#spatialexperiment"><i class="fa fa-check"></i><b>17.2</b> SpatialExperiment</a></li>
<li class="chapter" data-level="17.3" data-path="contributors-and-acknowledgments.html"><a href="contributors-and-acknowledgments.html#github-actions-workflow"><i class="fa fa-check"></i><b>17.3</b> GitHub Actions workflow</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>18</b> References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Best Practices for Spatial Transcriptomics Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="marker-genes" class="section level1 hasAnchor" number="10">
<h1><span class="header-section-number">Chapter 10</span> Marker genes<a href="marker-genes.html#marker-genes" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="background-4" class="section level2 hasAnchor" number="10.1">
<h2><span class="header-section-number">10.1</span> Background<a href="marker-genes.html#background-4" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this chapter, we identify marker genes for each cluster by testing for differential expression between clusters.</p>
</div>
<div id="previous-steps-4" class="section level2 hasAnchor" number="10.2">
<h2><span class="header-section-number">10.2</span> Previous steps<a href="marker-genes.html#previous-steps-4" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><em>Code to run steps from the previous chapters to generate the <code>SpatialExperiment</code> object required for this chapter.</em></p>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="marker-genes.html#cb112-1" tabindex="-1"></a><span class="co"># LOAD DATA</span></span>
<span id="cb112-2"><a href="marker-genes.html#cb112-2" tabindex="-1"></a></span>
<span id="cb112-3"><a href="marker-genes.html#cb112-3" tabindex="-1"></a><span class="fu">library</span>(SpatialExperiment)</span>
<span id="cb112-4"><a href="marker-genes.html#cb112-4" tabindex="-1"></a><span class="fu">library</span>(STexampleData)</span>
<span id="cb112-5"><a href="marker-genes.html#cb112-5" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">Visium_humanDLPFC</span>()</span>
<span id="cb112-6"><a href="marker-genes.html#cb112-6" tabindex="-1"></a></span>
<span id="cb112-7"><a href="marker-genes.html#cb112-7" tabindex="-1"></a><span class="co"># QUALITY CONTROL (QC)</span></span>
<span id="cb112-8"><a href="marker-genes.html#cb112-8" tabindex="-1"></a></span>
<span id="cb112-9"><a href="marker-genes.html#cb112-9" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb112-10"><a href="marker-genes.html#cb112-10" tabindex="-1"></a><span class="co"># subset to keep only spots over tissue</span></span>
<span id="cb112-11"><a href="marker-genes.html#cb112-11" tabindex="-1"></a>spe <span class="ot">&lt;-</span> spe[, <span class="fu">colData</span>(spe)<span class="sc">$</span>in_tissue <span class="sc">==</span> <span class="dv">1</span>]</span>
<span id="cb112-12"><a href="marker-genes.html#cb112-12" tabindex="-1"></a><span class="co"># identify mitochondrial genes</span></span>
<span id="cb112-13"><a href="marker-genes.html#cb112-13" tabindex="-1"></a>is_mito <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">&quot;(^MT-)|(^mt-)&quot;</span>, <span class="fu">rowData</span>(spe)<span class="sc">$</span>gene_name)</span>
<span id="cb112-14"><a href="marker-genes.html#cb112-14" tabindex="-1"></a><span class="co"># calculate per-spot QC metrics</span></span>
<span id="cb112-15"><a href="marker-genes.html#cb112-15" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">addPerCellQC</span>(spe, <span class="at">subsets =</span> <span class="fu">list</span>(<span class="at">mito =</span> is_mito))</span>
<span id="cb112-16"><a href="marker-genes.html#cb112-16" tabindex="-1"></a><span class="co"># select QC thresholds</span></span>
<span id="cb112-17"><a href="marker-genes.html#cb112-17" tabindex="-1"></a>qc_lib_size <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe)<span class="sc">$</span>sum <span class="sc">&lt;</span> <span class="dv">600</span></span>
<span id="cb112-18"><a href="marker-genes.html#cb112-18" tabindex="-1"></a>qc_detected <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe)<span class="sc">$</span>detected <span class="sc">&lt;</span> <span class="dv">400</span></span>
<span id="cb112-19"><a href="marker-genes.html#cb112-19" tabindex="-1"></a>qc_mito <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe)<span class="sc">$</span>subsets_mito_percent <span class="sc">&gt;</span> <span class="dv">28</span></span>
<span id="cb112-20"><a href="marker-genes.html#cb112-20" tabindex="-1"></a>qc_cell_count <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe)<span class="sc">$</span>cell_count <span class="sc">&gt;</span> <span class="dv">10</span></span>
<span id="cb112-21"><a href="marker-genes.html#cb112-21" tabindex="-1"></a><span class="co"># combined set of discarded spots</span></span>
<span id="cb112-22"><a href="marker-genes.html#cb112-22" tabindex="-1"></a>discard <span class="ot">&lt;-</span> qc_lib_size <span class="sc">|</span> qc_detected <span class="sc">|</span> qc_mito <span class="sc">|</span> qc_cell_count</span>
<span id="cb112-23"><a href="marker-genes.html#cb112-23" tabindex="-1"></a><span class="fu">colData</span>(spe)<span class="sc">$</span>discard <span class="ot">&lt;-</span> discard</span>
<span id="cb112-24"><a href="marker-genes.html#cb112-24" tabindex="-1"></a><span class="co"># filter low-quality spots</span></span>
<span id="cb112-25"><a href="marker-genes.html#cb112-25" tabindex="-1"></a>spe <span class="ot">&lt;-</span> spe[, <span class="sc">!</span><span class="fu">colData</span>(spe)<span class="sc">$</span>discard]</span>
<span id="cb112-26"><a href="marker-genes.html#cb112-26" tabindex="-1"></a></span>
<span id="cb112-27"><a href="marker-genes.html#cb112-27" tabindex="-1"></a><span class="co"># NORMALIZATION</span></span>
<span id="cb112-28"><a href="marker-genes.html#cb112-28" tabindex="-1"></a></span>
<span id="cb112-29"><a href="marker-genes.html#cb112-29" tabindex="-1"></a><span class="fu">library</span>(scran)</span>
<span id="cb112-30"><a href="marker-genes.html#cb112-30" tabindex="-1"></a><span class="co"># calculate logcounts using library size factors</span></span>
<span id="cb112-31"><a href="marker-genes.html#cb112-31" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(spe)</span>
<span id="cb112-32"><a href="marker-genes.html#cb112-32" tabindex="-1"></a></span>
<span id="cb112-33"><a href="marker-genes.html#cb112-33" tabindex="-1"></a><span class="co"># FEATURE SELECTION</span></span>
<span id="cb112-34"><a href="marker-genes.html#cb112-34" tabindex="-1"></a></span>
<span id="cb112-35"><a href="marker-genes.html#cb112-35" tabindex="-1"></a><span class="co"># remove mitochondrial genes</span></span>
<span id="cb112-36"><a href="marker-genes.html#cb112-36" tabindex="-1"></a>spe <span class="ot">&lt;-</span> spe[<span class="sc">!</span>is_mito, ]</span>
<span id="cb112-37"><a href="marker-genes.html#cb112-37" tabindex="-1"></a><span class="co"># fit mean-variance relationship</span></span>
<span id="cb112-38"><a href="marker-genes.html#cb112-38" tabindex="-1"></a>dec <span class="ot">&lt;-</span> <span class="fu">modelGeneVar</span>(spe)</span>
<span id="cb112-39"><a href="marker-genes.html#cb112-39" tabindex="-1"></a><span class="co"># select top HVGs</span></span>
<span id="cb112-40"><a href="marker-genes.html#cb112-40" tabindex="-1"></a>top_hvgs <span class="ot">&lt;-</span> <span class="fu">getTopHVGs</span>(dec, <span class="at">prop =</span> <span class="fl">0.1</span>)</span>
<span id="cb112-41"><a href="marker-genes.html#cb112-41" tabindex="-1"></a></span>
<span id="cb112-42"><a href="marker-genes.html#cb112-42" tabindex="-1"></a><span class="co"># DIMENSIONALITY REDUCTION</span></span>
<span id="cb112-43"><a href="marker-genes.html#cb112-43" tabindex="-1"></a></span>
<span id="cb112-44"><a href="marker-genes.html#cb112-44" tabindex="-1"></a><span class="co"># compute PCA</span></span>
<span id="cb112-45"><a href="marker-genes.html#cb112-45" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb112-46"><a href="marker-genes.html#cb112-46" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(spe, <span class="at">subset_row =</span> top_hvgs)</span>
<span id="cb112-47"><a href="marker-genes.html#cb112-47" tabindex="-1"></a><span class="co"># compute UMAP on top 50 PCs</span></span>
<span id="cb112-48"><a href="marker-genes.html#cb112-48" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb112-49"><a href="marker-genes.html#cb112-49" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">runUMAP</span>(spe, <span class="at">dimred =</span> <span class="st">&quot;PCA&quot;</span>)</span>
<span id="cb112-50"><a href="marker-genes.html#cb112-50" tabindex="-1"></a><span class="co"># update column names</span></span>
<span id="cb112-51"><a href="marker-genes.html#cb112-51" tabindex="-1"></a><span class="fu">colnames</span>(<span class="fu">reducedDim</span>(spe, <span class="st">&quot;UMAP&quot;</span>)) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;UMAP&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)</span>
<span id="cb112-52"><a href="marker-genes.html#cb112-52" tabindex="-1"></a></span>
<span id="cb112-53"><a href="marker-genes.html#cb112-53" tabindex="-1"></a><span class="co"># CLUSTERING</span></span>
<span id="cb112-54"><a href="marker-genes.html#cb112-54" tabindex="-1"></a></span>
<span id="cb112-55"><a href="marker-genes.html#cb112-55" tabindex="-1"></a><span class="co"># graph-based clustering</span></span>
<span id="cb112-56"><a href="marker-genes.html#cb112-56" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb112-57"><a href="marker-genes.html#cb112-57" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb112-58"><a href="marker-genes.html#cb112-58" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">buildSNNGraph</span>(spe, <span class="at">k =</span> k, <span class="at">use.dimred =</span> <span class="st">&quot;PCA&quot;</span>)</span>
<span id="cb112-59"><a href="marker-genes.html#cb112-59" tabindex="-1"></a>g_walk <span class="ot">&lt;-</span> igraph<span class="sc">::</span><span class="fu">cluster_walktrap</span>(g)</span>
<span id="cb112-60"><a href="marker-genes.html#cb112-60" tabindex="-1"></a>clus <span class="ot">&lt;-</span> g_walk<span class="sc">$</span>membership</span>
<span id="cb112-61"><a href="marker-genes.html#cb112-61" tabindex="-1"></a><span class="fu">colLabels</span>(spe) <span class="ot">&lt;-</span> <span class="fu">factor</span>(clus)</span></code></pre></div>
</div>
<div id="marker-genes-1" class="section level2 hasAnchor" number="10.3">
<h2><span class="header-section-number">10.3</span> Marker genes<a href="marker-genes.html#marker-genes-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Identify marker genes by testing for differential gene expression between clusters.</p>
<p>We use the <code>findMarkers</code> implementation in <code>scran</code> <span class="citation">(<a href="#ref-Lun2016">Lun, McCarthy, and Marioni 2016</a>)</span>, using a binomial test, which tests for genes that differ in the proportion expressed vs. not expressed between clusters. This is a more stringent test than the default t-tests, and tends to select genes that are easier to interpret and validate experimentally.</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="marker-genes.html#cb113-1" tabindex="-1"></a><span class="fu">library</span>(scran)</span>
<span id="cb113-2"><a href="marker-genes.html#cb113-2" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb113-3"><a href="marker-genes.html#cb113-3" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span></code></pre></div>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="marker-genes.html#cb114-1" tabindex="-1"></a><span class="co"># set gene names as row names for easier plotting</span></span>
<span id="cb114-2"><a href="marker-genes.html#cb114-2" tabindex="-1"></a><span class="fu">rownames</span>(spe) <span class="ot">&lt;-</span> <span class="fu">rowData</span>(spe)<span class="sc">$</span>gene_name</span>
<span id="cb114-3"><a href="marker-genes.html#cb114-3" tabindex="-1"></a></span>
<span id="cb114-4"><a href="marker-genes.html#cb114-4" tabindex="-1"></a><span class="co"># test for marker genes</span></span>
<span id="cb114-5"><a href="marker-genes.html#cb114-5" tabindex="-1"></a>markers <span class="ot">&lt;-</span> <span class="fu">findMarkers</span>(spe, <span class="at">test =</span> <span class="st">&quot;binom&quot;</span>, <span class="at">direction =</span> <span class="st">&quot;up&quot;</span>)</span>
<span id="cb114-6"><a href="marker-genes.html#cb114-6" tabindex="-1"></a></span>
<span id="cb114-7"><a href="marker-genes.html#cb114-7" tabindex="-1"></a><span class="co"># returns a list with one DataFrame per cluster</span></span>
<span id="cb114-8"><a href="marker-genes.html#cb114-8" tabindex="-1"></a>markers</span></code></pre></div>
<pre><code>## List of length 7
## names(7): 1 2 3 4 5 6 7</code></pre>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="marker-genes.html#cb116-1" tabindex="-1"></a><span class="co"># plot log-fold changes for one cluster over all other clusters</span></span>
<span id="cb116-2"><a href="marker-genes.html#cb116-2" tabindex="-1"></a><span class="co"># selecting cluster 1</span></span>
<span id="cb116-3"><a href="marker-genes.html#cb116-3" tabindex="-1"></a>interesting <span class="ot">&lt;-</span> markers[[<span class="dv">1</span>]]</span>
<span id="cb116-4"><a href="marker-genes.html#cb116-4" tabindex="-1"></a>best_set <span class="ot">&lt;-</span> interesting[interesting<span class="sc">$</span>Top <span class="sc">&lt;=</span> <span class="dv">5</span>, ]</span>
<span id="cb116-5"><a href="marker-genes.html#cb116-5" tabindex="-1"></a>logFCs <span class="ot">&lt;-</span> <span class="fu">getMarkerEffects</span>(best_set)</span>
<span id="cb116-6"><a href="marker-genes.html#cb116-6" tabindex="-1"></a></span>
<span id="cb116-7"><a href="marker-genes.html#cb116-7" tabindex="-1"></a><span class="fu">pheatmap</span>(logFCs, <span class="at">breaks =</span> <span class="fu">seq</span>(<span class="sc">-</span><span class="dv">5</span>, <span class="dv">5</span>, <span class="at">length.out =</span> <span class="dv">101</span>))</span></code></pre></div>
<p><img src="BestPracticesST_files/figure-html/unnamed-chunk-53-1.png" width="480" /></p>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="marker-genes.html#cb117-1" tabindex="-1"></a><span class="co"># plot log-transformed normalized expression of top genes for one cluster</span></span>
<span id="cb117-2"><a href="marker-genes.html#cb117-2" tabindex="-1"></a>top_genes <span class="ot">&lt;-</span> <span class="fu">head</span>(<span class="fu">rownames</span>(interesting))</span>
<span id="cb117-3"><a href="marker-genes.html#cb117-3" tabindex="-1"></a></span>
<span id="cb117-4"><a href="marker-genes.html#cb117-4" tabindex="-1"></a><span class="fu">plotExpression</span>(spe, <span class="at">x =</span> <span class="st">&quot;label&quot;</span>, <span class="at">features =</span> top_genes)</span></code></pre></div>
<p><img src="BestPracticesST_files/figure-html/unnamed-chunk-54-1.png" width="672" /></p>

</div>
</div>
<h3> References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Lun2016" class="csl-entry">
Lun, Aaron T L, Davis J McCarthy, and John C Marioni. 2016. <span>“A Step-by-Step Workflow for Low-Level Analysis of Single-Cell <span class="nocase">RNA-seq</span> Data with <span>Bioconductor</span>.”</span> <em>F1000Research</em> 5: 2122. <a href="https://doi.org/10.12688/f1000research.9501.2">https://doi.org/10.12688/f1000research.9501.2</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="clustering.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="spot-level-deconvolution.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
},
"lib_dir": "book_assets"
});
});
</script>

</body>

</html>
