<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 Spatially-resolved transcriptomics | Best Practices for Spatial Transcriptomics Analysis with Bioconductor</title>
  <meta name="description" content="Online book ‘Best Practices for Spatial Transcriptomics Analysis with Bioconductor’" />
  <meta name="generator" content="bookdown 0.35 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 Spatially-resolved transcriptomics | Best Practices for Spatial Transcriptomics Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="Online book ‘Best Practices for Spatial Transcriptomics Analysis with Bioconductor’" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Spatially-resolved transcriptomics | Best Practices for Spatial Transcriptomics Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Online book ‘Best Practices for Spatial Transcriptomics Analysis with Bioconductor’" />
  



<meta name="date" content="2023-09-12" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="introduction.html"/>
<link rel="next" href="bioconductor-data-classes.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 1em;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Best Practices ST</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#contents"><i class="fa fa-check"></i><b>1.1</b> Contents</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#who-this-book-is-for"><i class="fa fa-check"></i><b>1.2</b> Who this book is for</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#bioconductor"><i class="fa fa-check"></i><b>1.3</b> Bioconductor</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#contributions"><i class="fa fa-check"></i><b>1.4</b> Contributions</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html"><i class="fa fa-check"></i><b>2</b> Spatially-resolved transcriptomics</a>
<ul>
<li class="chapter" data-level="2.1" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#overview"><i class="fa fa-check"></i><b>2.1</b> Overview</a></li>
<li class="chapter" data-level="2.2" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#sequencing-based-platforms"><i class="fa fa-check"></i><b>2.2</b> Sequencing-based platforms</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#x-genomics-visium"><i class="fa fa-check"></i><b>2.2.1</b> 10x Genomics Visium</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#molecule-based-platforms"><i class="fa fa-check"></i><b>2.3</b> Molecule-based platforms</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#x-genomics-xenium"><i class="fa fa-check"></i><b>2.3.1</b> 10x Genomics Xenium</a></li>
<li class="chapter" data-level="2.3.2" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#vizgen-merscope"><i class="fa fa-check"></i><b>2.3.2</b> Vizgen MERSCOPE</a></li>
<li class="chapter" data-level="2.3.3" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#nanostring-cosmx"><i class="fa fa-check"></i><b>2.3.3</b> NanoString CosMx</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="bioconductor-data-classes.html"><a href="bioconductor-data-classes.html"><i class="fa fa-check"></i><b>3</b> Bioconductor data classes</a>
<ul>
<li class="chapter" data-level="3.1" data-path="bioconductor-data-classes.html"><a href="bioconductor-data-classes.html#spatialexperiment-class"><i class="fa fa-check"></i><b>3.1</b> SpatialExperiment class</a></li>
<li class="chapter" data-level="3.2" data-path="bioconductor-data-classes.html"><a href="bioconductor-data-classes.html#molecule-based-data"><i class="fa fa-check"></i><b>3.2</b> Molecule-based data</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="bioconductor-data-classes.html"><a href="bioconductor-data-classes.html#moleculeexperiment"><i class="fa fa-check"></i><b>3.2.1</b> MoleculeExperiment</a></li>
<li class="chapter" data-level="3.2.2" data-path="bioconductor-data-classes.html"><a href="bioconductor-data-classes.html#spatialfeatureexperiment"><i class="fa fa-check"></i><b>3.2.2</b> SpatialFeatureExperiment</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>II Analysis steps</b></span></li>
<li class="chapter" data-level="4" data-path="analysis-steps.html"><a href="analysis-steps.html"><i class="fa fa-check"></i><b>4</b> Analysis steps</a>
<ul>
<li class="chapter" data-level="4.1" data-path="analysis-steps.html"><a href="analysis-steps.html#load-data"><i class="fa fa-check"></i><b>4.1</b> Load data</a></li>
<li class="chapter" data-level="4.2" data-path="analysis-steps.html"><a href="analysis-steps.html#spatialexperiment-object"><i class="fa fa-check"></i><b>4.2</b> SpatialExperiment object</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>5</b> Quality control</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quality-control.html"><a href="quality-control.html#background"><i class="fa fa-check"></i><b>5.1</b> Background</a></li>
<li class="chapter" data-level="5.2" data-path="quality-control.html"><a href="quality-control.html#load-data-1"><i class="fa fa-check"></i><b>5.2</b> Load data</a></li>
<li class="chapter" data-level="5.3" data-path="quality-control.html"><a href="quality-control.html#plot-data"><i class="fa fa-check"></i><b>5.3</b> Plot data</a></li>
<li class="chapter" data-level="5.4" data-path="quality-control.html"><a href="quality-control.html#calculate-qc-metrics"><i class="fa fa-check"></i><b>5.4</b> Calculate QC metrics</a></li>
<li class="chapter" data-level="5.5" data-path="quality-control.html"><a href="quality-control.html#selecting-thresholds"><i class="fa fa-check"></i><b>5.5</b> Selecting thresholds</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="quality-control.html"><a href="quality-control.html#library-size"><i class="fa fa-check"></i><b>5.5.1</b> Library size</a></li>
<li class="chapter" data-level="5.5.2" data-path="quality-control.html"><a href="quality-control.html#number-of-expressed-features"><i class="fa fa-check"></i><b>5.5.2</b> Number of expressed features</a></li>
<li class="chapter" data-level="5.5.3" data-path="quality-control.html"><a href="quality-control.html#proportion-of-mitochondrial-reads"><i class="fa fa-check"></i><b>5.5.3</b> Proportion of mitochondrial reads</a></li>
<li class="chapter" data-level="5.5.4" data-path="quality-control.html"><a href="quality-control.html#number-of-cells-per-spot"><i class="fa fa-check"></i><b>5.5.4</b> Number of cells per spot</a></li>
<li class="chapter" data-level="5.5.5" data-path="quality-control.html"><a href="quality-control.html#remove-low-quality-spots"><i class="fa fa-check"></i><b>5.5.5</b> Remove low-quality spots</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="quality-control.html"><a href="quality-control.html#zero-cell-and-single-cell-spots"><i class="fa fa-check"></i><b>5.6</b> Zero-cell and single-cell spots</a></li>
<li class="chapter" data-level="5.7" data-path="quality-control.html"><a href="quality-control.html#quality-control-at-gene-level"><i class="fa fa-check"></i><b>5.7</b> Quality control at gene level</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="normalization.html"><a href="normalization.html"><i class="fa fa-check"></i><b>6</b> Normalization</a>
<ul>
<li class="chapter" data-level="6.1" data-path="normalization.html"><a href="normalization.html#background-1"><i class="fa fa-check"></i><b>6.1</b> Background</a></li>
<li class="chapter" data-level="6.2" data-path="normalization.html"><a href="normalization.html#previous-steps"><i class="fa fa-check"></i><b>6.2</b> Previous steps</a></li>
<li class="chapter" data-level="6.3" data-path="normalization.html"><a href="normalization.html#logcounts"><i class="fa fa-check"></i><b>6.3</b> Logcounts</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="feature-selection.html"><a href="feature-selection.html"><i class="fa fa-check"></i><b>7</b> Feature selection</a>
<ul>
<li class="chapter" data-level="7.1" data-path="feature-selection.html"><a href="feature-selection.html#background-2"><i class="fa fa-check"></i><b>7.1</b> Background</a></li>
<li class="chapter" data-level="7.2" data-path="feature-selection.html"><a href="feature-selection.html#previous-steps-1"><i class="fa fa-check"></i><b>7.2</b> Previous steps</a></li>
<li class="chapter" data-level="7.3" data-path="feature-selection.html"><a href="feature-selection.html#highly-variable-genes-hvgs"><i class="fa fa-check"></i><b>7.3</b> Highly variable genes (HVGs)</a></li>
<li class="chapter" data-level="7.4" data-path="feature-selection.html"><a href="feature-selection.html#spatially-variable-genes-svgs"><i class="fa fa-check"></i><b>7.4</b> Spatially variable genes (SVGs)</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html"><i class="fa fa-check"></i><b>8</b> Dimensionality reduction</a>
<ul>
<li class="chapter" data-level="8.1" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#background-3"><i class="fa fa-check"></i><b>8.1</b> Background</a></li>
<li class="chapter" data-level="8.2" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#previous-steps-2"><i class="fa fa-check"></i><b>8.2</b> Previous steps</a></li>
<li class="chapter" data-level="8.3" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>8.3</b> Principal component analysis (PCA)</a></li>
<li class="chapter" data-level="8.4" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#uniform-manifold-approximation-and-projection-umap"><i class="fa fa-check"></i><b>8.4</b> Uniform Manifold Approximation and Projection (UMAP)</a></li>
<li class="chapter" data-level="8.5" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#visualizations"><i class="fa fa-check"></i><b>8.5</b> Visualizations</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>9</b> Clustering</a>
<ul>
<li class="chapter" data-level="9.1" data-path="clustering.html"><a href="clustering.html#overview-1"><i class="fa fa-check"></i><b>9.1</b> Overview</a></li>
<li class="chapter" data-level="9.2" data-path="clustering.html"><a href="clustering.html#previous-steps-3"><i class="fa fa-check"></i><b>9.2</b> Previous steps</a></li>
<li class="chapter" data-level="9.3" data-path="clustering.html"><a href="clustering.html#non-spatial-clustering-on-hvgs"><i class="fa fa-check"></i><b>9.3</b> Non-spatial clustering on HVGs</a></li>
<li class="chapter" data-level="9.4" data-path="clustering.html"><a href="clustering.html#spatially-aware-clustering"><i class="fa fa-check"></i><b>9.4</b> Spatially-aware clustering</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="marker-genes.html"><a href="marker-genes.html"><i class="fa fa-check"></i><b>10</b> Marker genes</a>
<ul>
<li class="chapter" data-level="10.1" data-path="marker-genes.html"><a href="marker-genes.html#background-4"><i class="fa fa-check"></i><b>10.1</b> Background</a></li>
<li class="chapter" data-level="10.2" data-path="marker-genes.html"><a href="marker-genes.html#previous-steps-4"><i class="fa fa-check"></i><b>10.2</b> Previous steps</a></li>
<li class="chapter" data-level="10.3" data-path="marker-genes.html"><a href="marker-genes.html#marker-genes-1"><i class="fa fa-check"></i><b>10.3</b> Marker genes</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="spot-level-deconvolution.html"><a href="spot-level-deconvolution.html"><i class="fa fa-check"></i><b>11</b> Spot-level deconvolution</a>
<ul>
<li class="chapter" data-level="11.1" data-path="spot-level-deconvolution.html"><a href="spot-level-deconvolution.html#background-5"><i class="fa fa-check"></i><b>11.1</b> Background</a></li>
<li class="chapter" data-level="11.2" data-path="spot-level-deconvolution.html"><a href="spot-level-deconvolution.html#previous-steps-5"><i class="fa fa-check"></i><b>11.2</b> Previous steps</a></li>
<li class="chapter" data-level="11.3" data-path="spot-level-deconvolution.html"><a href="spot-level-deconvolution.html#number-of-cells-per-spot-1"><i class="fa fa-check"></i><b>11.3</b> Number of cells per spot</a></li>
</ul></li>
<li class="part"><span><b>III Workflows</b></span></li>
<li class="chapter" data-level="12" data-path="workflows.html"><a href="workflows.html"><i class="fa fa-check"></i><b>12</b> Workflows</a></li>
<li class="chapter" data-level="13" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html"><i class="fa fa-check"></i><b>13</b> Human DLPFC workflow</a>
<ul>
<li class="chapter" data-level="13.1" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#description-of-dataset"><i class="fa fa-check"></i><b>13.1</b> Description of dataset</a></li>
<li class="chapter" data-level="13.2" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#load-data-2"><i class="fa fa-check"></i><b>13.2</b> Load data</a></li>
<li class="chapter" data-level="13.3" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#plot-data-1"><i class="fa fa-check"></i><b>13.3</b> Plot data</a></li>
<li class="chapter" data-level="13.4" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#quality-control-qc"><i class="fa fa-check"></i><b>13.4</b> Quality control (QC)</a></li>
<li class="chapter" data-level="13.5" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#normalization-1"><i class="fa fa-check"></i><b>13.5</b> Normalization</a></li>
<li class="chapter" data-level="13.6" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#feature-selection-1"><i class="fa fa-check"></i><b>13.6</b> Feature selection</a></li>
<li class="chapter" data-level="13.7" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#spatially-aware-feature-selection"><i class="fa fa-check"></i><b>13.7</b> Spatially-aware feature selection</a></li>
<li class="chapter" data-level="13.8" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>13.8</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="13.9" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#clustering-1"><i class="fa fa-check"></i><b>13.9</b> Clustering</a></li>
<li class="chapter" data-level="13.10" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#marker-genes-2"><i class="fa fa-check"></i><b>13.10</b> Marker genes</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html"><i class="fa fa-check"></i><b>14</b> Mouse coronal workflow</a>
<ul>
<li class="chapter" data-level="14.1" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#description-of-dataset-1"><i class="fa fa-check"></i><b>14.1</b> Description of dataset</a></li>
<li class="chapter" data-level="14.2" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#load-data-3"><i class="fa fa-check"></i><b>14.2</b> Load data</a></li>
<li class="chapter" data-level="14.3" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#plot-data-2"><i class="fa fa-check"></i><b>14.3</b> Plot data</a></li>
<li class="chapter" data-level="14.4" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#quality-control-qc-1"><i class="fa fa-check"></i><b>14.4</b> Quality control (QC)</a></li>
<li class="chapter" data-level="14.5" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#normalization-2"><i class="fa fa-check"></i><b>14.5</b> Normalization</a></li>
<li class="chapter" data-level="14.6" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#feature-selection-2"><i class="fa fa-check"></i><b>14.6</b> Feature selection</a></li>
<li class="chapter" data-level="14.7" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#spatially-aware-feature-selection-1"><i class="fa fa-check"></i><b>14.7</b> Spatially-aware feature selection</a></li>
<li class="chapter" data-level="14.8" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#dimensionality-reduction-2"><i class="fa fa-check"></i><b>14.8</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="14.9" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#clustering-2"><i class="fa fa-check"></i><b>14.9</b> Clustering</a></li>
<li class="chapter" data-level="14.10" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#marker-genes-3"><i class="fa fa-check"></i><b>14.10</b> Marker genes</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html"><i class="fa fa-check"></i><b>15</b> spatialLIBD workflow</a>
<ul>
<li class="chapter" data-level="15.1" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#overview-2"><i class="fa fa-check"></i><b>15.1</b> Overview</a></li>
<li class="chapter" data-level="15.2" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#spatiallibd"><i class="fa fa-check"></i><b>15.2</b> spatialLIBD</a>
<ul>
<li class="chapter" data-level="15.2.1" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#why-use-spatiallibd"><i class="fa fa-check"></i><b>15.2.1</b> Why use spatialLIBD?</a></li>
<li class="chapter" data-level="15.2.2" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#want-to-learn-more-about-spatiallibd"><i class="fa fa-check"></i><b>15.2.2</b> Want to learn more about spatialLIBD?</a></li>
</ul></li>
<li class="chapter" data-level="15.3" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#code-prerequisites"><i class="fa fa-check"></i><b>15.3</b> Code prerequisites</a></li>
<li class="chapter" data-level="15.4" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#prepare-for-spatiallibd"><i class="fa fa-check"></i><b>15.4</b> Prepare for spatialLIBD</a>
<ul>
<li class="chapter" data-level="15.4.1" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#basic-information"><i class="fa fa-check"></i><b>15.4.1</b> Basic information</a></li>
<li class="chapter" data-level="15.4.2" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#gene-annotation"><i class="fa fa-check"></i><b>15.4.2</b> Gene annotation</a></li>
<li class="chapter" data-level="15.4.3" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#extra-information-and-filtering"><i class="fa fa-check"></i><b>15.4.3</b> Extra information and filtering</a></li>
</ul></li>
<li class="chapter" data-level="15.5" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#explore-the-data"><i class="fa fa-check"></i><b>15.5</b> Explore the data</a></li>
<li class="chapter" data-level="15.6" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#sharing-your-website"><i class="fa fa-check"></i><b>15.6</b> Sharing your website</a></li>
<li class="chapter" data-level="15.7" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#wrapping-up"><i class="fa fa-check"></i><b>15.7</b> Wrapping up</a></li>
<li class="chapter" data-level="" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#r-session-information"><i class="fa fa-check"></i>R session information</a></li>
</ul></li>
<li class="part"><span><b>IV Appendix</b></span></li>
<li class="chapter" data-level="16" data-path="related-resources.html"><a href="related-resources.html"><i class="fa fa-check"></i><b>16</b> Related resources</a>
<ul>
<li class="chapter" data-level="16.1" data-path="related-resources.html"><a href="related-resources.html#data-preprocessing-procedures"><i class="fa fa-check"></i><b>16.1</b> Data preprocessing procedures</a></li>
<li class="chapter" data-level="16.2" data-path="related-resources.html"><a href="related-resources.html#resources-for-other-spatially-resolved-platforms"><i class="fa fa-check"></i><b>16.2</b> Resources for other spatially-resolved platforms</a></li>
<li class="chapter" data-level="16.3" data-path="related-resources.html"><a href="related-resources.html#data-structures"><i class="fa fa-check"></i><b>16.3</b> Data structures</a></li>
<li class="chapter" data-level="16.4" data-path="related-resources.html"><a href="related-resources.html#statistical-concepts"><i class="fa fa-check"></i><b>16.4</b> Statistical concepts</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="contributors-and-acknowledgments.html"><a href="contributors-and-acknowledgments.html"><i class="fa fa-check"></i><b>17</b> Contributors and acknowledgments</a>
<ul>
<li class="chapter" data-level="17.1" data-path="contributors-and-acknowledgments.html"><a href="contributors-and-acknowledgments.html#best-practices-for-spatial-transcriptomics-analysis-with-bioconductor"><i class="fa fa-check"></i><b>17.1</b> Best Practices for Spatial Transcriptomics Analysis with Bioconductor</a></li>
<li class="chapter" data-level="17.2" data-path="contributors-and-acknowledgments.html"><a href="contributors-and-acknowledgments.html#spatialexperiment"><i class="fa fa-check"></i><b>17.2</b> SpatialExperiment</a></li>
<li class="chapter" data-level="17.3" data-path="contributors-and-acknowledgments.html"><a href="contributors-and-acknowledgments.html#github-actions-workflow"><i class="fa fa-check"></i><b>17.3</b> GitHub Actions workflow</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>18</b> References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Best Practices for Spatial Transcriptomics Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="spatially-resolved-transcriptomics" class="section level1 hasAnchor" number="2">
<h1><span class="header-section-number">Chapter 2</span> Spatially-resolved transcriptomics<a href="spatially-resolved-transcriptomics.html#spatially-resolved-transcriptomics" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Spatially-resolved transcriptomics (SRT) refers to high-throughput genomic technologies that enable the measurement of large sets of genes at a large number of spatial locations (e.g. up to thousands of genes at thousands of measurement locations), usually on two-dimensional tissue sections. A number of technological platforms have been developed, each with their unique advantages. SRT was named the <a href="https://doi.org/10.1038/s41592-020-01042-x">Method of the Year 2020</a> by Nature Methods, and has found widespread application in numerous biological systems.</p>
<p>For recent reviews of available platforms, computational analysis methods, and outstanding challenges, see <a href="https://www.science.org/doi/10.1126/science.abq4964">Bressan et al. 2023</a> or <a href="https://www.nature.com/articles/s41592-022-01409-2">Moses et al. 2022</a>.</p>
<div id="overview" class="section level2 hasAnchor" number="2.1">
<h2><span class="header-section-number">2.1</span> Overview<a href="spatially-resolved-transcriptomics.html#overview" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In this book, we concentrate on commercially available platforms, since these are most widely used and available. Here, we provide a short overview and introductory details on the platforms used to generate the datasets discussed in this book.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-1"></span>
<img src="images/Maynard2020_spatial.png" alt="Illustration of potential uses of spatially-resolved transcriptomics data and how it can be combined with single cell/nucleus RNA-seq to study tissues of interest such as the human brain. Source: [Maynard et al, _Neuropsychopharmacol._, 2020](https://doi.org/10.1038/s41386-019-0484-7)." width="100%" />
<p class="caption">
Figure 2.1: Illustration of potential uses of spatially-resolved transcriptomics data and how it can be combined with single cell/nucleus RNA-seq to study tissues of interest such as the human brain. Source: <a href="https://doi.org/10.1038/s41386-019-0484-7">Maynard et al, <em>Neuropsychopharmacol.</em>, 2020</a>.
</p>
</div>
</div>
<div id="sequencing-based-platforms" class="section level2 hasAnchor" number="2.2">
<h2><span class="header-section-number">2.2</span> Sequencing-based platforms<a href="spatially-resolved-transcriptomics.html#sequencing-based-platforms" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Sequencing-based platforms capture RNAs at a set of spatial measurement locations, tag RNAs with spatial barcodes, and generate a readout by next-generation sequencing.</p>
<p>These platforms can provide up to full-transcriptome feature resolution due to the use of next-generation sequencing. The spatial resolution depends on the size of the capture locations and tissue cell density, and may include multiple cells per measurement location.</p>
<div id="x-genomics-visium" class="section level3 hasAnchor" number="2.2.1">
<h3><span class="header-section-number">2.2.1</span> 10x Genomics Visium<a href="spatially-resolved-transcriptomics.html#x-genomics-visium" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The <a href="https://www.10xgenomics.com/products/spatial-gene-expression">10x Genomics Visium</a> platform measures transcriptome-wide gene expression at a grid of spatial locations (referred to as spots) on a tissue capture area. Either fresh-frozen or formalin-fixed paraffin-embedded (FFPE) tissue may be used. Each spot contains millions of spatially-barcoded capture oligonucleotides, which bind to mRNAs from the tissue. A cDNA library is then generated for sequencing, which includes the spatial barcodes, allowing reads to be mapped back to their spatial locations.</p>
<p>The array dimensions are 6.5mm x 6.5mm, with around 5000 barcoded spots. Spots are 55µm in diameter and spaced 100µm center-to-center in a hexagonal grid arrangement. The number of cells per spot depends on the tissue cell density, e.g. around 0-10 for human brain or ~50 for mouse brain. Each Visium slide contains four tissue capture areas. The following figure provides an illustration.</p>
<p>Histology images generated from hematoxylin and eosin (H&amp;E) staining can be used to identify anatomical and cell morphological features such as the number of cells per spot.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:Visium-schematic"></span>
<img src="images/Visium.png" alt="Schematic of 10x Genomics Visium platform. Source: [10x Genomics Visium website](https://www.10xgenomics.com/spatial-transcriptomics/)" width="574" />
<p class="caption">
Figure 2.2: Schematic of 10x Genomics Visium platform. Source: <a href="https://www.10xgenomics.com/spatial-transcriptomics/">10x Genomics Visium website</a>
</p>
</div>
</div>
</div>
<div id="molecule-based-platforms" class="section level2 hasAnchor" number="2.3">
<h2><span class="header-section-number">2.3</span> Molecule-based platforms<a href="spatially-resolved-transcriptomics.html#molecule-based-platforms" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Molecule-based platforms identify individual RNAs by in situ hybridization (ISH) or in situ sequencing (ISS) at up to sub-cellular spatial resolution. Counts may be aggregated to cellular (single-cell) spatial resolution by applying manual or data-driven cell segmentation algorithms.</p>
<p>Feature resolution is generally on the order of 100s to 1000s of RNAs, which may be selected as targeted panels of RNAs of interest.</p>
<div id="x-genomics-xenium" class="section level3 hasAnchor" number="2.3.1">
<h3><span class="header-section-number">2.3.1</span> 10x Genomics Xenium<a href="spatially-resolved-transcriptomics.html#x-genomics-xenium" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Details available on the <a href="https://www.10xgenomics.com/platforms/xenium">10x Genomics</a> website.</p>
</div>
<div id="vizgen-merscope" class="section level3 hasAnchor" number="2.3.2">
<h3><span class="header-section-number">2.3.2</span> Vizgen MERSCOPE<a href="spatially-resolved-transcriptomics.html#vizgen-merscope" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Details available on the <a href="https://vizgen.com/products/">Vizgen</a> website.</p>
</div>
<div id="nanostring-cosmx" class="section level3 hasAnchor" number="2.3.3">
<h3><span class="header-section-number">2.3.3</span> NanoString CosMx<a href="spatially-resolved-transcriptomics.html#nanostring-cosmx" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Details available on the <a href="https://nanostring.com/products/cosmx-spatial-molecular-imager/">NanoString</a> website.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introduction.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="bioconductor-data-classes.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
},
"lib_dir": "book_assets"
});
});
</script>

</body>

</html>
