<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 15 spatialLIBD workflow | Best Practices for Spatial Transcriptomics Analysis with Bioconductor</title>
  <meta name="description" content="Online book ‘Best Practices for Spatial Transcriptomics Analysis with Bioconductor’" />
  <meta name="generator" content="bookdown 0.35 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 15 spatialLIBD workflow | Best Practices for Spatial Transcriptomics Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="Online book ‘Best Practices for Spatial Transcriptomics Analysis with Bioconductor’" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 15 spatialLIBD workflow | Best Practices for Spatial Transcriptomics Analysis with Bioconductor" />
  
  <meta name="twitter:description" content="Online book ‘Best Practices for Spatial Transcriptomics Analysis with Bioconductor’" />
  



<meta name="date" content="2023-09-12" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="mouse-coronal-workflow.html"/>
<link rel="next" href="related-resources.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 1em;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Best Practices ST</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#contents"><i class="fa fa-check"></i><b>1.1</b> Contents</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#who-this-book-is-for"><i class="fa fa-check"></i><b>1.2</b> Who this book is for</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#bioconductor"><i class="fa fa-check"></i><b>1.3</b> Bioconductor</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#contributions"><i class="fa fa-check"></i><b>1.4</b> Contributions</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html"><i class="fa fa-check"></i><b>2</b> Spatially-resolved transcriptomics</a>
<ul>
<li class="chapter" data-level="2.1" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#overview"><i class="fa fa-check"></i><b>2.1</b> Overview</a></li>
<li class="chapter" data-level="2.2" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#sequencing-based-platforms"><i class="fa fa-check"></i><b>2.2</b> Sequencing-based platforms</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#x-genomics-visium"><i class="fa fa-check"></i><b>2.2.1</b> 10x Genomics Visium</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#molecule-based-platforms"><i class="fa fa-check"></i><b>2.3</b> Molecule-based platforms</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#x-genomics-xenium"><i class="fa fa-check"></i><b>2.3.1</b> 10x Genomics Xenium</a></li>
<li class="chapter" data-level="2.3.2" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#vizgen-merscope"><i class="fa fa-check"></i><b>2.3.2</b> Vizgen MERSCOPE</a></li>
<li class="chapter" data-level="2.3.3" data-path="spatially-resolved-transcriptomics.html"><a href="spatially-resolved-transcriptomics.html#nanostring-cosmx"><i class="fa fa-check"></i><b>2.3.3</b> NanoString CosMx</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="bioconductor-data-classes.html"><a href="bioconductor-data-classes.html"><i class="fa fa-check"></i><b>3</b> Bioconductor data classes</a>
<ul>
<li class="chapter" data-level="3.1" data-path="bioconductor-data-classes.html"><a href="bioconductor-data-classes.html#spatialexperiment-class"><i class="fa fa-check"></i><b>3.1</b> SpatialExperiment class</a></li>
<li class="chapter" data-level="3.2" data-path="bioconductor-data-classes.html"><a href="bioconductor-data-classes.html#molecule-based-data"><i class="fa fa-check"></i><b>3.2</b> Molecule-based data</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="bioconductor-data-classes.html"><a href="bioconductor-data-classes.html#moleculeexperiment"><i class="fa fa-check"></i><b>3.2.1</b> MoleculeExperiment</a></li>
<li class="chapter" data-level="3.2.2" data-path="bioconductor-data-classes.html"><a href="bioconductor-data-classes.html#spatialfeatureexperiment"><i class="fa fa-check"></i><b>3.2.2</b> SpatialFeatureExperiment</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>II Analysis steps</b></span></li>
<li class="chapter" data-level="4" data-path="analysis-steps.html"><a href="analysis-steps.html"><i class="fa fa-check"></i><b>4</b> Analysis steps</a>
<ul>
<li class="chapter" data-level="4.1" data-path="analysis-steps.html"><a href="analysis-steps.html#load-data"><i class="fa fa-check"></i><b>4.1</b> Load data</a></li>
<li class="chapter" data-level="4.2" data-path="analysis-steps.html"><a href="analysis-steps.html#spatialexperiment-object"><i class="fa fa-check"></i><b>4.2</b> SpatialExperiment object</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>5</b> Quality control</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quality-control.html"><a href="quality-control.html#background"><i class="fa fa-check"></i><b>5.1</b> Background</a></li>
<li class="chapter" data-level="5.2" data-path="quality-control.html"><a href="quality-control.html#load-data-1"><i class="fa fa-check"></i><b>5.2</b> Load data</a></li>
<li class="chapter" data-level="5.3" data-path="quality-control.html"><a href="quality-control.html#plot-data"><i class="fa fa-check"></i><b>5.3</b> Plot data</a></li>
<li class="chapter" data-level="5.4" data-path="quality-control.html"><a href="quality-control.html#calculate-qc-metrics"><i class="fa fa-check"></i><b>5.4</b> Calculate QC metrics</a></li>
<li class="chapter" data-level="5.5" data-path="quality-control.html"><a href="quality-control.html#selecting-thresholds"><i class="fa fa-check"></i><b>5.5</b> Selecting thresholds</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="quality-control.html"><a href="quality-control.html#library-size"><i class="fa fa-check"></i><b>5.5.1</b> Library size</a></li>
<li class="chapter" data-level="5.5.2" data-path="quality-control.html"><a href="quality-control.html#number-of-expressed-features"><i class="fa fa-check"></i><b>5.5.2</b> Number of expressed features</a></li>
<li class="chapter" data-level="5.5.3" data-path="quality-control.html"><a href="quality-control.html#proportion-of-mitochondrial-reads"><i class="fa fa-check"></i><b>5.5.3</b> Proportion of mitochondrial reads</a></li>
<li class="chapter" data-level="5.5.4" data-path="quality-control.html"><a href="quality-control.html#number-of-cells-per-spot"><i class="fa fa-check"></i><b>5.5.4</b> Number of cells per spot</a></li>
<li class="chapter" data-level="5.5.5" data-path="quality-control.html"><a href="quality-control.html#remove-low-quality-spots"><i class="fa fa-check"></i><b>5.5.5</b> Remove low-quality spots</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="quality-control.html"><a href="quality-control.html#zero-cell-and-single-cell-spots"><i class="fa fa-check"></i><b>5.6</b> Zero-cell and single-cell spots</a></li>
<li class="chapter" data-level="5.7" data-path="quality-control.html"><a href="quality-control.html#quality-control-at-gene-level"><i class="fa fa-check"></i><b>5.7</b> Quality control at gene level</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="normalization.html"><a href="normalization.html"><i class="fa fa-check"></i><b>6</b> Normalization</a>
<ul>
<li class="chapter" data-level="6.1" data-path="normalization.html"><a href="normalization.html#background-1"><i class="fa fa-check"></i><b>6.1</b> Background</a></li>
<li class="chapter" data-level="6.2" data-path="normalization.html"><a href="normalization.html#previous-steps"><i class="fa fa-check"></i><b>6.2</b> Previous steps</a></li>
<li class="chapter" data-level="6.3" data-path="normalization.html"><a href="normalization.html#logcounts"><i class="fa fa-check"></i><b>6.3</b> Logcounts</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="feature-selection.html"><a href="feature-selection.html"><i class="fa fa-check"></i><b>7</b> Feature selection</a>
<ul>
<li class="chapter" data-level="7.1" data-path="feature-selection.html"><a href="feature-selection.html#background-2"><i class="fa fa-check"></i><b>7.1</b> Background</a></li>
<li class="chapter" data-level="7.2" data-path="feature-selection.html"><a href="feature-selection.html#previous-steps-1"><i class="fa fa-check"></i><b>7.2</b> Previous steps</a></li>
<li class="chapter" data-level="7.3" data-path="feature-selection.html"><a href="feature-selection.html#highly-variable-genes-hvgs"><i class="fa fa-check"></i><b>7.3</b> Highly variable genes (HVGs)</a></li>
<li class="chapter" data-level="7.4" data-path="feature-selection.html"><a href="feature-selection.html#spatially-variable-genes-svgs"><i class="fa fa-check"></i><b>7.4</b> Spatially variable genes (SVGs)</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html"><i class="fa fa-check"></i><b>8</b> Dimensionality reduction</a>
<ul>
<li class="chapter" data-level="8.1" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#background-3"><i class="fa fa-check"></i><b>8.1</b> Background</a></li>
<li class="chapter" data-level="8.2" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#previous-steps-2"><i class="fa fa-check"></i><b>8.2</b> Previous steps</a></li>
<li class="chapter" data-level="8.3" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>8.3</b> Principal component analysis (PCA)</a></li>
<li class="chapter" data-level="8.4" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#uniform-manifold-approximation-and-projection-umap"><i class="fa fa-check"></i><b>8.4</b> Uniform Manifold Approximation and Projection (UMAP)</a></li>
<li class="chapter" data-level="8.5" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#visualizations"><i class="fa fa-check"></i><b>8.5</b> Visualizations</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>9</b> Clustering</a>
<ul>
<li class="chapter" data-level="9.1" data-path="clustering.html"><a href="clustering.html#overview-1"><i class="fa fa-check"></i><b>9.1</b> Overview</a></li>
<li class="chapter" data-level="9.2" data-path="clustering.html"><a href="clustering.html#previous-steps-3"><i class="fa fa-check"></i><b>9.2</b> Previous steps</a></li>
<li class="chapter" data-level="9.3" data-path="clustering.html"><a href="clustering.html#non-spatial-clustering-on-hvgs"><i class="fa fa-check"></i><b>9.3</b> Non-spatial clustering on HVGs</a></li>
<li class="chapter" data-level="9.4" data-path="clustering.html"><a href="clustering.html#spatially-aware-clustering"><i class="fa fa-check"></i><b>9.4</b> Spatially-aware clustering</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="marker-genes.html"><a href="marker-genes.html"><i class="fa fa-check"></i><b>10</b> Marker genes</a>
<ul>
<li class="chapter" data-level="10.1" data-path="marker-genes.html"><a href="marker-genes.html#background-4"><i class="fa fa-check"></i><b>10.1</b> Background</a></li>
<li class="chapter" data-level="10.2" data-path="marker-genes.html"><a href="marker-genes.html#previous-steps-4"><i class="fa fa-check"></i><b>10.2</b> Previous steps</a></li>
<li class="chapter" data-level="10.3" data-path="marker-genes.html"><a href="marker-genes.html#marker-genes-1"><i class="fa fa-check"></i><b>10.3</b> Marker genes</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="spot-level-deconvolution.html"><a href="spot-level-deconvolution.html"><i class="fa fa-check"></i><b>11</b> Spot-level deconvolution</a>
<ul>
<li class="chapter" data-level="11.1" data-path="spot-level-deconvolution.html"><a href="spot-level-deconvolution.html#background-5"><i class="fa fa-check"></i><b>11.1</b> Background</a></li>
<li class="chapter" data-level="11.2" data-path="spot-level-deconvolution.html"><a href="spot-level-deconvolution.html#previous-steps-5"><i class="fa fa-check"></i><b>11.2</b> Previous steps</a></li>
<li class="chapter" data-level="11.3" data-path="spot-level-deconvolution.html"><a href="spot-level-deconvolution.html#number-of-cells-per-spot-1"><i class="fa fa-check"></i><b>11.3</b> Number of cells per spot</a></li>
</ul></li>
<li class="part"><span><b>III Workflows</b></span></li>
<li class="chapter" data-level="12" data-path="workflows.html"><a href="workflows.html"><i class="fa fa-check"></i><b>12</b> Workflows</a></li>
<li class="chapter" data-level="13" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html"><i class="fa fa-check"></i><b>13</b> Human DLPFC workflow</a>
<ul>
<li class="chapter" data-level="13.1" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#description-of-dataset"><i class="fa fa-check"></i><b>13.1</b> Description of dataset</a></li>
<li class="chapter" data-level="13.2" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#load-data-2"><i class="fa fa-check"></i><b>13.2</b> Load data</a></li>
<li class="chapter" data-level="13.3" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#plot-data-1"><i class="fa fa-check"></i><b>13.3</b> Plot data</a></li>
<li class="chapter" data-level="13.4" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#quality-control-qc"><i class="fa fa-check"></i><b>13.4</b> Quality control (QC)</a></li>
<li class="chapter" data-level="13.5" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#normalization-1"><i class="fa fa-check"></i><b>13.5</b> Normalization</a></li>
<li class="chapter" data-level="13.6" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#feature-selection-1"><i class="fa fa-check"></i><b>13.6</b> Feature selection</a></li>
<li class="chapter" data-level="13.7" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#spatially-aware-feature-selection"><i class="fa fa-check"></i><b>13.7</b> Spatially-aware feature selection</a></li>
<li class="chapter" data-level="13.8" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#dimensionality-reduction-1"><i class="fa fa-check"></i><b>13.8</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="13.9" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#clustering-1"><i class="fa fa-check"></i><b>13.9</b> Clustering</a></li>
<li class="chapter" data-level="13.10" data-path="human-dlpfc-workflow.html"><a href="human-dlpfc-workflow.html#marker-genes-2"><i class="fa fa-check"></i><b>13.10</b> Marker genes</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html"><i class="fa fa-check"></i><b>14</b> Mouse coronal workflow</a>
<ul>
<li class="chapter" data-level="14.1" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#description-of-dataset-1"><i class="fa fa-check"></i><b>14.1</b> Description of dataset</a></li>
<li class="chapter" data-level="14.2" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#load-data-3"><i class="fa fa-check"></i><b>14.2</b> Load data</a></li>
<li class="chapter" data-level="14.3" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#plot-data-2"><i class="fa fa-check"></i><b>14.3</b> Plot data</a></li>
<li class="chapter" data-level="14.4" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#quality-control-qc-1"><i class="fa fa-check"></i><b>14.4</b> Quality control (QC)</a></li>
<li class="chapter" data-level="14.5" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#normalization-2"><i class="fa fa-check"></i><b>14.5</b> Normalization</a></li>
<li class="chapter" data-level="14.6" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#feature-selection-2"><i class="fa fa-check"></i><b>14.6</b> Feature selection</a></li>
<li class="chapter" data-level="14.7" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#spatially-aware-feature-selection-1"><i class="fa fa-check"></i><b>14.7</b> Spatially-aware feature selection</a></li>
<li class="chapter" data-level="14.8" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#dimensionality-reduction-2"><i class="fa fa-check"></i><b>14.8</b> Dimensionality reduction</a></li>
<li class="chapter" data-level="14.9" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#clustering-2"><i class="fa fa-check"></i><b>14.9</b> Clustering</a></li>
<li class="chapter" data-level="14.10" data-path="mouse-coronal-workflow.html"><a href="mouse-coronal-workflow.html#marker-genes-3"><i class="fa fa-check"></i><b>14.10</b> Marker genes</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html"><i class="fa fa-check"></i><b>15</b> spatialLIBD workflow</a>
<ul>
<li class="chapter" data-level="15.1" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#overview-2"><i class="fa fa-check"></i><b>15.1</b> Overview</a></li>
<li class="chapter" data-level="15.2" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#spatiallibd"><i class="fa fa-check"></i><b>15.2</b> spatialLIBD</a>
<ul>
<li class="chapter" data-level="15.2.1" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#why-use-spatiallibd"><i class="fa fa-check"></i><b>15.2.1</b> Why use spatialLIBD?</a></li>
<li class="chapter" data-level="15.2.2" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#want-to-learn-more-about-spatiallibd"><i class="fa fa-check"></i><b>15.2.2</b> Want to learn more about spatialLIBD?</a></li>
</ul></li>
<li class="chapter" data-level="15.3" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#code-prerequisites"><i class="fa fa-check"></i><b>15.3</b> Code prerequisites</a></li>
<li class="chapter" data-level="15.4" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#prepare-for-spatiallibd"><i class="fa fa-check"></i><b>15.4</b> Prepare for spatialLIBD</a>
<ul>
<li class="chapter" data-level="15.4.1" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#basic-information"><i class="fa fa-check"></i><b>15.4.1</b> Basic information</a></li>
<li class="chapter" data-level="15.4.2" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#gene-annotation"><i class="fa fa-check"></i><b>15.4.2</b> Gene annotation</a></li>
<li class="chapter" data-level="15.4.3" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#extra-information-and-filtering"><i class="fa fa-check"></i><b>15.4.3</b> Extra information and filtering</a></li>
</ul></li>
<li class="chapter" data-level="15.5" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#explore-the-data"><i class="fa fa-check"></i><b>15.5</b> Explore the data</a></li>
<li class="chapter" data-level="15.6" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#sharing-your-website"><i class="fa fa-check"></i><b>15.6</b> Sharing your website</a></li>
<li class="chapter" data-level="15.7" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#wrapping-up"><i class="fa fa-check"></i><b>15.7</b> Wrapping up</a></li>
<li class="chapter" data-level="" data-path="spatiallibd-workflow.html"><a href="spatiallibd-workflow.html#r-session-information"><i class="fa fa-check"></i>R session information</a></li>
</ul></li>
<li class="part"><span><b>IV Appendix</b></span></li>
<li class="chapter" data-level="16" data-path="related-resources.html"><a href="related-resources.html"><i class="fa fa-check"></i><b>16</b> Related resources</a>
<ul>
<li class="chapter" data-level="16.1" data-path="related-resources.html"><a href="related-resources.html#data-preprocessing-procedures"><i class="fa fa-check"></i><b>16.1</b> Data preprocessing procedures</a></li>
<li class="chapter" data-level="16.2" data-path="related-resources.html"><a href="related-resources.html#resources-for-other-spatially-resolved-platforms"><i class="fa fa-check"></i><b>16.2</b> Resources for other spatially-resolved platforms</a></li>
<li class="chapter" data-level="16.3" data-path="related-resources.html"><a href="related-resources.html#data-structures"><i class="fa fa-check"></i><b>16.3</b> Data structures</a></li>
<li class="chapter" data-level="16.4" data-path="related-resources.html"><a href="related-resources.html#statistical-concepts"><i class="fa fa-check"></i><b>16.4</b> Statistical concepts</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="contributors-and-acknowledgments.html"><a href="contributors-and-acknowledgments.html"><i class="fa fa-check"></i><b>17</b> Contributors and acknowledgments</a>
<ul>
<li class="chapter" data-level="17.1" data-path="contributors-and-acknowledgments.html"><a href="contributors-and-acknowledgments.html#best-practices-for-spatial-transcriptomics-analysis-with-bioconductor"><i class="fa fa-check"></i><b>17.1</b> Best Practices for Spatial Transcriptomics Analysis with Bioconductor</a></li>
<li class="chapter" data-level="17.2" data-path="contributors-and-acknowledgments.html"><a href="contributors-and-acknowledgments.html#spatialexperiment"><i class="fa fa-check"></i><b>17.2</b> SpatialExperiment</a></li>
<li class="chapter" data-level="17.3" data-path="contributors-and-acknowledgments.html"><a href="contributors-and-acknowledgments.html#github-actions-workflow"><i class="fa fa-check"></i><b>17.3</b> GitHub Actions workflow</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>18</b> References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Best Practices for Spatial Transcriptomics Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="spatiallibd-workflow" class="section level1 hasAnchor" number="15">
<h1><span class="header-section-number">Chapter 15</span> spatialLIBD workflow<a href="spatiallibd-workflow.html#spatiallibd-workflow" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="overview-2" class="section level2 hasAnchor" number="15.1">
<h2><span class="header-section-number">15.1</span> Overview<a href="spatiallibd-workflow.html#overview-2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In the previous workflow, <a href="human-dlpfc-workflow.html#human-dlpfc-workflow">Human DLPFC workflow</a>, you practiced some of the basics with a portion of the postmortem human brain dataset <span class="citation">Maynard et al. (<a href="#ref-Maynard2021">2021</a>)</span>. The goal of this workflow is to learn what steps you need to carry out in order to create an interactive website to visualize this type of data. For this, we’ll use the <a href="http://bioconductor.org/packages/spatialLIBD"><code>spatialLIBD</code></a> Bioconductor package <span class="citation">Pardo et al. (<a href="#ref-Pardo2021">2021</a>)</span>.</p>
<div class="sourceCode" id="cb259"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb259-1"><a href="spatiallibd-workflow.html#cb259-1" tabindex="-1"></a><span class="co"># clear workspace from previous chapters</span></span>
<span id="cb259-2"><a href="spatiallibd-workflow.html#cb259-2" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>(<span class="at">all =</span> <span class="cn">TRUE</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-109"></span>
<img src="images/spatialLIBD_Figure1.png" alt="`spatialLIBD` overview. Source: [Pardo et al, _bioRxiv_, 2021](https://doi.org/10.1101/2021.04.29.440149)." width="100%" />
<p class="caption">
Figure 15.1: <code>spatialLIBD</code> overview. Source: <a href="https://doi.org/10.1101/2021.04.29.440149">Pardo et al, <em>bioRxiv</em>, 2021</a>.
</p>
</div>
</div>
<div id="spatiallibd" class="section level2 hasAnchor" number="15.2">
<h2><span class="header-section-number">15.2</span> spatialLIBD<a href="spatiallibd-workflow.html#spatiallibd" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="why-use-spatiallibd" class="section level3 hasAnchor" number="15.2.1">
<h3><span class="header-section-number">15.2.1</span> Why use spatialLIBD?<a href="spatiallibd-workflow.html#why-use-spatiallibd" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Before we dive into the R code, let’s first revisit why you might want to use <code>spatialLIBD</code>. This package has a function, <code>spatialLIBD::run_app(spe)</code>, which will create an interactive website using a <code>SpatialExperiment</code> object (<code>spe</code>). The interactive website it creates has several features that were initially designed for a specific dataset <span class="citation">Maynard et al. (<a href="#ref-Maynard2021">2021</a>)</span> and later made flexible for any dataset <span class="citation">Pardo et al. (<a href="#ref-Pardo2021">2021</a>)</span>. These features include panels to visualize spots from the Visium platform by 10x Genomics:</p>
<ul>
<li>for one tissue section at a time, either with interactive or static versions</li>
<li>multiple tissue sections at a time, either interactively or statically</li>
</ul>
<p>Both options work with continuous and discrete variables such as the gene expression and clusters, respectively. The interactive version for discrete variables such as clusters is useful if you want to manually annotate Visium spots, as it was done in the initial project <span class="citation">Maynard et al. (<a href="#ref-Maynard2021">2021</a>)</span>. <code>spatialLIBD</code> allows users to download the annotated spots and resume your spot annotation work later.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-110"></span>
<img src="images/spatialLIBD_interactive_cluster.png" alt="Screenshot of the 'clusters (interactive)' section of the 'spot-level data' panel created with the full spatialLIBD dataset. The website was created with `spatialLIBD::run_app(spatialLIBD::fetch_data('spe'))` version 1.4.0 and then using the lasso selection, we selected a set of spots in the UMAP interactive plot colored by the estimated number of cells per spot (`cell_count`) on the bottom left, which automatically updated the other three plots." width="100%" />
<p class="caption">
Figure 15.2: Screenshot of the ‘clusters (interactive)’ section of the ‘spot-level data’ panel created with the full spatialLIBD dataset. The website was created with <code>spatialLIBD::run_app(spatialLIBD::fetch_data('spe'))</code> version 1.4.0 and then using the lasso selection, we selected a set of spots in the UMAP interactive plot colored by the estimated number of cells per spot (<code>cell_count</code>) on the bottom left, which automatically updated the other three plots.
</p>
</div>
<p>Visualizing genes or clusters across multiple tissue sections can be quite useful. For example, here we show the expression levels of <em>PCP4</em> across two sets of spatially adjacent replicates. <em>PCP4</em> is a marker gene for layer 5 in the grey matter of the dorsolateral prefrontal cortex (DLPFC) in the human brain. Spatially adjacent replicates are about 10 microns apart from each other and visualizations like the one below help assess the technical variability in the Visium technology.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-111"></span>
<img src="images/spatialLIBD_gene_grid.png" alt="Screenshot of the 'gene grid (static)' section of the 'spot-level data' panel created with the full spatialLIBD dataset. The website was created with `spatialLIBD::run_app(spatialLIBD::fetch_data('spe'))` version 1.4.0, selecting the _PCP4_ gene, selecting the _paper_ gene color scale, changing the number of rows and columns in the grid 2, selecting two pairs of spatially adjacent replicate samples (151507, 151508, 151673, and 151674), and clicking on the _upgrade grid plot_ button. Note that the default _viridis_ gene color scale is color-blind friendly." width="100%" />
<p class="caption">
Figure 15.3: Screenshot of the ‘gene grid (static)’ section of the ‘spot-level data’ panel created with the full spatialLIBD dataset. The website was created with <code>spatialLIBD::run_app(spatialLIBD::fetch_data('spe'))</code> version 1.4.0, selecting the <em>PCP4</em> gene, selecting the <em>paper</em> gene color scale, changing the number of rows and columns in the grid 2, selecting two pairs of spatially adjacent replicate samples (151507, 151508, 151673, and 151674), and clicking on the <em>upgrade grid plot</em> button. Note that the default <em>viridis</em> gene color scale is color-blind friendly.
</p>
</div>
<p>You can try out a <code>spatialLIBD</code>-powered website yourself by opening <a href="http://spatial.libd.org/spatialLIBD">it on your browser</a> <a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>.</p>
</div>
<div id="want-to-learn-more-about-spatiallibd" class="section level3 hasAnchor" number="15.2.2">
<h3><span class="header-section-number">15.2.2</span> Want to learn more about spatialLIBD?<a href="spatiallibd-workflow.html#want-to-learn-more-about-spatiallibd" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>If you are interested in learning more about <code>spatialLIBD</code>, please check the <a href="http://bioconductor.org/packages/spatialLIBD"><code>spatialLIBD</code> Bioconductor landing page</a> or the <a href="http://lieberinstitute.github.io/spatialLIBD/"><code>pkgdown</code> documentation website</a>. In particular, we have two vignettes documents:</p>
<ul>
<li><a href="http://research.libd.org/spatialLIBD/articles/spatialLIBD.html">Introduction to spatialLIBD</a></li>
<li><a href="http://research.libd.org/spatialLIBD/articles/TenX_data_download.html">Using spatialLIBD with 10x Genomics public datasets</a></li>
</ul>
<p>You can also read more about <code>spatialLIBD</code> in the associated publication.</p>
<div class="sourceCode" id="cb260"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb260-1"><a href="spatiallibd-workflow.html#cb260-1" tabindex="-1"></a><span class="fu">citation</span>(<span class="st">&quot;spatialLIBD&quot;</span>)[<span class="dv">1</span>]</span></code></pre></div>
<pre><code>## Pardo B, Spangler A, Weber LM, Hicks SC, Jaffe AE, Martinowich K,
## Maynard KR, Collado-Torres L (2022). &quot;spatialLIBD: an R/Bioconductor
## package to visualize spatially-resolved transcriptomics data.&quot; _BMC
## Genomics_. doi:10.1186/s12864-022-08601-w
## &lt;https://doi.org/10.1186/s12864-022-08601-w&gt;,
## &lt;https://doi.org/10.1186/s12864-022-08601-w&gt;.
## 
## A BibTeX entry for LaTeX users is
## 
##   @Article{,
##     title = {spatialLIBD: an R/Bioconductor package to visualize spatially-resolved transcriptomics data},
##     author = {Brenda Pardo and Abby Spangler and Lukas M. Weber and Stephanie C. Hicks and Andrew E. Jaffe and Keri Martinowich and Kristen R. Maynard and Leonardo Collado-Torres},
##     year = {2022},
##     journal = {BMC Genomics},
##     doi = {10.1186/s12864-022-08601-w},
##     url = {https://doi.org/10.1186/s12864-022-08601-w},
##   }</code></pre>
<div id="recordings" class="section level4 hasAnchor" number="15.2.2.1">
<h4><span class="header-section-number">15.2.2.1</span> Recordings<a href="spatiallibd-workflow.html#recordings" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p>If you prefer to watch recordings of presentations related to the dataset <span class="citation">Maynard et al. (<a href="#ref-Maynard2021">2021</a>)</span> or <code>spatialLIBD</code> <span class="citation">Pardo et al. (<a href="#ref-Pardo2021">2021</a>)</span>, check the following <a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>:</p>
<iframe class="speakerdeck-iframe" frameborder="0" src="https://speakerdeck.com/player/dde92cd6dfc04f9589770e074915658f" title="BioTuring_spatialLIBD" allowfullscreen="true" style="border: 0px; background: padding-box padding-box rgba(0, 0, 0, 0.1); margin: 0px; padding: 0px; border-radius: 6px; box-shadow: rgba(0, 0, 0, 0.2) 0px 5px 40px; width: 100%; height: auto; aspect-ratio: 560 / 420;" data-ratio="1.3333333333333333">
</iframe>
<p>These slides were part of our 2021-04-27 webinar for BioTuring that you can watch on YouTube:</p>
<iframe width="560" height="315" src="https://www.youtube.com/embed/S8884Kde-1U" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen>
</iframe>
<p>A recording of an earlier version of this talk is also available on YouTube.</p>
<iframe width="560" height="315" src="https://www.youtube.com/embed/aD2JU-vUv54" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen>
</iframe>
<p>You might also be interested in this video demonstration of <code>spatialLIBD</code> for the <a href="http://research.libd.org/rstatsclub/">LIBD rstats club</a>.</p>
<iframe width="560" height="315" src="https://www.youtube.com/embed/LZ2kvCiRVdM" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen>
</iframe>
</div>
</div>
</div>
<div id="code-prerequisites" class="section level2 hasAnchor" number="15.3">
<h2><span class="header-section-number">15.3</span> Code prerequisites<a href="spatiallibd-workflow.html#code-prerequisites" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Ok, let’s get started! First we need to re-create the <code>spe</code> object from the <a href="human-dlpfc-workflow.html#human-dlpfc-workflow">Human DLPFC workflow</a> in the previous chapter. That chapter included code for visualizing results along the way, which we’ll skip here. Thus, we’ll use the following packages in addition to <code>spatialLIBD</code>:</p>
<ul>
<li><a href="http://bioconductor.org/packages/SpatialExperiment"><code>SpatialExperiment</code></a>: for storing our data in a common object</li>
<li><a href="http://bioconductor.org/packages/STexampleData"><code>STexampleData</code></a>: for accessing the example data</li>
<li><a href="http://bioconductor.org/packages/scater"><code>scater</code></a>: for quality control checks</li>
<li><a href="http://bioconductor.org/packages/scran"><code>scran</code></a>: for normalization, dimension reduction, and clustering</li>
<li><a href="https://cran.r-project.org/package=igraph"><code>igraph</code></a>: for clustering algorithms</li>
<li><a href="http://bioconductor.org/packages/BiocFileCache"><code>BiocFileCache</code></a>: for downloading and storing data</li>
<li><a href="http://bioconductor.org/packages/rtracklayer"><code>rtracklayer</code></a>: for importing gene annotation files</li>
<li><a href="https://cran.r-project.org/package=igraph"><code>lobstr</code></a>: for checking object memory usage</li>
</ul>
<p>You’ll need to have the R version compatible with <code>bioc-release</code> <a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a> installed in your computer, as <a href="http://bioconductor.org/install/">documented by Bioconductor</a>. Alternatively, you can use the <a href="http://bioconductor.org/help/docker/">Bioconductor docker images</a>. Next, if you haven’t installed these packages, please do so with the following R code.</p>
<div class="sourceCode" id="cb262"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb262-1"><a href="spatiallibd-workflow.html#cb262-1" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) {</span>
<span id="cb262-2"><a href="spatiallibd-workflow.html#cb262-2" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb262-3"><a href="spatiallibd-workflow.html#cb262-3" tabindex="-1"></a>}</span>
<span id="cb262-4"><a href="spatiallibd-workflow.html#cb262-4" tabindex="-1"></a></span>
<span id="cb262-5"><a href="spatiallibd-workflow.html#cb262-5" tabindex="-1"></a><span class="do">## Check that you have a valid installation</span></span>
<span id="cb262-6"><a href="spatiallibd-workflow.html#cb262-6" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">valid</span>()</span>
<span id="cb262-7"><a href="spatiallibd-workflow.html#cb262-7" tabindex="-1"></a></span>
<span id="cb262-8"><a href="spatiallibd-workflow.html#cb262-8" tabindex="-1"></a><span class="do">## Install the required R packages for this workflow</span></span>
<span id="cb262-9"><a href="spatiallibd-workflow.html#cb262-9" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="fu">c</span>(</span>
<span id="cb262-10"><a href="spatiallibd-workflow.html#cb262-10" tabindex="-1"></a>    <span class="st">&quot;SpatialExperiment&quot;</span>,</span>
<span id="cb262-11"><a href="spatiallibd-workflow.html#cb262-11" tabindex="-1"></a>    <span class="st">&quot;STexampleData&quot;</span>,</span>
<span id="cb262-12"><a href="spatiallibd-workflow.html#cb262-12" tabindex="-1"></a>    <span class="st">&quot;scater&quot;</span>,</span>
<span id="cb262-13"><a href="spatiallibd-workflow.html#cb262-13" tabindex="-1"></a>    <span class="st">&quot;scran&quot;</span>,</span>
<span id="cb262-14"><a href="spatiallibd-workflow.html#cb262-14" tabindex="-1"></a>    <span class="st">&quot;igraph&quot;</span>,</span>
<span id="cb262-15"><a href="spatiallibd-workflow.html#cb262-15" tabindex="-1"></a>    <span class="st">&quot;BiocFileCache&quot;</span>,</span>
<span id="cb262-16"><a href="spatiallibd-workflow.html#cb262-16" tabindex="-1"></a>    <span class="st">&quot;rtracklayer&quot;</span>,</span>
<span id="cb262-17"><a href="spatiallibd-workflow.html#cb262-17" tabindex="-1"></a>    <span class="st">&quot;lobstr&quot;</span>,</span>
<span id="cb262-18"><a href="spatiallibd-workflow.html#cb262-18" tabindex="-1"></a>    <span class="st">&quot;spatialLIBD&quot;</span></span>
<span id="cb262-19"><a href="spatiallibd-workflow.html#cb262-19" tabindex="-1"></a>))</span></code></pre></div>
<p>We can now run the following R code to re-make the <code>spe</code> object from the <a href="human-dlpfc-workflow.html#human-dlpfc-workflow">Human DLPFC workflow</a>. This will take a bit of time.</p>
<div class="sourceCode" id="cb263"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb263-1"><a href="spatiallibd-workflow.html#cb263-1" tabindex="-1"></a><span class="do">## Load packages required for the</span></span>
<span id="cb263-2"><a href="spatiallibd-workflow.html#cb263-2" tabindex="-1"></a><span class="do">## &quot;Visium human DLPFC workflow&quot;</span></span>
<span id="cb263-3"><a href="spatiallibd-workflow.html#cb263-3" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;SpatialExperiment&quot;</span>)</span>
<span id="cb263-4"><a href="spatiallibd-workflow.html#cb263-4" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;STexampleData&quot;</span>)</span>
<span id="cb263-5"><a href="spatiallibd-workflow.html#cb263-5" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;scater&quot;</span>)</span>
<span id="cb263-6"><a href="spatiallibd-workflow.html#cb263-6" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;scran&quot;</span>)</span>
<span id="cb263-7"><a href="spatiallibd-workflow.html#cb263-7" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;igraph&quot;</span>)</span>
<span id="cb263-8"><a href="spatiallibd-workflow.html#cb263-8" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;BiocFileCache&quot;</span>)</span>
<span id="cb263-9"><a href="spatiallibd-workflow.html#cb263-9" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;rtracklayer&quot;</span>)</span>
<span id="cb263-10"><a href="spatiallibd-workflow.html#cb263-10" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;lobstr&quot;</span>)</span>
<span id="cb263-11"><a href="spatiallibd-workflow.html#cb263-11" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;spatialLIBD&quot;</span>)</span>
<span id="cb263-12"><a href="spatiallibd-workflow.html#cb263-12" tabindex="-1"></a></span>
<span id="cb263-13"><a href="spatiallibd-workflow.html#cb263-13" tabindex="-1"></a><span class="do">## Start tracking time</span></span>
<span id="cb263-14"><a href="spatiallibd-workflow.html#cb263-14" tabindex="-1"></a>time_start <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span>
<span id="cb263-15"><a href="spatiallibd-workflow.html#cb263-15" tabindex="-1"></a></span>
<span id="cb263-16"><a href="spatiallibd-workflow.html#cb263-16" tabindex="-1"></a><span class="co"># load object</span></span>
<span id="cb263-17"><a href="spatiallibd-workflow.html#cb263-17" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">Visium_humanDLPFC</span>()</span>
<span id="cb263-18"><a href="spatiallibd-workflow.html#cb263-18" tabindex="-1"></a></span>
<span id="cb263-19"><a href="spatiallibd-workflow.html#cb263-19" tabindex="-1"></a><span class="co"># subset to keep only spots over tissue</span></span>
<span id="cb263-20"><a href="spatiallibd-workflow.html#cb263-20" tabindex="-1"></a>spe <span class="ot">&lt;-</span> spe[, <span class="fu">colData</span>(spe)<span class="sc">$</span>in_tissue <span class="sc">==</span> <span class="dv">1</span>]</span>
<span id="cb263-21"><a href="spatiallibd-workflow.html#cb263-21" tabindex="-1"></a></span>
<span id="cb263-22"><a href="spatiallibd-workflow.html#cb263-22" tabindex="-1"></a><span class="co"># identify mitochondrial genes</span></span>
<span id="cb263-23"><a href="spatiallibd-workflow.html#cb263-23" tabindex="-1"></a>is_mito <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">&quot;(^MT-)|(^mt-)&quot;</span>, <span class="fu">rowData</span>(spe)<span class="sc">$</span>gene_name)</span>
<span id="cb263-24"><a href="spatiallibd-workflow.html#cb263-24" tabindex="-1"></a></span>
<span id="cb263-25"><a href="spatiallibd-workflow.html#cb263-25" tabindex="-1"></a><span class="co"># calculate per-spot QC metrics and store in colData</span></span>
<span id="cb263-26"><a href="spatiallibd-workflow.html#cb263-26" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">addPerCellQC</span>(spe, <span class="at">subsets =</span> <span class="fu">list</span>(<span class="at">mito =</span> is_mito))</span>
<span id="cb263-27"><a href="spatiallibd-workflow.html#cb263-27" tabindex="-1"></a></span>
<span id="cb263-28"><a href="spatiallibd-workflow.html#cb263-28" tabindex="-1"></a><span class="co"># select QC thresholds</span></span>
<span id="cb263-29"><a href="spatiallibd-workflow.html#cb263-29" tabindex="-1"></a>qc_lib_size <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe)<span class="sc">$</span>sum <span class="sc">&lt;</span> <span class="dv">600</span></span>
<span id="cb263-30"><a href="spatiallibd-workflow.html#cb263-30" tabindex="-1"></a>qc_detected <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe)<span class="sc">$</span>detected <span class="sc">&lt;</span> <span class="dv">400</span></span>
<span id="cb263-31"><a href="spatiallibd-workflow.html#cb263-31" tabindex="-1"></a>qc_mito <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe)<span class="sc">$</span>subsets_mito_percent <span class="sc">&gt;</span> <span class="dv">28</span></span>
<span id="cb263-32"><a href="spatiallibd-workflow.html#cb263-32" tabindex="-1"></a>qc_cell_count <span class="ot">&lt;-</span> <span class="fu">colData</span>(spe)<span class="sc">$</span>cell_count <span class="sc">&gt;</span> <span class="dv">10</span></span>
<span id="cb263-33"><a href="spatiallibd-workflow.html#cb263-33" tabindex="-1"></a></span>
<span id="cb263-34"><a href="spatiallibd-workflow.html#cb263-34" tabindex="-1"></a><span class="co"># combined set of discarded spots</span></span>
<span id="cb263-35"><a href="spatiallibd-workflow.html#cb263-35" tabindex="-1"></a>discard <span class="ot">&lt;-</span> qc_lib_size <span class="sc">|</span> qc_detected <span class="sc">|</span> qc_mito <span class="sc">|</span> qc_cell_count</span>
<span id="cb263-36"><a href="spatiallibd-workflow.html#cb263-36" tabindex="-1"></a></span>
<span id="cb263-37"><a href="spatiallibd-workflow.html#cb263-37" tabindex="-1"></a><span class="co"># store in object</span></span>
<span id="cb263-38"><a href="spatiallibd-workflow.html#cb263-38" tabindex="-1"></a><span class="fu">colData</span>(spe)<span class="sc">$</span>discard <span class="ot">&lt;-</span> discard</span>
<span id="cb263-39"><a href="spatiallibd-workflow.html#cb263-39" tabindex="-1"></a></span>
<span id="cb263-40"><a href="spatiallibd-workflow.html#cb263-40" tabindex="-1"></a><span class="co"># filter low-quality spots</span></span>
<span id="cb263-41"><a href="spatiallibd-workflow.html#cb263-41" tabindex="-1"></a>spe <span class="ot">&lt;-</span> spe[, <span class="sc">!</span><span class="fu">colData</span>(spe)<span class="sc">$</span>discard]</span>
<span id="cb263-42"><a href="spatiallibd-workflow.html#cb263-42" tabindex="-1"></a></span>
<span id="cb263-43"><a href="spatiallibd-workflow.html#cb263-43" tabindex="-1"></a><span class="co"># calculate logcounts using library size factors</span></span>
<span id="cb263-44"><a href="spatiallibd-workflow.html#cb263-44" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(spe)</span>
<span id="cb263-45"><a href="spatiallibd-workflow.html#cb263-45" tabindex="-1"></a></span>
<span id="cb263-46"><a href="spatiallibd-workflow.html#cb263-46" tabindex="-1"></a><span class="co"># remove mitochondrial genes</span></span>
<span id="cb263-47"><a href="spatiallibd-workflow.html#cb263-47" tabindex="-1"></a>spe <span class="ot">&lt;-</span> spe[<span class="sc">!</span>is_mito, ]</span>
<span id="cb263-48"><a href="spatiallibd-workflow.html#cb263-48" tabindex="-1"></a></span>
<span id="cb263-49"><a href="spatiallibd-workflow.html#cb263-49" tabindex="-1"></a><span class="co"># fit mean-variance relationship</span></span>
<span id="cb263-50"><a href="spatiallibd-workflow.html#cb263-50" tabindex="-1"></a>dec <span class="ot">&lt;-</span> <span class="fu">modelGeneVar</span>(spe)</span>
<span id="cb263-51"><a href="spatiallibd-workflow.html#cb263-51" tabindex="-1"></a></span>
<span id="cb263-52"><a href="spatiallibd-workflow.html#cb263-52" tabindex="-1"></a><span class="co"># select top HVGs</span></span>
<span id="cb263-53"><a href="spatiallibd-workflow.html#cb263-53" tabindex="-1"></a>top_hvgs <span class="ot">&lt;-</span> <span class="fu">getTopHVGs</span>(dec, <span class="at">prop =</span> <span class="fl">0.1</span>)</span>
<span id="cb263-54"><a href="spatiallibd-workflow.html#cb263-54" tabindex="-1"></a></span>
<span id="cb263-55"><a href="spatiallibd-workflow.html#cb263-55" tabindex="-1"></a><span class="co"># compute PCA</span></span>
<span id="cb263-56"><a href="spatiallibd-workflow.html#cb263-56" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb263-57"><a href="spatiallibd-workflow.html#cb263-57" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(spe, <span class="at">subset_row =</span> top_hvgs)</span>
<span id="cb263-58"><a href="spatiallibd-workflow.html#cb263-58" tabindex="-1"></a></span>
<span id="cb263-59"><a href="spatiallibd-workflow.html#cb263-59" tabindex="-1"></a><span class="co"># compute UMAP on top 50 PCs</span></span>
<span id="cb263-60"><a href="spatiallibd-workflow.html#cb263-60" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb263-61"><a href="spatiallibd-workflow.html#cb263-61" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">runUMAP</span>(spe, <span class="at">dimred =</span> <span class="st">&quot;PCA&quot;</span>)</span>
<span id="cb263-62"><a href="spatiallibd-workflow.html#cb263-62" tabindex="-1"></a></span>
<span id="cb263-63"><a href="spatiallibd-workflow.html#cb263-63" tabindex="-1"></a><span class="co"># update column names for easier plotting</span></span>
<span id="cb263-64"><a href="spatiallibd-workflow.html#cb263-64" tabindex="-1"></a><span class="fu">colnames</span>(<span class="fu">reducedDim</span>(spe, <span class="st">&quot;UMAP&quot;</span>)) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;UMAP&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)</span>
<span id="cb263-65"><a href="spatiallibd-workflow.html#cb263-65" tabindex="-1"></a></span>
<span id="cb263-66"><a href="spatiallibd-workflow.html#cb263-66" tabindex="-1"></a><span class="co"># graph-based clustering</span></span>
<span id="cb263-67"><a href="spatiallibd-workflow.html#cb263-67" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb263-68"><a href="spatiallibd-workflow.html#cb263-68" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb263-69"><a href="spatiallibd-workflow.html#cb263-69" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">buildSNNGraph</span>(spe, <span class="at">k =</span> k, <span class="at">use.dimred =</span> <span class="st">&quot;PCA&quot;</span>)</span>
<span id="cb263-70"><a href="spatiallibd-workflow.html#cb263-70" tabindex="-1"></a>g_walk <span class="ot">&lt;-</span> igraph<span class="sc">::</span><span class="fu">cluster_walktrap</span>(g)</span>
<span id="cb263-71"><a href="spatiallibd-workflow.html#cb263-71" tabindex="-1"></a>clus <span class="ot">&lt;-</span> g_walk<span class="sc">$</span>membership</span>
<span id="cb263-72"><a href="spatiallibd-workflow.html#cb263-72" tabindex="-1"></a></span>
<span id="cb263-73"><a href="spatiallibd-workflow.html#cb263-73" tabindex="-1"></a><span class="co"># store cluster labels in column &#39;label&#39; in colData</span></span>
<span id="cb263-74"><a href="spatiallibd-workflow.html#cb263-74" tabindex="-1"></a><span class="fu">colLabels</span>(spe) <span class="ot">&lt;-</span> <span class="fu">factor</span>(clus)</span>
<span id="cb263-75"><a href="spatiallibd-workflow.html#cb263-75" tabindex="-1"></a></span>
<span id="cb263-76"><a href="spatiallibd-workflow.html#cb263-76" tabindex="-1"></a><span class="co"># set gene names as row names for easier plotting</span></span>
<span id="cb263-77"><a href="spatiallibd-workflow.html#cb263-77" tabindex="-1"></a><span class="fu">rownames</span>(spe) <span class="ot">&lt;-</span> <span class="fu">rowData</span>(spe)<span class="sc">$</span>gene_name</span>
<span id="cb263-78"><a href="spatiallibd-workflow.html#cb263-78" tabindex="-1"></a></span>
<span id="cb263-79"><a href="spatiallibd-workflow.html#cb263-79" tabindex="-1"></a><span class="co"># test for marker genes</span></span>
<span id="cb263-80"><a href="spatiallibd-workflow.html#cb263-80" tabindex="-1"></a>markers <span class="ot">&lt;-</span> <span class="fu">findMarkers</span>(spe, <span class="at">test =</span> <span class="st">&quot;binom&quot;</span>, <span class="at">direction =</span> <span class="st">&quot;up&quot;</span>)</span>
<span id="cb263-81"><a href="spatiallibd-workflow.html#cb263-81" tabindex="-1"></a></span>
<span id="cb263-82"><a href="spatiallibd-workflow.html#cb263-82" tabindex="-1"></a><span class="do">## Find the interesting markers for each cluster</span></span>
<span id="cb263-83"><a href="spatiallibd-workflow.html#cb263-83" tabindex="-1"></a>interesting <span class="ot">&lt;-</span> <span class="fu">sapply</span>(markers, <span class="cf">function</span>(x) x<span class="sc">$</span>Top <span class="sc">&lt;=</span> <span class="dv">5</span>)</span>
<span id="cb263-84"><a href="spatiallibd-workflow.html#cb263-84" tabindex="-1"></a><span class="fu">colnames</span>(interesting) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;gene_interest_&quot;</span>, <span class="fu">seq_len</span>(<span class="fu">length</span>(markers)))</span>
<span id="cb263-85"><a href="spatiallibd-workflow.html#cb263-85" tabindex="-1"></a><span class="fu">rowData</span>(spe) <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">rowData</span>(spe), interesting)</span>
<span id="cb263-86"><a href="spatiallibd-workflow.html#cb263-86" tabindex="-1"></a></span>
<span id="cb263-87"><a href="spatiallibd-workflow.html#cb263-87" tabindex="-1"></a><span class="do">## How long this code took to run</span></span>
<span id="cb263-88"><a href="spatiallibd-workflow.html#cb263-88" tabindex="-1"></a>time_prereqs <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span>
<span id="cb263-89"><a href="spatiallibd-workflow.html#cb263-89" tabindex="-1"></a>time_prereqs <span class="sc">-</span> time_start</span></code></pre></div>
<pre><code>## Time difference of 1.423712 mins</code></pre>
</div>
<div id="prepare-for-spatiallibd" class="section level2 hasAnchor" number="15.4">
<h2><span class="header-section-number">15.4</span> Prepare for spatialLIBD<a href="spatiallibd-workflow.html#prepare-for-spatiallibd" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Now that we have a <code>spe</code> object with quality control information, dimension reduction results, clustering data, among other things, we can proceed to visualize the object using <code>spatialLIBD</code>. Well, almost. First we need to modify the <code>spe</code> object, similar to steps we need to carry out when <a href="http://research.libd.org/spatialLIBD/articles/TenX_data_download.html#modify-spe-for-spatiallibd-1">using spatialLIBD with 10x Genomics public datasets</a></p>
<div id="basic-information" class="section level3 hasAnchor" number="15.4.1">
<h3><span class="header-section-number">15.4.1</span> Basic information<a href="spatiallibd-workflow.html#basic-information" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb265"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb265-1"><a href="spatiallibd-workflow.html#cb265-1" tabindex="-1"></a><span class="do">## Add some information used by spatialLIBD</span></span>
<span id="cb265-2"><a href="spatiallibd-workflow.html#cb265-2" tabindex="-1"></a>spe<span class="sc">$</span>key <span class="ot">&lt;-</span> <span class="fu">paste0</span>(spe<span class="sc">$</span>sample_id, <span class="st">&quot;_&quot;</span>, <span class="fu">colnames</span>(spe))</span>
<span id="cb265-3"><a href="spatiallibd-workflow.html#cb265-3" tabindex="-1"></a>spe<span class="sc">$</span>sum_umi <span class="ot">&lt;-</span> <span class="fu">colSums</span>(<span class="fu">counts</span>(spe))</span>
<span id="cb265-4"><a href="spatiallibd-workflow.html#cb265-4" tabindex="-1"></a>spe<span class="sc">$</span>sum_gene <span class="ot">&lt;-</span> <span class="fu">colSums</span>(<span class="fu">counts</span>(spe) <span class="sc">&gt;</span> <span class="dv">0</span>)</span></code></pre></div>
</div>
<div id="gene-annotation" class="section level3 hasAnchor" number="15.4.2">
<h3><span class="header-section-number">15.4.2</span> Gene annotation<a href="spatiallibd-workflow.html#gene-annotation" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Since the gene information is missing, we’ll <a href="http://research.libd.org/spatialLIBD/articles/TenX_data_download.html#add-gene-annotation-information-1">add the gene annotation data from Gencode</a> although you would ideally add this information from the same gene annotation you used for running <code>spaceranger</code>.</p>
<div class="sourceCode" id="cb266"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb266-1"><a href="spatiallibd-workflow.html#cb266-1" tabindex="-1"></a><span class="do">## Download the Gencode v32 GTF file and cache it</span></span>
<span id="cb266-2"><a href="spatiallibd-workflow.html#cb266-2" tabindex="-1"></a>bfc <span class="ot">&lt;-</span> BiocFileCache<span class="sc">::</span><span class="fu">BiocFileCache</span>()</span>
<span id="cb266-3"><a href="spatiallibd-workflow.html#cb266-3" tabindex="-1"></a>gtf_cache <span class="ot">&lt;-</span> BiocFileCache<span class="sc">::</span><span class="fu">bfcrpath</span>(</span>
<span id="cb266-4"><a href="spatiallibd-workflow.html#cb266-4" tabindex="-1"></a>    bfc,</span>
<span id="cb266-5"><a href="spatiallibd-workflow.html#cb266-5" tabindex="-1"></a>    <span class="fu">paste0</span>(</span>
<span id="cb266-6"><a href="spatiallibd-workflow.html#cb266-6" tabindex="-1"></a>        <span class="st">&quot;ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/&quot;</span>,</span>
<span id="cb266-7"><a href="spatiallibd-workflow.html#cb266-7" tabindex="-1"></a>        <span class="st">&quot;release_32/gencode.v32.annotation.gtf.gz&quot;</span></span>
<span id="cb266-8"><a href="spatiallibd-workflow.html#cb266-8" tabindex="-1"></a>    )</span>
<span id="cb266-9"><a href="spatiallibd-workflow.html#cb266-9" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## adding rname &#39;ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/gencode.v32.annotation.gtf.gz&#39;</code></pre>
<div class="sourceCode" id="cb268"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb268-1"><a href="spatiallibd-workflow.html#cb268-1" tabindex="-1"></a><span class="do">## Show the GTF cache location</span></span>
<span id="cb268-2"><a href="spatiallibd-workflow.html#cb268-2" tabindex="-1"></a>gtf_cache</span></code></pre></div>
<pre><code>##                                                                            BFC1 
## &quot;/github/home/.cache/R/BiocFileCache/355276e4844_gencode.v32.annotation.gtf.gz&quot;</code></pre>
<div class="sourceCode" id="cb270"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb270-1"><a href="spatiallibd-workflow.html#cb270-1" tabindex="-1"></a><span class="do">## Import into R (takes ~1 min)</span></span>
<span id="cb270-2"><a href="spatiallibd-workflow.html#cb270-2" tabindex="-1"></a>gtf <span class="ot">&lt;-</span> rtracklayer<span class="sc">::</span><span class="fu">import</span>(gtf_cache)</span>
<span id="cb270-3"><a href="spatiallibd-workflow.html#cb270-3" tabindex="-1"></a></span>
<span id="cb270-4"><a href="spatiallibd-workflow.html#cb270-4" tabindex="-1"></a><span class="do">## Subset to genes only</span></span>
<span id="cb270-5"><a href="spatiallibd-workflow.html#cb270-5" tabindex="-1"></a>gtf <span class="ot">&lt;-</span> gtf[gtf<span class="sc">$</span>type <span class="sc">==</span> <span class="st">&quot;gene&quot;</span>]</span>
<span id="cb270-6"><a href="spatiallibd-workflow.html#cb270-6" tabindex="-1"></a></span>
<span id="cb270-7"><a href="spatiallibd-workflow.html#cb270-7" tabindex="-1"></a><span class="do">## Remove the .x part of the gene IDs</span></span>
<span id="cb270-8"><a href="spatiallibd-workflow.html#cb270-8" tabindex="-1"></a>gtf<span class="sc">$</span>gene_id <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">&quot;</span><span class="sc">\\</span><span class="st">..*&quot;</span>, <span class="st">&quot;&quot;</span>, gtf<span class="sc">$</span>gene_id)</span>
<span id="cb270-9"><a href="spatiallibd-workflow.html#cb270-9" tabindex="-1"></a></span>
<span id="cb270-10"><a href="spatiallibd-workflow.html#cb270-10" tabindex="-1"></a><span class="do">## Set the names to be the gene IDs</span></span>
<span id="cb270-11"><a href="spatiallibd-workflow.html#cb270-11" tabindex="-1"></a><span class="fu">names</span>(gtf) <span class="ot">&lt;-</span> gtf<span class="sc">$</span>gene_id</span>
<span id="cb270-12"><a href="spatiallibd-workflow.html#cb270-12" tabindex="-1"></a></span>
<span id="cb270-13"><a href="spatiallibd-workflow.html#cb270-13" tabindex="-1"></a><span class="do">## Match the genes</span></span>
<span id="cb270-14"><a href="spatiallibd-workflow.html#cb270-14" tabindex="-1"></a>match_genes <span class="ot">&lt;-</span> <span class="fu">match</span>(<span class="fu">rowData</span>(spe)<span class="sc">$</span>gene_id, gtf<span class="sc">$</span>gene_id)</span>
<span id="cb270-15"><a href="spatiallibd-workflow.html#cb270-15" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">is.na</span>(match_genes))</span></code></pre></div>
<pre><code>## 
## FALSE  TRUE 
## 33267   258</code></pre>
<div class="sourceCode" id="cb272"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb272-1"><a href="spatiallibd-workflow.html#cb272-1" tabindex="-1"></a><span class="do">## Drop the few genes for which we don&#39;t have information</span></span>
<span id="cb272-2"><a href="spatiallibd-workflow.html#cb272-2" tabindex="-1"></a>spe <span class="ot">&lt;-</span> spe[<span class="sc">!</span><span class="fu">is.na</span>(match_genes), ]</span>
<span id="cb272-3"><a href="spatiallibd-workflow.html#cb272-3" tabindex="-1"></a>match_genes <span class="ot">&lt;-</span> match_genes[<span class="sc">!</span><span class="fu">is.na</span>(match_genes)]</span>
<span id="cb272-4"><a href="spatiallibd-workflow.html#cb272-4" tabindex="-1"></a></span>
<span id="cb272-5"><a href="spatiallibd-workflow.html#cb272-5" tabindex="-1"></a><span class="do">## Keep only some columns from the gtf</span></span>
<span id="cb272-6"><a href="spatiallibd-workflow.html#cb272-6" tabindex="-1"></a><span class="fu">mcols</span>(gtf) <span class="ot">&lt;-</span> <span class="fu">mcols</span>(gtf)[, <span class="fu">c</span>(<span class="st">&quot;source&quot;</span>, <span class="st">&quot;type&quot;</span>, <span class="st">&quot;gene_id&quot;</span>, <span class="st">&quot;gene_name&quot;</span>, <span class="st">&quot;gene_type&quot;</span>)]</span>
<span id="cb272-7"><a href="spatiallibd-workflow.html#cb272-7" tabindex="-1"></a></span>
<span id="cb272-8"><a href="spatiallibd-workflow.html#cb272-8" tabindex="-1"></a><span class="do">## Save the &quot;interest&quot;ing columns from our original spe object</span></span>
<span id="cb272-9"><a href="spatiallibd-workflow.html#cb272-9" tabindex="-1"></a>interesting <span class="ot">&lt;-</span> <span class="fu">rowData</span>(spe)[, <span class="fu">grepl</span>(<span class="st">&quot;interest&quot;</span>, <span class="fu">colnames</span>(<span class="fu">rowData</span>(spe)))]</span>
<span id="cb272-10"><a href="spatiallibd-workflow.html#cb272-10" tabindex="-1"></a></span>
<span id="cb272-11"><a href="spatiallibd-workflow.html#cb272-11" tabindex="-1"></a><span class="do">## Add the gene info to our SPE object</span></span>
<span id="cb272-12"><a href="spatiallibd-workflow.html#cb272-12" tabindex="-1"></a><span class="fu">rowRanges</span>(spe) <span class="ot">&lt;-</span> gtf[match_genes]</span>
<span id="cb272-13"><a href="spatiallibd-workflow.html#cb272-13" tabindex="-1"></a></span>
<span id="cb272-14"><a href="spatiallibd-workflow.html#cb272-14" tabindex="-1"></a><span class="do">## Add back the &quot;interest&quot; coolumns</span></span>
<span id="cb272-15"><a href="spatiallibd-workflow.html#cb272-15" tabindex="-1"></a><span class="fu">rowData</span>(spe) <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">rowData</span>(spe), interesting)</span>
<span id="cb272-16"><a href="spatiallibd-workflow.html#cb272-16" tabindex="-1"></a></span>
<span id="cb272-17"><a href="spatiallibd-workflow.html#cb272-17" tabindex="-1"></a><span class="do">## Inspect the gene annotation data we added</span></span>
<span id="cb272-18"><a href="spatiallibd-workflow.html#cb272-18" tabindex="-1"></a><span class="fu">rowRanges</span>(spe)</span></code></pre></div>
<pre><code>## GRanges object with 33267 ranges and 12 metadata columns:
##                   seqnames            ranges strand |   source     type
##                      &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;factor&gt;
##   ENSG00000243485     chr1       29554-31109      + |   HAVANA     gene
##   ENSG00000237613     chr1       34554-36081      - |   HAVANA     gene
##   ENSG00000186092     chr1       65419-71585      + |   HAVANA     gene
##   ENSG00000238009     chr1      89295-133723      - |   HAVANA     gene
##   ENSG00000239945     chr1       89551-91105      - |   HAVANA     gene
##               ...      ...               ...    ... .      ...      ...
##   ENSG00000160298    chr21 46300181-46323875      - |   HAVANA     gene
##   ENSG00000160299    chr21 46324141-46445769      + |   HAVANA     gene
##   ENSG00000160305    chr21 46458891-46570015      + |   HAVANA     gene
##   ENSG00000160307    chr21 46598604-46605208      - |   HAVANA     gene
##   ENSG00000160310    chr21 46635595-46665124      + |   HAVANA     gene
##                           gene_id   gene_name      gene_type gene_interest_1
##                       &lt;character&gt; &lt;character&gt;    &lt;character&gt;       &lt;logical&gt;
##   ENSG00000243485 ENSG00000243485 MIR1302-2HG         lncRNA            TRUE
##   ENSG00000237613 ENSG00000237613     FAM138A         lncRNA            TRUE
##   ENSG00000186092 ENSG00000186092       OR4F5 protein_coding            TRUE
##   ENSG00000238009 ENSG00000238009  AL627309.1         lncRNA            TRUE
##   ENSG00000239945 ENSG00000239945  AL627309.3         lncRNA            TRUE
##               ...             ...         ...            ...             ...
##   ENSG00000160298 ENSG00000160298    C21orf58 protein_coding           FALSE
##   ENSG00000160299 ENSG00000160299        PCNT protein_coding           FALSE
##   ENSG00000160305 ENSG00000160305       DIP2A protein_coding           FALSE
##   ENSG00000160307 ENSG00000160307       S100B protein_coding           FALSE
##   ENSG00000160310 ENSG00000160310       PRMT2 protein_coding           FALSE
##                   gene_interest_2 gene_interest_3 gene_interest_4
##                         &lt;logical&gt;       &lt;logical&gt;       &lt;logical&gt;
##   ENSG00000243485            TRUE            TRUE            TRUE
##   ENSG00000237613            TRUE            TRUE            TRUE
##   ENSG00000186092            TRUE            TRUE            TRUE
##   ENSG00000238009            TRUE            TRUE            TRUE
##   ENSG00000239945            TRUE            TRUE            TRUE
##               ...             ...             ...             ...
##   ENSG00000160298           FALSE           FALSE           FALSE
##   ENSG00000160299           FALSE           FALSE           FALSE
##   ENSG00000160305           FALSE           FALSE           FALSE
##   ENSG00000160307           FALSE           FALSE           FALSE
##   ENSG00000160310           FALSE           FALSE           FALSE
##                   gene_interest_5 gene_interest_6 gene_interest_7
##                         &lt;logical&gt;       &lt;logical&gt;       &lt;logical&gt;
##   ENSG00000243485            TRUE            TRUE            TRUE
##   ENSG00000237613            TRUE            TRUE            TRUE
##   ENSG00000186092            TRUE            TRUE            TRUE
##   ENSG00000238009            TRUE            TRUE            TRUE
##   ENSG00000239945            TRUE            TRUE            TRUE
##               ...             ...             ...             ...
##   ENSG00000160298           FALSE           FALSE           FALSE
##   ENSG00000160299           FALSE           FALSE           FALSE
##   ENSG00000160305           FALSE           FALSE           FALSE
##   ENSG00000160307           FALSE           FALSE           FALSE
##   ENSG00000160310           FALSE           FALSE           FALSE
##   -------
##   seqinfo: 25 sequences from an unspecified genome; no seqlengths</code></pre>
<p>Now that we have the gene annotation information, we can use it to add a few more pieces to our <code>spe</code> object that <code>spatialLIBD</code> will use. For example, we want to enable users to search genes by either their gene symbol or their Ensembl ID. We also like to visualize the amount and percent of the mitochondrial gene expression.</p>
<div class="sourceCode" id="cb274"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb274-1"><a href="spatiallibd-workflow.html#cb274-1" tabindex="-1"></a><span class="do">## Add information used by spatialLIBD</span></span>
<span id="cb274-2"><a href="spatiallibd-workflow.html#cb274-2" tabindex="-1"></a><span class="fu">rowData</span>(spe)<span class="sc">$</span>gene_search <span class="ot">&lt;-</span> <span class="fu">paste0</span>(</span>
<span id="cb274-3"><a href="spatiallibd-workflow.html#cb274-3" tabindex="-1"></a>    <span class="fu">rowData</span>(spe)<span class="sc">$</span>gene_name, <span class="st">&quot;; &quot;</span>, <span class="fu">rowData</span>(spe)<span class="sc">$</span>gene_id</span>
<span id="cb274-4"><a href="spatiallibd-workflow.html#cb274-4" tabindex="-1"></a>)</span>
<span id="cb274-5"><a href="spatiallibd-workflow.html#cb274-5" tabindex="-1"></a></span>
<span id="cb274-6"><a href="spatiallibd-workflow.html#cb274-6" tabindex="-1"></a><span class="do">## Compute chrM expression and chrM expression ratio</span></span>
<span id="cb274-7"><a href="spatiallibd-workflow.html#cb274-7" tabindex="-1"></a>is_mito <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">seqnames</span>(spe) <span class="sc">==</span> <span class="st">&quot;chrM&quot;</span>)</span>
<span id="cb274-8"><a href="spatiallibd-workflow.html#cb274-8" tabindex="-1"></a>spe<span class="sc">$</span>expr_chrM <span class="ot">&lt;-</span> <span class="fu">colSums</span>(<span class="fu">counts</span>(spe)[is_mito, , <span class="at">drop =</span> <span class="cn">FALSE</span>])</span>
<span id="cb274-9"><a href="spatiallibd-workflow.html#cb274-9" tabindex="-1"></a>spe<span class="sc">$</span>expr_chrM_ratio <span class="ot">&lt;-</span> spe<span class="sc">$</span>expr_chrM <span class="sc">/</span> spe<span class="sc">$</span>sum_umi</span></code></pre></div>
</div>
<div id="extra-information-and-filtering" class="section level3 hasAnchor" number="15.4.3">
<h3><span class="header-section-number">15.4.3</span> Extra information and filtering<a href="spatiallibd-workflow.html#extra-information-and-filtering" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Now that we have the full gene annotation information we need, we can proceed to add some last touches as well as <a href="http://research.libd.org/spatialLIBD/articles/TenX_data_download.html#filter-the-spe-object-1">filter the object</a> to reduce the memory required for visualizing the data.</p>
<div class="sourceCode" id="cb275"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb275-1"><a href="spatiallibd-workflow.html#cb275-1" tabindex="-1"></a><span class="do">## Add a variable for saving the manual annotations</span></span>
<span id="cb275-2"><a href="spatiallibd-workflow.html#cb275-2" tabindex="-1"></a>spe<span class="sc">$</span>ManualAnnotation <span class="ot">&lt;-</span> <span class="st">&quot;NA&quot;</span></span>
<span id="cb275-3"><a href="spatiallibd-workflow.html#cb275-3" tabindex="-1"></a></span>
<span id="cb275-4"><a href="spatiallibd-workflow.html#cb275-4" tabindex="-1"></a><span class="do">## Remove genes with no data</span></span>
<span id="cb275-5"><a href="spatiallibd-workflow.html#cb275-5" tabindex="-1"></a>no_expr <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">rowSums</span>(<span class="fu">counts</span>(spe)) <span class="sc">==</span> <span class="dv">0</span>)</span>
<span id="cb275-6"><a href="spatiallibd-workflow.html#cb275-6" tabindex="-1"></a></span>
<span id="cb275-7"><a href="spatiallibd-workflow.html#cb275-7" tabindex="-1"></a><span class="do">## Number of genes with no counts</span></span>
<span id="cb275-8"><a href="spatiallibd-workflow.html#cb275-8" tabindex="-1"></a><span class="fu">length</span>(no_expr)</span></code></pre></div>
<pre><code>## [1] 11596</code></pre>
<div class="sourceCode" id="cb277"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb277-1"><a href="spatiallibd-workflow.html#cb277-1" tabindex="-1"></a><span class="do">## Compute the percent of genes with no counts</span></span>
<span id="cb277-2"><a href="spatiallibd-workflow.html#cb277-2" tabindex="-1"></a><span class="fu">length</span>(no_expr) <span class="sc">/</span> <span class="fu">nrow</span>(spe) <span class="sc">*</span> <span class="dv">100</span></span></code></pre></div>
<pre><code>## [1] 34.85737</code></pre>
<div class="sourceCode" id="cb279"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb279-1"><a href="spatiallibd-workflow.html#cb279-1" tabindex="-1"></a>spe <span class="ot">&lt;-</span> spe[<span class="sc">-</span>no_expr, , drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb279-2"><a href="spatiallibd-workflow.html#cb279-2" tabindex="-1"></a></span>
<span id="cb279-3"><a href="spatiallibd-workflow.html#cb279-3" tabindex="-1"></a><span class="do">## Remove spots without counts</span></span>
<span id="cb279-4"><a href="spatiallibd-workflow.html#cb279-4" tabindex="-1"></a><span class="fu">summary</span>(spe<span class="sc">$</span>sum_umi)</span></code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##     537    2414    3466    3870    4938   15862</code></pre>
<div class="sourceCode" id="cb281"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb281-1"><a href="spatiallibd-workflow.html#cb281-1" tabindex="-1"></a><span class="do">## If we had spots with no counts, we would remove them</span></span>
<span id="cb281-2"><a href="spatiallibd-workflow.html#cb281-2" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">any</span>(spe<span class="sc">$</span>sum_umi <span class="sc">==</span> <span class="dv">0</span>)) {</span>
<span id="cb281-3"><a href="spatiallibd-workflow.html#cb281-3" tabindex="-1"></a>    spots_no_counts <span class="ot">&lt;-</span> <span class="fu">which</span>(spe<span class="sc">$</span>sum_umi <span class="sc">==</span> <span class="dv">0</span>)</span>
<span id="cb281-4"><a href="spatiallibd-workflow.html#cb281-4" tabindex="-1"></a>    <span class="do">## Number of spots with no counts</span></span>
<span id="cb281-5"><a href="spatiallibd-workflow.html#cb281-5" tabindex="-1"></a>    <span class="fu">print</span>(<span class="fu">length</span>(spots_no_counts))</span>
<span id="cb281-6"><a href="spatiallibd-workflow.html#cb281-6" tabindex="-1"></a>    <span class="do">## Percent of spots with no counts</span></span>
<span id="cb281-7"><a href="spatiallibd-workflow.html#cb281-7" tabindex="-1"></a>    <span class="fu">print</span>(<span class="fu">length</span>(spots_no_counts) <span class="sc">/</span> <span class="fu">ncol</span>(spe) <span class="sc">*</span> <span class="dv">100</span>)</span>
<span id="cb281-8"><a href="spatiallibd-workflow.html#cb281-8" tabindex="-1"></a>    spe <span class="ot">&lt;-</span> spe[, <span class="sc">-</span>spots_no_counts, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb281-9"><a href="spatiallibd-workflow.html#cb281-9" tabindex="-1"></a>}</span></code></pre></div>
<p>We <em>think</em> that we are ready to proceed to making our interactive website. Let’s use the <code>spatialLIBD::check_spe()</code> function, just to verify that we are right. If we aren’t, then it’ll try to tell us what we missed.</p>
<div class="sourceCode" id="cb282"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb282-1"><a href="spatiallibd-workflow.html#cb282-1" tabindex="-1"></a><span class="do">## Run check_spe() function</span></span>
<span id="cb282-2"><a href="spatiallibd-workflow.html#cb282-2" tabindex="-1"></a>spatialLIBD<span class="sc">::</span><span class="fu">check_spe</span>(spe)</span></code></pre></div>
<pre><code>## class: SpatialExperiment 
## dim: 21671 3524 
## metadata(0):
## assays(2): counts logcounts
## rownames(21671): ENSG00000243485 ENSG00000238009 ... ENSG00000160307
##   ENSG00000160310
## rowData names(13): source type ... gene_interest_7 gene_search
## colnames(3524): AAACAAGTATCTCCCA-1 AAACACCAATAACTGC-1 ...
##   TTGTTTCCATACAACT-1 TTGTTTGTGTAAATTC-1
## colData names(22): barcode_id sample_id ... expr_chrM_ratio
##   ManualAnnotation
## reducedDimNames(2): PCA UMAP
## mainExpName: NULL
## altExpNames(0):
## spatialCoords names(2) : pxl_col_in_fullres pxl_row_in_fullres
## imgData names(4): sample_id image_id data scaleFactor</code></pre>
<div class="sourceCode" id="cb284"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb284-1"><a href="spatiallibd-workflow.html#cb284-1" tabindex="-1"></a><span class="do">## End tracking time</span></span>
<span id="cb284-2"><a href="spatiallibd-workflow.html#cb284-2" tabindex="-1"></a>time_end <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span>
<span id="cb284-3"><a href="spatiallibd-workflow.html#cb284-3" tabindex="-1"></a></span>
<span id="cb284-4"><a href="spatiallibd-workflow.html#cb284-4" tabindex="-1"></a><span class="do">## How long this code took to run</span></span>
<span id="cb284-5"><a href="spatiallibd-workflow.html#cb284-5" tabindex="-1"></a>time_end <span class="sc">-</span> time_prereqs</span></code></pre></div>
<pre><code>## Time difference of 47.56244 secs</code></pre>
<p>Creating our final <code>spe</code> object took 2.21641933520635 to run. So you might want to save this object for later use.</p>
<div class="sourceCode" id="cb286"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb286-1"><a href="spatiallibd-workflow.html#cb286-1" tabindex="-1"></a><span class="fu">saveRDS</span>(spe, <span class="at">file =</span> <span class="st">&quot;spe_workflow_Visium_spatialLIBD.rds&quot;</span>)</span></code></pre></div>
<p>You can then re-load it with the following code on a later session.</p>
<div class="sourceCode" id="cb287"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb287-1"><a href="spatiallibd-workflow.html#cb287-1" tabindex="-1"></a>spe <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;spe_workflow_Visium_spatialLIBD.rds&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="explore-the-data" class="section level2 hasAnchor" number="15.5">
<h2><span class="header-section-number">15.5</span> Explore the data<a href="spatiallibd-workflow.html#explore-the-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In order to visualize the data, we can then use <code>spatialLIBD::vis_gene()</code>. Note that we didn’t need to do all that hard work just for that. But well, this is a nice quick check before we try launching our interactive website.</p>
<div class="sourceCode" id="cb288"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb288-1"><a href="spatiallibd-workflow.html#cb288-1" tabindex="-1"></a><span class="do">## Sum of UMI</span></span>
<span id="cb288-2"><a href="spatiallibd-workflow.html#cb288-2" tabindex="-1"></a>spatialLIBD<span class="sc">::</span><span class="fu">vis_gene</span>(</span>
<span id="cb288-3"><a href="spatiallibd-workflow.html#cb288-3" tabindex="-1"></a>    <span class="at">spe =</span> spe,</span>
<span id="cb288-4"><a href="spatiallibd-workflow.html#cb288-4" tabindex="-1"></a>    <span class="at">sampleid =</span> <span class="st">&quot;sample_151673&quot;</span>,</span>
<span id="cb288-5"><a href="spatiallibd-workflow.html#cb288-5" tabindex="-1"></a>    <span class="at">geneid =</span> <span class="st">&quot;sum_umi&quot;</span></span>
<span id="cb288-6"><a href="spatiallibd-workflow.html#cb288-6" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="BestPracticesST_files/figure-html/spatialLIBD_vis_gene-1.png" width="768" /></p>
<div class="sourceCode" id="cb289"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb289-1"><a href="spatiallibd-workflow.html#cb289-1" tabindex="-1"></a><span class="do">## PCP4, a layer 5 marker gene</span></span>
<span id="cb289-2"><a href="spatiallibd-workflow.html#cb289-2" tabindex="-1"></a>spatialLIBD<span class="sc">::</span><span class="fu">vis_gene</span>(</span>
<span id="cb289-3"><a href="spatiallibd-workflow.html#cb289-3" tabindex="-1"></a>    <span class="at">spe =</span> spe,</span>
<span id="cb289-4"><a href="spatiallibd-workflow.html#cb289-4" tabindex="-1"></a>    <span class="at">sampleid =</span> <span class="st">&quot;sample_151673&quot;</span>,</span>
<span id="cb289-5"><a href="spatiallibd-workflow.html#cb289-5" tabindex="-1"></a>    <span class="at">geneid =</span> <span class="fu">rowData</span>(spe)<span class="sc">$</span>gene_search[<span class="fu">which</span>(<span class="fu">rowData</span>(spe)<span class="sc">$</span>gene_name <span class="sc">==</span> <span class="st">&quot;PCP4&quot;</span>)]</span>
<span id="cb289-6"><a href="spatiallibd-workflow.html#cb289-6" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="BestPracticesST_files/figure-html/spatialLIBD_vis_gene-2.png" width="768" /></p>
<p>As we wanted let’s proceed to <a href="http://research.libd.org/spatialLIBD/articles/TenX_data_download.html#run-the-interactive-website-1">visualize the data interactively</a> with a <code>spatialLIBD</code>-powered website. We have lots of variables to choose from. We’ll specify which are our continuous and discrete variables in our <code>spatialLIBD::run_app()</code> call.</p>
<div class="sourceCode" id="cb290"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb290-1"><a href="spatiallibd-workflow.html#cb290-1" tabindex="-1"></a><span class="do">## Explore all the variables we can use</span></span>
<span id="cb290-2"><a href="spatiallibd-workflow.html#cb290-2" tabindex="-1"></a><span class="fu">colData</span>(spe)</span></code></pre></div>
<pre><code>## DataFrame with 3524 rows and 22 columns
##                            barcode_id     sample_id in_tissue array_row
##                           &lt;character&gt;   &lt;character&gt; &lt;integer&gt; &lt;integer&gt;
## AAACAAGTATCTCCCA-1 AAACAAGTATCTCCCA-1 sample_151673         1        50
## AAACACCAATAACTGC-1 AAACACCAATAACTGC-1 sample_151673         1        59
## AAACAGAGCGACTCCT-1 AAACAGAGCGACTCCT-1 sample_151673         1        14
## AAACAGCTTTCAGAAG-1 AAACAGCTTTCAGAAG-1 sample_151673         1        43
## AAACAGGGTCTATATT-1 AAACAGGGTCTATATT-1 sample_151673         1        47
## ...                               ...           ...       ...       ...
## TTGTTGTGTGTCAAGA-1 TTGTTGTGTGTCAAGA-1 sample_151673         1        31
## TTGTTTCACATCCAGG-1 TTGTTTCACATCCAGG-1 sample_151673         1        58
## TTGTTTCATTAGTCTA-1 TTGTTTCATTAGTCTA-1 sample_151673         1        60
## TTGTTTCCATACAACT-1 TTGTTTCCATACAACT-1 sample_151673         1        45
## TTGTTTGTGTAAATTC-1 TTGTTTGTGTAAATTC-1 sample_151673         1         7
##                    array_col ground_truth cell_count       sum  detected
##                    &lt;integer&gt;  &lt;character&gt;  &lt;integer&gt; &lt;numeric&gt; &lt;numeric&gt;
## AAACAAGTATCTCCCA-1       102       Layer3          6      8458      3586
## AAACACCAATAACTGC-1        19           WM          5      3769      1960
## AAACAGAGCGACTCCT-1        94       Layer3          2      5433      2424
## AAACAGCTTTCAGAAG-1         9       Layer5          4      4278      2264
## AAACAGGGTCTATATT-1        13       Layer6          6      4004      2178
## ...                      ...          ...        ...       ...       ...
## TTGTTGTGTGTCAAGA-1        77       Layer5          3      3966      1982
## TTGTTTCACATCCAGG-1        42           WM          3      4324      2170
## TTGTTTCATTAGTCTA-1        30           WM          4      2761      1560
## TTGTTTCCATACAACT-1        27       Layer6          3      2322      1343
## TTGTTTGTGTAAATTC-1        51       Layer2          5      6281      2927
##                    subsets_mito_sum subsets_mito_detected subsets_mito_percent
##                           &lt;numeric&gt;             &lt;numeric&gt;            &lt;numeric&gt;
## AAACAAGTATCTCCCA-1             1407                    13              16.6351
## AAACACCAATAACTGC-1              430                    13              11.4089
## AAACAGAGCGACTCCT-1             1316                    13              24.2223
## AAACAGCTTTCAGAAG-1              651                    12              15.2174
## AAACAGGGTCTATATT-1              621                    13              15.5095
## ...                             ...                   ...                  ...
## TTGTTGTGTGTCAAGA-1              789                    13             19.89410
## TTGTTTCACATCCAGG-1              370                    12              8.55689
## TTGTTTCATTAGTCTA-1              314                    12             11.37269
## TTGTTTCCATACAACT-1              476                    13             20.49957
## TTGTTTGTGTAAATTC-1              991                    13             15.77774
##                        total   discard sizeFactor    label
##                    &lt;numeric&gt; &lt;logical&gt;  &lt;numeric&gt; &lt;factor&gt;
## AAACAAGTATCTCCCA-1      8458     FALSE   1.822839        3
## AAACACCAATAACTGC-1      3769     FALSE   0.812282        1
## AAACAGAGCGACTCCT-1      5433     FALSE   1.170902        3
## AAACAGCTTTCAGAAG-1      4278     FALSE   0.921980        3
## AAACAGGGTCTATATT-1      4004     FALSE   0.862928        6
## ...                      ...       ...        ...      ...
## TTGTTGTGTGTCAAGA-1      3966     FALSE   0.854739        4
## TTGTTTCACATCCAGG-1      4324     FALSE   0.931894        1
## TTGTTTCATTAGTCTA-1      2761     FALSE   0.595041        1
## TTGTTTCCATACAACT-1      2322     FALSE   0.500429        2
## TTGTTTGTGTAAATTC-1      6281     FALSE   1.353660        7
##                                       key   sum_umi  sum_gene expr_chrM
##                               &lt;character&gt; &lt;numeric&gt; &lt;integer&gt; &lt;numeric&gt;
## AAACAAGTATCTCCCA-1 sample_151673_AAACAA..      7051      3573         0
## AAACACCAATAACTGC-1 sample_151673_AAACAC..      3339      1947         0
## AAACAGAGCGACTCCT-1 sample_151673_AAACAG..      4117      2411         0
## AAACAGCTTTCAGAAG-1 sample_151673_AAACAG..      3627      2252         0
## AAACAGGGTCTATATT-1 sample_151673_AAACAG..      3383      2165         0
## ...                                   ...       ...       ...       ...
## TTGTTGTGTGTCAAGA-1 sample_151673_TTGTTG..      3177      1969         0
## TTGTTTCACATCCAGG-1 sample_151673_TTGTTT..      3954      2158         0
## TTGTTTCATTAGTCTA-1 sample_151673_TTGTTT..      2447      1548         0
## TTGTTTCCATACAACT-1 sample_151673_TTGTTT..      1846      1330         0
## TTGTTTGTGTAAATTC-1 sample_151673_TTGTTT..      5290      2914         0
##                    expr_chrM_ratio ManualAnnotation
##                          &lt;numeric&gt;      &lt;character&gt;
## AAACAAGTATCTCCCA-1               0               NA
## AAACACCAATAACTGC-1               0               NA
## AAACAGAGCGACTCCT-1               0               NA
## AAACAGCTTTCAGAAG-1               0               NA
## AAACAGGGTCTATATT-1               0               NA
## ...                            ...              ...
## TTGTTGTGTGTCAAGA-1               0               NA
## TTGTTTCACATCCAGG-1               0               NA
## TTGTTTCATTAGTCTA-1               0               NA
## TTGTTTCCATACAACT-1               0               NA
## TTGTTTGTGTAAATTC-1               0               NA</code></pre>
<div class="sourceCode" id="cb292"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb292-1"><a href="spatiallibd-workflow.html#cb292-1" tabindex="-1"></a><span class="do">## Run our shiny app</span></span>
<span id="cb292-2"><a href="spatiallibd-workflow.html#cb292-2" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">interactive</span>()) {</span>
<span id="cb292-3"><a href="spatiallibd-workflow.html#cb292-3" tabindex="-1"></a>    spatialLIBD<span class="sc">::</span><span class="fu">run_app</span>(</span>
<span id="cb292-4"><a href="spatiallibd-workflow.html#cb292-4" tabindex="-1"></a>        spe,</span>
<span id="cb292-5"><a href="spatiallibd-workflow.html#cb292-5" tabindex="-1"></a>        <span class="at">sce_layer =</span> <span class="cn">NULL</span>,</span>
<span id="cb292-6"><a href="spatiallibd-workflow.html#cb292-6" tabindex="-1"></a>        <span class="at">modeling_results =</span> <span class="cn">NULL</span>,</span>
<span id="cb292-7"><a href="spatiallibd-workflow.html#cb292-7" tabindex="-1"></a>        <span class="at">sig_genes =</span> <span class="cn">NULL</span>,</span>
<span id="cb292-8"><a href="spatiallibd-workflow.html#cb292-8" tabindex="-1"></a>        <span class="at">title =</span> <span class="st">&quot;OSTA spatialLIBD workflow example&quot;</span>,</span>
<span id="cb292-9"><a href="spatiallibd-workflow.html#cb292-9" tabindex="-1"></a>        <span class="at">spe_discrete_vars =</span> <span class="fu">c</span>(<span class="st">&quot;ground_truth&quot;</span>, <span class="st">&quot;label&quot;</span>, <span class="st">&quot;ManualAnnotation&quot;</span>),</span>
<span id="cb292-10"><a href="spatiallibd-workflow.html#cb292-10" tabindex="-1"></a>        <span class="at">spe_continuous_vars =</span> <span class="fu">c</span>(</span>
<span id="cb292-11"><a href="spatiallibd-workflow.html#cb292-11" tabindex="-1"></a>            <span class="st">&quot;cell_count&quot;</span>,</span>
<span id="cb292-12"><a href="spatiallibd-workflow.html#cb292-12" tabindex="-1"></a>            <span class="st">&quot;sum_umi&quot;</span>,</span>
<span id="cb292-13"><a href="spatiallibd-workflow.html#cb292-13" tabindex="-1"></a>            <span class="st">&quot;sum_gene&quot;</span>,</span>
<span id="cb292-14"><a href="spatiallibd-workflow.html#cb292-14" tabindex="-1"></a>            <span class="st">&quot;expr_chrM&quot;</span>,</span>
<span id="cb292-15"><a href="spatiallibd-workflow.html#cb292-15" tabindex="-1"></a>            <span class="st">&quot;expr_chrM_ratio&quot;</span>,</span>
<span id="cb292-16"><a href="spatiallibd-workflow.html#cb292-16" tabindex="-1"></a>            <span class="st">&quot;sum&quot;</span>,</span>
<span id="cb292-17"><a href="spatiallibd-workflow.html#cb292-17" tabindex="-1"></a>            <span class="st">&quot;detected&quot;</span>,</span>
<span id="cb292-18"><a href="spatiallibd-workflow.html#cb292-18" tabindex="-1"></a>            <span class="st">&quot;subsets_mito_sum&quot;</span>,</span>
<span id="cb292-19"><a href="spatiallibd-workflow.html#cb292-19" tabindex="-1"></a>            <span class="st">&quot;subsets_mito_detected&quot;</span>,</span>
<span id="cb292-20"><a href="spatiallibd-workflow.html#cb292-20" tabindex="-1"></a>            <span class="st">&quot;subsets_mito_percent&quot;</span>,</span>
<span id="cb292-21"><a href="spatiallibd-workflow.html#cb292-21" tabindex="-1"></a>            <span class="st">&quot;total&quot;</span>,</span>
<span id="cb292-22"><a href="spatiallibd-workflow.html#cb292-22" tabindex="-1"></a>            <span class="st">&quot;sizeFactor&quot;</span></span>
<span id="cb292-23"><a href="spatiallibd-workflow.html#cb292-23" tabindex="-1"></a>        ),</span>
<span id="cb292-24"><a href="spatiallibd-workflow.html#cb292-24" tabindex="-1"></a>        <span class="at">default_cluster =</span> <span class="st">&quot;label&quot;</span></span>
<span id="cb292-25"><a href="spatiallibd-workflow.html#cb292-25" tabindex="-1"></a>    )</span>
<span id="cb292-26"><a href="spatiallibd-workflow.html#cb292-26" tabindex="-1"></a>}</span></code></pre></div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-112"></span>
<img src="images/spatialLIBD_demo_result.png" alt="Screenshot of the 'clusters (interactive)' section of the 'spot-level data' panel created with with the data from this workflow." width="100%" />
<p class="caption">
Figure 15.4: Screenshot of the ‘clusters (interactive)’ section of the ‘spot-level data’ panel created with with the data from this workflow.
</p>
</div>
</div>
<div id="sharing-your-website" class="section level2 hasAnchor" number="15.6">
<h2><span class="header-section-number">15.6</span> Sharing your website<a href="spatiallibd-workflow.html#sharing-your-website" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Now that you have created a <code>spatialLIBD</code>-powered website, you might be interested in sharing it. To do so, it’ll be useful to have saved a small <code>spe</code> object using <code>saveRDS()</code> like we did earlier. The smaller the object, the better in terms of performance. For example, you might want to remove the lowly expressed genes to save space. One function you can use to measure how big your object is is <code>lobstr::obj_size()</code> as shown below.</p>
<div class="sourceCode" id="cb293"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb293-1"><a href="spatiallibd-workflow.html#cb293-1" tabindex="-1"></a><span class="do">## Object size</span></span>
<span id="cb293-2"><a href="spatiallibd-workflow.html#cb293-2" tabindex="-1"></a>lobstr<span class="sc">::</span><span class="fu">obj_size</span>(spe) <span class="sc">/</span> <span class="dv">1024</span><span class="sc">^</span><span class="dv">2</span> <span class="do">## Convert to MB</span></span></code></pre></div>
<pre><code>## 293.12 B</code></pre>
<p>If your data is small enough, you might want to share your website by hosting on <a href="https://www.shinyapps.io/">shinyapps.io</a> by RStudio, which you can try for free. Once you have created your account, you’ll want to create an <code>app.R</code> file like the one we have <a href="https://github.com/lmweber/BestPracticesST/blob/main/spatialLIBD_demo">on the <code>spatialLIBD_demo</code> directory</a>.</p>
<pre><code>## library(&quot;spatialLIBD&quot;)
##  library(&quot;markdown&quot;) # for shinyapps.io
##  
##  ## spatialLIBD uses golem
##  options(&quot;golem.app.prod&quot; = TRUE)
##  
##  ## You need this to enable shinyapps to install Bioconductor packages
##  options(repos = BiocManager::repositories())
##  
##  ## Load the data
##  spe &lt;- readRDS(&quot;spe_workflow_Visium_spatialLIBD.rds&quot;)
##  
##  ## Deploy the website
##  spatialLIBD::run_app(
##      spe,
##      sce_layer = NULL,
##      modeling_results = NULL,
##      sig_genes = NULL,
##      title = &quot;OSTA spatialLIBD workflow example&quot;,
##      spe_discrete_vars = c(&quot;ground_truth&quot;, &quot;label&quot;, &quot;ManualAnnotation&quot;),
##      spe_continuous_vars = c(
##          &quot;cell_count&quot;,
##          &quot;sum_umi&quot;,
##          &quot;sum_gene&quot;,
##          &quot;expr_chrM&quot;,
##          &quot;expr_chrM_ratio&quot;,
##          &quot;sum&quot;,
##          &quot;detected&quot;,
##          &quot;subsets_mito_sum&quot;,
##          &quot;subsets_mito_detected&quot;,
##          &quot;subsets_mito_percent&quot;,
##          &quot;total&quot;,
##          &quot;sizeFactor&quot;
##      ),
##      default_cluster = &quot;label&quot;,
##      docs_path = &quot;www&quot;
##  )</code></pre>
<p>You can then open R in a new session in the same directory where you saved the <code>app.R</code> file, run the code and click on the “publish” blue button at the top right of your RStudio window. You’ll then need to upload the <code>app.R</code> file, your <code>spe_workflow_Visium_spatialLIBD.rds</code> file and the files under the <code>www</code> directory which enable you to customize your <code>spatialLIDB</code> website.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:unnamed-chunk-113"></span>
<img src="images/spatialLIBD_publish.png" alt="Screenshot of the RStudio window for publishing your spatialLIBD-powered website to shinyapps.io" width="80%" />
<p class="caption">
Figure 15.5: Screenshot of the RStudio window for publishing your spatialLIBD-powered website to shinyapps.io
</p>
</div>
<p>The RStudio prompts will guide you along the process for authenticating to your shinyapps.io account, which will involve copy pasting some code that starts with <code>rsconnect::setAccountInfo()</code>. Alternatively, you can create a <code>deploy.R</code> script and write the code for uploading your files to shinyapps.io as shown below.</p>
<pre><code>## library(&#39;rsconnect&#39;)
##  
##  ## Or you can go to your shinyapps.io account and copy this
##  ## Here we do this to keep our information hidden.
##  load(&quot;.deploy_info.Rdata&quot;)
##  rsconnect::setAccountInfo(
##      name = deploy_info$name,
##      token = deploy_info$token,
##      secret = deploy_info$secret
##  )
##  
##  ## You need this to enable shinyapps to install Bioconductor packages
##  options(repos = BiocManager::repositories())
##  
##  ## Deploy the app, that is, upload it to shinyapps.io
##  rsconnect::deployApp(
##      appFiles = c(
##          &quot;app.R&quot;,
##          &quot;spe_workflow_Visium_spatialLIBD.rds&quot;,
##          dir(&quot;www&quot;, full.names = TRUE)
##      ),
##      appName = &#39;OSTA_spatialLIBD_demo&#39;,
##      account = &#39;libd&#39;,
##      server = &#39;shinyapps.io&#39;
##  )</code></pre>
<p>Note that we have copied the default <a href="https://github.com/LieberInstitute/spatialLIBD/tree/master/inst/app/www"><code>www</code> directory files from the <code>spatialLIBD</code> repository</a> and <a href="(https://github.com/lmweber/BestPracticesST/blob/main/spatialLIBD_demo/www)">adapted them to our liking</a>. We then use these files with <code>spatialLIBD::run_app(docs_path)</code> in our <code>app.R</code> script. These files help us control portions of our <code>spatialLIBD</code>-powered website and customize it to our liking.</p>
<p>If you follow this workflow, you’ll end up with a website <a href="https://libd.shinyapps.io/OSTA_spatialLIBD_demo/">just like this one</a>. In our case, we further configured our website through the shinyapps.io dashboard. We selected the following options:</p>
<ul>
<li><em>General</em> <code>Instance Size</code>: 3X-Large (8GB)</li>
<li><em>Advanced</em> <code>Max Worker Processes</code>: 1</li>
<li><em>Advanced</em> <code>Max Connections</code>: 15</li>
</ul>
<p>The <code>Max Worker Processes</code> determines how many R sessions are open per instance. Then <code>Max Connections</code> specifies the number of connections to each R session. The <code>Instance Size</code> determines the memory available. In this case, 8000 / 300 is approximately 27, but we decided to be conservative and set the total number of users per instance to be 15. This is why it can be important to reduce the size of your <code>spe</code> object before sharing the website. Alternatively, you can rent an AWS Instance and deploy your app there, which is how <a href="http://spatial.libd.org/spatialLIBD" class="uri">http://spatial.libd.org/spatialLIBD</a> is hosted along with these <a href="https://github.com/LieberInstitute/spatialLIBD/tree/master/dev/shiny-server-files">error configuration files</a>.</p>
</div>
<div id="wrapping-up" class="section level2 hasAnchor" number="15.7">
<h2><span class="header-section-number">15.7</span> Wrapping up<a href="spatiallibd-workflow.html#wrapping-up" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Thank you for reading this far! In this workflow we showed you:</p>
<ul>
<li>why you might be interested in using <code>spatialLIBD</code>,</li>
<li>we re-used the <code>spe</code> object from the <a href="human-dlpfc-workflow.html#human-dlpfc-workflow">Human DLPFC workflow</a> chapter,</li>
<li>we adapted the <code>spe</code> object to make it compatible with <code>spatialLIBD</code>,</li>
<li>we created an interactive website in our laptops,</li>
<li>we shared the website with others using shinyapps.io.</li>
</ul>
<p>Overall, we hope that you found this information useful and we wish you the best luck with exploring and annotating your own Visium data!</p>
</div>
<div id="r-session-information" class="section level2 unnumbered hasAnchor">
<h2>R session information<a href="spatiallibd-workflow.html#r-session-information" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Here’s the R session information for this workflow.</p>
<div class="sourceCode" id="cb297"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb297-1"><a href="spatiallibd-workflow.html#cb297-1" tabindex="-1"></a><span class="fu">options</span>(<span class="at">width =</span> <span class="dv">120</span>)</span>
<span id="cb297-2"><a href="spatiallibd-workflow.html#cb297-2" tabindex="-1"></a>sessioninfo<span class="sc">::</span><span class="fu">session_info</span>()</span></code></pre></div>
<pre><code>## ─ Session info ───────────────────────────────────────────────────────────────────────────────────────────────────────
##  setting  value
##  version  R version 4.3.1 (2023-06-16)
##  os       Ubuntu 22.04.3 LTS
##  system   x86_64, linux-gnu
##  ui       X11
##  language (EN)
##  collate  en_US.UTF-8
##  ctype    en_US.UTF-8
##  tz       UTC
##  date     2023-09-12
##  pandoc   3.1.1 @ /usr/local/bin/ (via rmarkdown)
## 
## ─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────
##  package                * version   date (UTC) lib source
##  abind                    1.4-5     2016-07-21 [1] RSPM (R 4.3.0)
##  AnnotationDbi            1.62.2    2023-07-02 [1] Bioconductor
##  AnnotationHub          * 3.8.0     2023-04-25 [1] Bioconductor
##  attempt                  0.3.1     2020-05-03 [1] RSPM (R 4.3.0)
##  beachmat                 2.16.0    2023-04-25 [1] Bioconductor
##  beeswarm                 0.4.0     2021-06-01 [1] RSPM (R 4.3.0)
##  benchmarkme              1.0.8     2022-06-12 [1] RSPM (R 4.3.0)
##  benchmarkmeData          1.0.4     2020-04-23 [1] RSPM (R 4.3.0)
##  Biobase                * 2.60.0    2023-04-25 [1] Bioconductor
##  BiocFileCache          * 2.8.0     2023-04-25 [1] Bioconductor
##  BiocGenerics           * 0.46.0    2023-04-25 [1] Bioconductor
##  BiocIO                   1.10.0    2023-04-25 [1] Bioconductor
##  BiocManager              1.30.22   2023-08-08 [2] CRAN (R 4.3.1)
##  BiocNeighbors            1.18.0    2023-04-25 [1] Bioconductor
##  BiocParallel             1.34.2    2023-05-22 [1] Bioconductor
##  BiocSingular             1.16.0    2023-04-25 [1] Bioconductor
##  BiocVersion              3.17.1    2022-11-04 [2] Bioconductor
##  Biostrings               2.68.1    2023-05-16 [1] Bioconductor
##  bit                      4.0.5     2022-11-15 [1] RSPM (R 4.3.0)
##  bit64                    4.0.5     2020-08-30 [1] RSPM (R 4.3.0)
##  bitops                   1.0-7     2021-04-24 [1] RSPM (R 4.3.0)
##  blob                     1.2.4     2023-03-17 [1] RSPM (R 4.3.0)
##  bluster                  1.10.0    2023-04-25 [1] Bioconductor
##  bookdown                 0.35      2023-08-09 [1] RSPM (R 4.3.0)
##  BRISC                    1.0.5     2022-04-29 [1] RSPM (R 4.3.0)
##  bslib                    0.5.1     2023-08-11 [2] RSPM (R 4.3.0)
##  cachem                   1.0.8     2023-05-01 [2] RSPM (R 4.3.0)
##  cli                      3.6.1     2023-03-23 [2] RSPM (R 4.3.0)
##  cluster                  2.1.4     2022-08-22 [3] CRAN (R 4.3.1)
##  codetools                0.2-19    2023-02-01 [3] CRAN (R 4.3.1)
##  colorspace               2.1-0     2023-01-23 [1] RSPM (R 4.3.0)
##  config                   0.3.2     2023-08-30 [1] RSPM (R 4.3.0)
##  cowplot                  1.1.1     2020-12-30 [1] RSPM (R 4.3.0)
##  crayon                   1.5.2     2022-09-29 [2] RSPM (R 4.3.0)
##  curl                     5.0.2     2023-08-14 [2] RSPM (R 4.3.0)
##  data.table               1.14.8    2023-02-17 [1] RSPM (R 4.3.0)
##  DBI                      1.1.3     2022-06-18 [1] RSPM (R 4.3.0)
##  dbplyr                 * 2.3.3     2023-07-07 [1] RSPM (R 4.3.0)
##  DelayedArray             0.26.7    2023-07-28 [1] Bioconductor
##  DelayedMatrixStats       1.22.6    2023-08-28 [1] Bioconductor
##  digest                   0.6.33    2023-07-07 [2] RSPM (R 4.3.0)
##  doParallel               1.0.17    2022-02-07 [1] RSPM (R 4.3.0)
##  dotCall64                1.0-2     2022-10-03 [1] RSPM (R 4.3.0)
##  dplyr                    1.1.3     2023-09-03 [1] RSPM (R 4.3.0)
##  dqrng                    0.3.1     2023-08-30 [1] RSPM (R 4.3.0)
##  DropletUtils             1.20.0    2023-04-25 [1] Bioconductor
##  DT                       0.29      2023-08-29 [1] RSPM (R 4.3.0)
##  edgeR                    3.42.4    2023-05-31 [1] Bioconductor
##  ellipsis                 0.3.2     2021-04-29 [2] RSPM (R 4.3.0)
##  evaluate                 0.21      2023-05-05 [2] RSPM (R 4.3.0)
##  ExperimentHub          * 2.8.1     2023-07-12 [1] Bioconductor
##  fansi                    1.0.4     2023-01-22 [2] RSPM (R 4.3.0)
##  farver                   2.1.1     2022-07-06 [1] RSPM (R 4.3.0)
##  fastmap                  1.1.1     2023-02-24 [2] RSPM (R 4.3.0)
##  fields                   15.2      2023-08-17 [1] RSPM (R 4.3.0)
##  filelock                 1.0.2     2018-10-05 [1] RSPM (R 4.3.0)
##  FNN                      1.1.3.2   2023-03-20 [1] RSPM (R 4.3.0)
##  foreach                  1.5.2     2022-02-02 [1] RSPM (R 4.3.0)
##  generics                 0.1.3     2022-07-05 [1] RSPM (R 4.3.0)
##  GenomeInfoDb           * 1.36.3    2023-09-07 [1] Bioconductor
##  GenomeInfoDbData         1.2.10    2023-08-25 [1] Bioconductor
##  GenomicAlignments        1.36.0    2023-04-25 [1] Bioconductor
##  GenomicRanges          * 1.52.0    2023-04-25 [1] Bioconductor
##  ggbeeswarm               0.7.2     2023-04-29 [1] RSPM (R 4.3.0)
##  ggplot2                * 3.4.3     2023-08-14 [1] RSPM (R 4.3.0)
##  ggrepel                  0.9.3     2023-02-03 [1] RSPM (R 4.3.0)
##  ggside                   0.2.2     2022-12-04 [1] RSPM (R 4.3.0)
##  ggspavis               * 1.6.0     2023-04-25 [1] Bioconductor
##  glue                     1.6.2     2022-02-24 [2] RSPM (R 4.3.0)
##  golem                    0.4.1     2023-06-05 [1] RSPM (R 4.3.0)
##  gridExtra                2.3       2017-09-09 [1] RSPM (R 4.3.0)
##  gtable                   0.3.4     2023-08-21 [1] RSPM (R 4.3.0)
##  HDF5Array                1.28.1    2023-05-01 [1] Bioconductor
##  highr                    0.10      2022-12-22 [2] RSPM (R 4.3.0)
##  htmltools                0.5.6     2023-08-10 [2] RSPM (R 4.3.0)
##  htmlwidgets              1.6.2     2023-03-17 [2] RSPM (R 4.3.0)
##  httpuv                   1.6.11    2023-05-11 [2] RSPM (R 4.3.0)
##  httr                     1.4.7     2023-08-15 [2] RSPM (R 4.3.0)
##  igraph                 * 1.5.1     2023-08-10 [1] RSPM (R 4.3.0)
##  interactiveDisplayBase   1.38.0    2023-04-25 [1] Bioconductor
##  IRanges                * 2.34.1    2023-06-22 [1] Bioconductor
##  irlba                    2.3.5.1   2022-10-03 [1] RSPM (R 4.3.0)
##  iterators                1.0.14    2022-02-05 [1] RSPM (R 4.3.0)
##  jquerylib                0.1.4     2021-04-26 [2] RSPM (R 4.3.0)
##  jsonlite                 1.8.7     2023-06-29 [2] RSPM (R 4.3.0)
##  KEGGREST                 1.40.0    2023-04-25 [1] Bioconductor
##  knitr                    1.43      2023-05-25 [2] RSPM (R 4.3.0)
##  labeling                 0.4.3     2023-08-29 [1] RSPM (R 4.3.0)
##  later                    1.3.1     2023-05-02 [2] RSPM (R 4.3.0)
##  lattice                  0.21-8    2023-04-05 [3] CRAN (R 4.3.1)
##  lazyeval                 0.2.2     2019-03-15 [1] RSPM (R 4.3.0)
##  lifecycle                1.0.3     2022-10-07 [2] RSPM (R 4.3.0)
##  limma                    3.56.2    2023-06-04 [1] Bioconductor
##  lobstr                 * 1.1.2     2022-06-22 [1] RSPM (R 4.3.0)
##  locfit                   1.5-9.8   2023-06-11 [1] RSPM (R 4.3.0)
##  magick                   2.7.5     2023-08-07 [1] RSPM (R 4.3.0)
##  magrittr                 2.0.3     2022-03-30 [2] RSPM (R 4.3.0)
##  maps                     3.4.1     2022-10-30 [1] RSPM (R 4.3.0)
##  Matrix                   1.6-1     2023-08-14 [3] RSPM (R 4.3.0)
##  MatrixGenerics         * 1.12.3    2023-07-30 [1] Bioconductor
##  matrixStats            * 1.0.0     2023-06-02 [1] RSPM (R 4.3.0)
##  memoise                  2.0.1     2021-11-26 [2] RSPM (R 4.3.0)
##  metapod                  1.8.0     2023-04-25 [1] Bioconductor
##  mgcv                     1.9-0     2023-07-11 [3] RSPM (R 4.3.0)
##  mime                     0.12      2021-09-28 [2] RSPM (R 4.3.0)
##  munsell                  0.5.0     2018-06-12 [1] RSPM (R 4.3.0)
##  nlme                     3.1-163   2023-08-09 [3] RSPM (R 4.3.0)
##  nnSVG                  * 1.4.2     2023-08-14 [1] Bioconductor
##  paletteer                1.5.0     2022-10-19 [1] RSPM (R 4.3.0)
##  pbapply                  1.7-2     2023-06-27 [1] RSPM (R 4.3.0)
##  pheatmap               * 1.0.12    2019-01-04 [1] RSPM (R 4.3.0)
##  pillar                   1.9.0     2023-03-22 [2] RSPM (R 4.3.0)
##  pkgconfig                2.0.3     2019-09-22 [2] RSPM (R 4.3.0)
##  plotly                   4.10.2    2023-06-03 [1] RSPM (R 4.3.0)
##  png                      0.1-8     2022-11-29 [1] RSPM (R 4.3.0)
##  prettyunits              1.1.1     2020-01-24 [2] RSPM (R 4.3.0)
##  promises                 1.2.1     2023-08-10 [2] RSPM (R 4.3.0)
##  purrr                    1.0.2     2023-08-10 [2] RSPM (R 4.3.0)
##  R.methodsS3              1.8.2     2022-06-13 [1] RSPM (R 4.3.0)
##  R.oo                     1.25.0    2022-06-12 [1] RSPM (R 4.3.0)
##  R.utils                  2.12.2    2022-11-11 [1] RSPM (R 4.3.0)
##  R6                       2.5.1     2021-08-19 [2] RSPM (R 4.3.0)
##  RANN                     2.6.1     2019-01-08 [1] RSPM (R 4.3.0)
##  rappdirs                 0.3.3     2021-01-31 [2] RSPM (R 4.3.0)
##  RColorBrewer             1.1-3     2022-04-03 [1] RSPM (R 4.3.0)
##  Rcpp                     1.0.11    2023-07-06 [2] RSPM (R 4.3.0)
##  RCurl                    1.98-1.12 2023-03-27 [1] RSPM (R 4.3.0)
##  rdist                    0.0.5     2020-05-04 [1] RSPM (R 4.3.0)
##  rematch2                 2.1.2     2020-05-01 [2] RSPM (R 4.3.0)
##  restfulr                 0.0.15    2022-06-16 [1] RSPM (R 4.3.0)
##  rhdf5                    2.44.0    2023-04-25 [1] Bioconductor
##  rhdf5filters             1.12.1    2023-04-30 [1] Bioconductor
##  Rhdf5lib                 1.22.1    2023-09-10 [1] Bioconductor
##  rjson                    0.2.21    2022-01-09 [1] RSPM (R 4.3.0)
##  rlang                    1.1.1     2023-04-28 [2] RSPM (R 4.3.0)
##  rmarkdown                2.24      2023-08-14 [2] RSPM (R 4.3.0)
##  Rsamtools                2.16.0    2023-04-25 [1] Bioconductor
##  RSQLite                  2.3.1     2023-04-03 [1] RSPM (R 4.3.0)
##  rstudioapi               0.15.0    2023-07-07 [2] RSPM (R 4.3.0)
##  rsvd                     1.0.5     2021-04-16 [1] RSPM (R 4.3.0)
##  rtracklayer            * 1.60.1    2023-08-15 [1] Bioconductor
##  S4Arrays                 1.0.6     2023-08-30 [1] Bioconductor
##  S4Vectors              * 0.38.1    2023-05-02 [1] Bioconductor
##  sass                     0.4.7     2023-07-15 [2] RSPM (R 4.3.0)
##  ScaledMatrix             1.8.1     2023-05-03 [1] Bioconductor
##  scales                   1.2.1     2022-08-20 [1] RSPM (R 4.3.0)
##  scater                 * 1.28.0    2023-04-25 [1] Bioconductor
##  scran                  * 1.28.2    2023-07-23 [1] Bioconductor
##  scuttle                * 1.10.2    2023-08-03 [1] Bioconductor
##  sessioninfo              1.2.2     2021-12-06 [2] RSPM (R 4.3.0)
##  shiny                    1.7.5     2023-08-12 [2] RSPM (R 4.3.0)
##  shinyWidgets             0.8.0     2023-08-30 [1] RSPM (R 4.3.0)
##  SingleCellExperiment   * 1.22.0    2023-04-25 [1] Bioconductor
##  spam                     2.9-1     2022-08-07 [1] RSPM (R 4.3.0)
##  sparseMatrixStats        1.12.2    2023-07-02 [1] Bioconductor
##  SpatialExperiment      * 1.10.0    2023-04-25 [1] Bioconductor
##  spatialLIBD            * 1.12.0    2023-04-27 [1] Bioconductor
##  statmod                  1.5.0     2023-01-06 [1] RSPM (R 4.3.0)
##  STexampleData          * 1.8.0     2023-04-27 [1] Bioconductor
##  SummarizedExperiment   * 1.30.2    2023-06-06 [1] Bioconductor
##  tibble                   3.2.1     2023-03-20 [2] RSPM (R 4.3.0)
##  tidyr                    1.3.0     2023-01-24 [1] RSPM (R 4.3.0)
##  tidyselect               1.2.0     2022-10-10 [1] RSPM (R 4.3.0)
##  utf8                     1.2.3     2023-01-31 [2] RSPM (R 4.3.0)
##  uwot                     0.1.16    2023-06-29 [1] RSPM (R 4.3.0)
##  vctrs                    0.6.3     2023-06-14 [2] RSPM (R 4.3.0)
##  vipor                    0.4.5     2017-03-22 [1] RSPM (R 4.3.0)
##  viridis                  0.6.4     2023-07-22 [1] RSPM (R 4.3.0)
##  viridisLite              0.4.2     2023-05-02 [1] RSPM (R 4.3.0)
##  withr                    2.5.0     2022-03-03 [2] RSPM (R 4.3.0)
##  xfun                     0.40      2023-08-09 [2] RSPM (R 4.3.0)
##  XML                      3.99-0.14 2023-03-19 [1] RSPM (R 4.3.0)
##  xtable                   1.8-4     2019-04-21 [2] RSPM (R 4.3.0)
##  XVector                  0.40.0    2023-04-25 [1] Bioconductor
##  yaml                     2.3.7     2023-01-23 [2] RSPM (R 4.3.0)
##  zlibbioc                 1.46.0    2023-04-25 [1] Bioconductor
## 
##  [1] /__w/_temp/Library
##  [2] /usr/local/lib/R/site-library
##  [3] /usr/local/lib/R/library
## 
## ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</code></pre>

</div>
</div>



<h3> References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Maynard2021" class="csl-entry">
Maynard, Kristen R., Leonardo Collado-Torres, Lukas M. Weber, Cedric Uytingco, Brianna K. Barry, Stephen R. Williams, Joseph L. Catallini II, et al. 2021. <span>“Transcriptome-Scale Spatial Gene Expression in the Human Dorsolateral Prefrontal Cortex.”</span> <em>Nature Neuroscience</em>. <a href="https://doi.org/10.1038/s41593-020-00787-0">https://doi.org/10.1038/s41593-020-00787-0</a>.
</div>
<div id="ref-Pardo2021" class="csl-entry">
Pardo, Brenda, Abby Spangler, Lukas M. Weber, Stephanie C. Hicks, Andrew E. Jaffe, Keri Martinowich, Kristen R. Maynard, and Leonardo Collado-Torres. 2021. <span>“spatialLIBD: An r/Bioconductor Package to Visualize Spatially-Resolved Transcriptomics Data.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2021.04.29.440149">https://doi.org/10.1101/2021.04.29.440149</a>.
</div>
</div>
<div class="footnotes">
<hr />
<ol start="1">
<li id="fn1"><p>Check <a href="https://github.com/LieberInstitute/spatialLIBD#shiny-website-mirrors" class="uri">https://github.com/LieberInstitute/spatialLIBD#shiny-website-mirrors</a> in case you need to use a mirror. <code>shiny</code>-powered websites work best on browsers such as Google Chrome and Mozilla Firefox, among others.<a href="spatiallibd-workflow.html#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>Originally available at <a href="https://github.com/LieberInstitute/spatialLIBD/blob/master/inst/app/www/documentation_spe.md#slides-and-videos" class="uri">https://github.com/LieberInstitute/spatialLIBD/blob/master/inst/app/www/documentation_spe.md#slides-and-videos</a>.<a href="spatiallibd-workflow.html#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>This book is under development, so right now you actually need to use <code>bioc-devel</code>. Once this book is submitted to Bioconductor, then it’ll work with <code>bioc-release</code>. TODO: remove this comment when the book is available on <code>bioc-release</code>.<a href="spatiallibd-workflow.html#fnref3" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="mouse-coronal-workflow.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="related-resources.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
},
"lib_dir": "book_assets"
});
});
</script>

</body>

</html>
