# Normalization

## Background

Here we apply normalization methods developed for scRNA-seq data, treating each spot as equivalent to one cell.


## Previous steps

*Code to run steps from the previous chapters to generate the `SpatialExperiment` object required for this chapter.*

```{r}
# clear workspace from previous chapters
rm(list = ls(all = TRUE))
```

```{r, message=FALSE, results='hide'}
# LOAD DATA

library(SpatialExperiment)
library(STexampleData)
spe <- Visium_humanDLPFC()

# QUALITY CONTROL (QC)

library(scater)
# subset to keep only spots over tissue
spe <- spe[, colData(spe)$in_tissue == 1]
# identify mitochondrial genes
is_mito <- grepl("(^MT-)|(^mt-)", rowData(spe)$gene_name)
# calculate per-spot QC metrics
spe <- addPerCellQC(spe, subsets = list(mito = is_mito))
# select QC thresholds
qc_lib_size <- colData(spe)$sum < 500
qc_detected <- colData(spe)$detected < 250
qc_mito <- colData(spe)$subsets_mito_percent > 30
qc_cell_count <- colData(spe)$cell_count > 12
# combined set of discarded spots
discard <- qc_lib_size | qc_detected | qc_mito | qc_cell_count
colData(spe)$discard <- discard
# filter low-quality spots
spe <- spe[, !colData(spe)$discard]
```


## Log-transformed normalized counts

Calculate log-transformed normalized counts (abbreviated as "logcounts") using library size factors.

We use normalization methods from `scater` [@McCarthy2017] and `scran` [@Lun2016], which were originally developed for single-cell RNA sequencing data. We make the assumption that these methods can be applied here by treating spots as equivalent to cells.

Since we have only a single sample, there are no blocking factors in the experimental design.

```{r, message=FALSE, fig.height=3.5}
library(scran)

# calculate library size factors
spe <- computeLibraryFactors(spe)

summary(sizeFactors(spe))
hist(sizeFactors(spe), breaks = 20)

# calculate logcounts (log-transformed normalized counts) and store in object
spe <- logNormCounts(spe)

# check
assayNames(spe)
dim(counts(spe))
dim(logcounts(spe))
```

